

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Python API &mdash; tskit 0.3.3.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Statistics" href="stats.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> tskit
          

          
            
            <img src="_static/tskit_logo_pale.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Python API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#trees-and-tree-sequences">Trees and tree sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-treesequence-class">The <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code> class</a></li>
<li class="toctree-l3"><a class="reference internal" href="#the-tree-class">The <code class="docutils literal notranslate"><span class="pre">Tree</span></code> class</a></li>
<li class="toctree-l3"><a class="reference internal" href="#constants">Constants</a></li>
<li class="toctree-l3"><a class="reference internal" href="#simple-container-classes">Simple container classes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#loading-data">Loading data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tables-and-table-collections">Tables and Table Collections</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-tablecollection-class">The <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> class</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tables">Tables</a></li>
<li class="toctree-l3"><a class="reference internal" href="#text-columns">Text columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#binary-columns">Binary columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#table-classes">Table classes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#table-functions">Table functions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#metadata-api">Metadata API</a></li>
<li class="toctree-l2"><a class="reference internal" href="#combinatorics-api">Combinatorics API</a></li>
<li class="toctree-l2"><a class="reference internal" href="#linkage-disequilibrium">Linkage disequilibrium</a></li>
<li class="toctree-l2"><a class="reference internal" href="#provenance">Provenance</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="stats.html">Statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="c-api.html">C API</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="data-model.html">Data model</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="combinatorics.html">Combinatorics</a></li>
<li class="toctree-l1"><a class="reference internal" href="provenance.html">Provenance</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">tskit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Python API</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/tskit-dev/tskit/blob/main/docs/python-api.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="python-api">
<span id="sec-python-api"></span><h1>Python API<a class="headerlink" href="#python-api" title="Permalink to this headline">¶</a></h1>
<p>This page provides detailed documentation for the <code class="docutils literal notranslate"><span class="pre">tskit</span></code> Python API.</p>
<div class="section" id="trees-and-tree-sequences">
<span id="sec-python-api-trees-and-tree-sequences"></span><h2>Trees and tree sequences<a class="headerlink" href="#trees-and-tree-sequences" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> class represents a sequence of correlated
evolutionary trees along a genome. The <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> class represents a
single tree in this sequence. These classes are the interfaces used to
interact with the trees and mutational information stored in a tree sequence,
for example as returned from a simulation or inferred from a set of DNA
sequences. This library also provides methods for loading stored tree
sequences, for example using <a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load()</span></code></a>.</p>
<div class="section" id="the-treesequence-class">
<h3>The <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code> class<a class="headerlink" href="#the-treesequence-class" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt id="tskit.TreeSequence">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">TreeSequence</code><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>A single tree sequence, as defined by the <a class="reference internal" href="data-model.html#sec-data-model"><span class="std std-ref">data model</span></a>.
A TreeSequence instance can be created from a set of
<a class="reference internal" href="data-model.html#sec-table-definitions"><span class="std std-ref">tables</span></a> using
<a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a>, or loaded from a set of text files
using <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a>, or loaded from a native binary file using
<a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load()</span></code></a>.</p>
<p>TreeSequences are immutable. To change the data held in a particular
tree sequence, first get the table information as a <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>
instance (using <a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dump_tables()</span></code></a>), edit those tables using the
<a class="reference internal" href="#sec-tables-api"><span class="std std-ref">tables api</span></a>, and create a new tree sequence using
<a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a>.</p>
<p>The <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method iterates over all trees in a tree sequence, and
the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> method iterates over all sites and their genotypes.</p>
<p><strong>Methods</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.Fst" title="tskit.TreeSequence.Fst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Fst</span></code></a>(sample_sets[, indexes, windows, mode, …])</p></td>
<td><p>Computes “windowed” Fst between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.Tajimas_D" title="tskit.TreeSequence.Tajimas_D"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tajimas_D</span></code></a>([sample_sets, windows, mode])</p></td>
<td><p>Computes Tajima’s D of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> in windows.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.Y1" title="tskit.TreeSequence.Y1"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Y1</span></code></a>(sample_sets[, windows, mode, span_normalise])</p></td>
<td><p>Computes the ‘Y1’ statistic within each of the sets of nodes given by <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.Y2" title="tskit.TreeSequence.Y2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Y2</span></code></a>(sample_sets[, indexes, windows, mode, …])</p></td>
<td><p>Computes the ‘Y2’ statistic between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.Y3" title="tskit.TreeSequence.Y3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Y3</span></code></a>(sample_sets[, indexes, windows, mode, …])</p></td>
<td><p>Computes the ‘Y’ statistic between triples of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.allele_frequency_spectrum" title="tskit.TreeSequence.allele_frequency_spectrum"><code class="xref py py-obj docutils literal notranslate"><span class="pre">allele_frequency_spectrum</span></code></a>([sample_sets, …])</p></td>
<td><p>Computes the allele frequency spectrum (AFS) in windows across the genome for with respect to the specified <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.aslist" title="tskit.TreeSequence.aslist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aslist</span></code></a>(**kwargs)</p></td>
<td><p>Returns the trees in this tree sequence as a list.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.at" title="tskit.TreeSequence.at"><code class="xref py py-obj docutils literal notranslate"><span class="pre">at</span></code></a>(position, **kwargs)</p></td>
<td><p>Returns the tree covering the specified genomic location.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.at_index" title="tskit.TreeSequence.at_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">at_index</span></code></a>(index, **kwargs)</p></td>
<td><p>Returns the tree at the specified index.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.breakpoints" title="tskit.TreeSequence.breakpoints"><code class="xref py py-obj docutils literal notranslate"><span class="pre">breakpoints</span></code></a>([as_array])</p></td>
<td><p>Returns the breakpoints along the chromosome, including the two extreme points 0 and L.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.coiterate" title="tskit.TreeSequence.coiterate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">coiterate</span></code></a>(other, **kwargs)</p></td>
<td><p>Returns an iterator over the pairs of trees for each distinct interval in the specified pair of tree sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.count_topologies" title="tskit.TreeSequence.count_topologies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_topologies</span></code></a>([sample_sets])</p></td>
<td><p>Returns a generator that produces the same distribution of topologies as <a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a> but sequentially for every tree in a tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.delete_intervals" title="tskit.TreeSequence.delete_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_intervals</span></code></a>(intervals[, simplify, …])</p></td>
<td><p>Returns a copy of this tree sequence for which information in the specified list of genomic intervals has been deleted.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.delete_sites" title="tskit.TreeSequence.delete_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_sites</span></code></a>(site_ids[, record_provenance])</p></td>
<td><p>Returns a copy of this tree sequence with the specified sites (and their associated mutations) entirely removed.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">divergence</span></code></a>(sample_sets[, indexes, windows, …])</p></td>
<td><p>Computes mean genetic divergence between (and within) pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diversity</span></code></a>([sample_sets, windows, mode, …])</p></td>
<td><p>Computes mean genetic diversity (also knowns as “Tajima’s pi”) in each of the sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.draw_svg" title="tskit.TreeSequence.draw_svg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">draw_svg</span></code></a>([path, size, x_scale, …])</p></td>
<td><p>Return an SVG representation of a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.dump" title="tskit.TreeSequence.dump"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dump</span></code></a>(file_or_path)</p></td>
<td><p>Writes the tree sequence to the specified path or file object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dump_tables</span></code></a>()</p></td>
<td><p>A copy of the tables defining this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.dump_text" title="tskit.TreeSequence.dump_text"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dump_text</span></code></a>([nodes, edges, sites, mutations, …])</p></td>
<td><p>Writes a text representation of the tables underlying the tree sequence to the specified connections.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.edge" title="tskit.TreeSequence.edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edge</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.edge_diffs" title="tskit.TreeSequence.edge_diffs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edge_diffs</span></code></a>([include_terminal])</p></td>
<td><p>Returns an iterator over all the edges that are inserted and removed to build the trees as we move from left-to-right along the tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.edges" title="tskit.TreeSequence.edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edges</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.equals" title="tskit.TreeSequence.equals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">equals</span></code></a>(other, *[, ignore_metadata, …])</p></td>
<td><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.f2" title="tskit.TreeSequence.f2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">f2</span></code></a>(sample_sets[, indexes, windows, mode, …])</p></td>
<td><p>Computes Patterson’s f3 statistic between two groups of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.f3" title="tskit.TreeSequence.f3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">f3</span></code></a>(sample_sets[, indexes, windows, mode, …])</p></td>
<td><p>Computes Patterson’s f3 statistic between three groups of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-obj docutils literal notranslate"><span class="pre">f4</span></code></a>(sample_sets[, indexes, windows, mode, …])</p></td>
<td><p>Computes Patterson’s f4 statistic between four groups of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.first" title="tskit.TreeSequence.first"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first</span></code></a>(**kwargs)</p></td>
<td><p>Returns the first tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.genealogical_nearest_neighbours" title="tskit.TreeSequence.genealogical_nearest_neighbours"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genealogical_nearest_neighbours</span></code></a>(focal, …)</p></td>
<td><p>Return the genealogical nearest neighbours (GNN) proportions for the given focal nodes, with reference to two or more sets of interest, averaged over all trees in the tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">general_stat</span></code></a>(W, f, output_dim[, windows, …])</p></td>
<td><p>Compute a windowed statistic from weights and a summary function.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genetic_relatedness</span></code></a>(sample_sets[, indexes, …])</p></td>
<td><p>Computes genetic relatedness between (and within) pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.genotype_matrix" title="tskit.TreeSequence.genotype_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genotype_matrix</span></code></a>(*[, isolated_as_missing, …])</p></td>
<td><p>Returns an <span class="math notranslate nohighlight">\(m \times n\)</span> numpy array of the genotypes in this tree sequence, where <span class="math notranslate nohighlight">\(m\)</span> is the number of sites and <span class="math notranslate nohighlight">\(n\)</span> the number of samples.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.haplotypes" title="tskit.TreeSequence.haplotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">haplotypes</span></code></a>(*[, isolated_as_missing, …])</p></td>
<td><p>Returns an iterator over the strings of haplotypes that result from the trees and mutations in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.individual" title="tskit.TreeSequence.individual"><code class="xref py py-obj docutils literal notranslate"><span class="pre">individual</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.individuals" title="tskit.TreeSequence.individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">individuals</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.kc_distance" title="tskit.TreeSequence.kc_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kc_distance</span></code></a>(other[, lambda_])</p></td>
<td><p>Returns the average <a class="reference internal" href="#tskit.Tree.kc_distance" title="tskit.Tree.kc_distance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.kc_distance()</span></code></a> between pairs of trees along the sequence whose intervals overlap.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">keep_intervals</span></code></a>(intervals[, simplify, …])</p></td>
<td><p>Returns a copy of this tree sequence which includes only information in the specified list of genomic intervals.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.last" title="tskit.TreeSequence.last"><code class="xref py py-obj docutils literal notranslate"><span class="pre">last</span></code></a>(**kwargs)</p></td>
<td><p>Returns the last tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ltrim</span></code></a>([record_provenance])</p></td>
<td><p>Returns a copy of this tree sequence with a potentially changed coordinate system, such that empty regions (i.e.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mean_descendants" title="tskit.TreeSequence.mean_descendants"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mean_descendants</span></code></a>(sample_sets)</p></td>
<td><p>Computes for every node the mean number of samples in each of the <cite>sample_sets</cite> that descend from that node, averaged over the portions of the genome for which the node is ancestral to <em>any</em> sample.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.migration" title="tskit.TreeSequence.migration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">migration</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations" title="tskit.TreeSequence.migrations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">migrations</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutation" title="tskit.TreeSequence.mutation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mutation</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations" title="tskit.TreeSequence.mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mutations</span></code></a>()</p></td>
<td><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">node</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.nodes" title="tskit.TreeSequence.nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nodes</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.pairwise_diversity" title="tskit.TreeSequence.pairwise_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pairwise_diversity</span></code></a>([samples])</p></td>
<td><p>Returns the pairwise nucleotide site diversity, the average number of sites that differ between a randomly chosen pair of samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.population" title="tskit.TreeSequence.population"><code class="xref py py-obj docutils literal notranslate"><span class="pre">population</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.populations" title="tskit.TreeSequence.populations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">populations</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.provenances" title="tskit.TreeSequence.provenances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">provenances</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rtrim</span></code></a>([record_provenance])</p></td>
<td><p>Returns a copy of this tree sequence with the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property reset so that the sequence ends at the end of the rightmost edge.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.sample_count_stat" title="tskit.TreeSequence.sample_count_stat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample_count_stat</span></code></a>(sample_sets, f, output_dim)</p></td>
<td><p>Compute a windowed statistic from sample counts and a summary function.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.samples" title="tskit.TreeSequence.samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">samples</span></code></a>([population, population_id])</p></td>
<td><p>Returns an array of the sample node IDs in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.segregating_sites" title="tskit.TreeSequence.segregating_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">segregating_sites</span></code></a>([sample_sets, windows, …])</p></td>
<td><p>Computes the density of segregating sites for each of the sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>, and related quantities.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-obj docutils literal notranslate"><span class="pre">simplify</span></code></a>([samples, map_nodes, …])</p></td>
<td><p>Returns a simplified tree sequence that retains only the history of the nodes given in the list <code class="docutils literal notranslate"><span class="pre">samples</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.site" title="tskit.TreeSequence.site"><code class="xref py py-obj docutils literal notranslate"><span class="pre">site</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.sites" title="tskit.TreeSequence.sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sites</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.subset" title="tskit.TreeSequence.subset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subset</span></code></a>(nodes[, record_provenance])</p></td>
<td><p>Returns a tree sequence modified to contain only the entries referring to the provided list of nodes, with nodes reordered according to the order they appear in the <code class="docutils literal notranslate"><span class="pre">nodes</span></code> argument.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.to_macs" title="tskit.TreeSequence.to_macs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_macs</span></code></a>()</p></td>
<td><p><p>Return a <a class="reference external" href="https://github.com/gchen98/macs">macs encoding</a> of this tree sequence.</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.to_nexus" title="tskit.TreeSequence.to_nexus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_nexus</span></code></a>([precision])</p></td>
<td><p><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Nexus_file">nexus encoding</a> of this tree sequence.</p>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.trait_correlation" title="tskit.TreeSequence.trait_correlation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trait_correlation</span></code></a>(W[, windows, mode, …])</p></td>
<td><p>Computes the mean squared correlations between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code> (the “phenotypes”) and inheritance along the tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trait_covariance</span></code></a>(W[, windows, mode, …])</p></td>
<td><p>Computes the mean squared covariances between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code> (the “phenotypes”) and inheritance along the tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.trait_linear_model" title="tskit.TreeSequence.trait_linear_model"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trait_linear_model</span></code></a>(W[, Z, windows, mode, …])</p></td>
<td><p>Finds the relationship between trait and genotype after accounting for covariates.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.trait_regression" title="tskit.TreeSequence.trait_regression"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trait_regression</span></code></a>(*args, **kwargs)</p></td>
<td><p>Deprecated synonym for <a class="reference internal" href="#tskit.TreeSequence.trait_linear_model" title="tskit.TreeSequence.trait_linear_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_linear_model</span></code></a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trees</span></code></a>([tracked_samples, sample_lists, …])</p></td>
<td><p>Returns an iterator over the trees in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.trim" title="tskit.TreeSequence.trim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim</span></code></a>([record_provenance])</p></td>
<td><p>Returns a copy of this tree sequence with any empty regions (i.e.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.union" title="tskit.TreeSequence.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></a>(other, node_mapping[, …])</p></td>
<td><p>Returns an expanded tree sequence which contains the node-wise union of <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>, obtained by adding the non-shared portions of <code class="docutils literal notranslate"><span class="pre">other</span></code> onto <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variants</span></code></a>(*[, as_bytes, samples, …])</p></td>
<td><p>Returns an iterator over the variants in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.write_vcf" title="tskit.TreeSequence.write_vcf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_vcf</span></code></a>(output[, ploidy, contig_id, …])</p></td>
<td><p>Writes a VCF formatted file to the specified file-like object.</p></td>
</tr>
</tbody>
</table>
<p><strong>Attributes</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.max_root_time" title="tskit.TreeSequence.max_root_time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_root_time</span></code></a></p></td>
<td><p>Returns time of the oldest root in any of the trees in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.metadata" title="tskit.TreeSequence.metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata</span></code></a></p></td>
<td><p>The decoded metadata for this TreeSequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.metadata_schema" title="tskit.TreeSequence.metadata_schema"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata_schema</span></code></a></p></td>
<td><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this TreeSequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.nbytes" title="tskit.TreeSequence.nbytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nbytes</span></code></a></p></td>
<td><p>Returns the total number of bytes required to store the data in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_edges" title="tskit.TreeSequence.num_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_edges</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_individuals" title="tskit.TreeSequence.num_individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_individuals</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_migrations" title="tskit.TreeSequence.num_migrations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_migrations</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_mutations" title="tskit.TreeSequence.num_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_mutations</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_nodes" title="tskit.TreeSequence.num_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_nodes</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_populations" title="tskit.TreeSequence.num_populations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_populations</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_provenances" title="tskit.TreeSequence.num_provenances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_provenances</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_samples" title="tskit.TreeSequence.num_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_samples</span></code></a></p></td>
<td><p>Returns the number of samples in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_sites" title="tskit.TreeSequence.num_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_sites</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_trees" title="tskit.TreeSequence.num_trees"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_trees</span></code></a></p></td>
<td><p>Returns the number of distinct trees in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence_length</span></code></a></p></td>
<td><p>Returns the sequence length in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.table_metadata_schemas" title="tskit.TreeSequence.table_metadata_schemas"><code class="xref py py-obj docutils literal notranslate"><span class="pre">table_metadata_schemas</span></code></a></p></td>
<td><p>The set of metadata schemas for the tables in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.tables" title="tskit.TreeSequence.tables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tables</span></code></a></p></td>
<td><p>A copy of the tables underlying this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.tables_dict" title="tskit.TreeSequence.tables_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tables_dict</span></code></a></p></td>
<td><p>Returns a dictionary mapping names to tables in the underlying <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="tskit.TreeSequence.Fst">
<code class="sig-name descname">Fst</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Fst"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.Fst" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes “windowed” Fst between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>For sample sets <code class="docutils literal notranslate"><span class="pre">X</span></code> and <code class="docutils literal notranslate"><span class="pre">Y</span></code>, if <code class="docutils literal notranslate"><span class="pre">d(X,</span> <span class="pre">Y)</span></code> is the
<a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence</span></code></a>
between <code class="docutils literal notranslate"><span class="pre">X</span></code> and <code class="docutils literal notranslate"><span class="pre">Y</span></code>, and <code class="docutils literal notranslate"><span class="pre">d(X)</span></code> is the
<a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity</span></code></a> of <code class="docutils literal notranslate"><span class="pre">X</span></code>, then what is
computed is</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Fst</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">d</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">(</span><span class="n">Y</span><span class="p">))</span> <span class="o">/</span> <span class="p">(</span><span class="n">d</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">d</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">(</span><span class="n">Y</span><span class="p">))</span>
</pre></div>
</div>
<p>What is computed for diversity and divergence depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>;
see those functions for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 2-tuples.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.Tajimas_D">
<code class="sig-name descname">Tajimas_D</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Tajimas_D"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.Tajimas_D" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes Tajima’s D of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> in windows.
Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>, <a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">1</span></code> sample sets at a
time. For a sample set <code class="docutils literal notranslate"><span class="pre">X</span></code> of <code class="docutils literal notranslate"><span class="pre">n</span></code> nodes, if and <code class="docutils literal notranslate"><span class="pre">T</span></code> is the mean
number of pairwise differing sites in <code class="docutils literal notranslate"><span class="pre">X</span></code> and <code class="docutils literal notranslate"><span class="pre">S</span></code> is the number of
sites segregating in <code class="docutils literal notranslate"><span class="pre">X</span></code> (computed with <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity</span></code></a> and <a class="reference internal" href="#tskit.TreeSequence.segregating_sites" title="tskit.TreeSequence.segregating_sites"><code class="xref py py-meth docutils literal notranslate"><span class="pre">segregating</span> <span class="pre">sites</span></code></a>, respectively, both not span
normalised), then Tajima’s D is</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">D</span> <span class="o">=</span> <span class="p">(</span><span class="n">T</span> <span class="o">-</span> <span class="n">S</span> <span class="o">/</span> <span class="n">h</span><span class="p">)</span> <span class="o">/</span> <span class="n">sqrt</span><span class="p">(</span><span class="n">a</span> <span class="o">*</span> <span class="n">S</span> <span class="o">+</span> <span class="p">(</span><span class="n">b</span> <span class="o">/</span> <span class="n">c</span><span class="p">)</span> <span class="o">*</span> <span class="n">S</span> <span class="o">*</span> <span class="p">(</span><span class="n">S</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
<span class="n">h</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">+</span> <span class="o">...</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">g</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="o">...</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>
<span class="n">a</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">h</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">h</span> <span class="o">**</span> <span class="mi">2</span>
<span class="n">b</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">n</span> <span class="o">+</span> <span class="mi">3</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mi">9</span> <span class="o">*</span> <span class="n">n</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="o">-</span> <span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">h</span> <span class="o">*</span> <span class="n">n</span><span class="p">)</span> <span class="o">+</span> <span class="n">g</span> <span class="o">/</span> <span class="n">h</span> <span class="o">**</span> <span class="mi">2</span>
<span class="n">c</span> <span class="o">=</span> <span class="n">h</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">g</span>
</pre></div>
</div>
<p>What is computed for diversity and divergence depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>;
see those functions for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.Y1">
<code class="sig-name descname">Y1</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Y1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.Y1" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the ‘Y1’ statistic within each of the sets of nodes given by
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>, <a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">1</span></code> sample set at a time.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is computed exactly as
<code class="docutils literal notranslate"><span class="pre">Y3</span></code>, except that the average is across a randomly chosen trio of
samples <code class="docutils literal notranslate"><span class="pre">(a1,</span> <span class="pre">a2,</span> <span class="pre">a3)</span></code> all chosen without replacement from the same
sample set. See <a class="reference internal" href="#tskit.TreeSequence.Y3" title="tskit.TreeSequence.Y3"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Y3</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.Y2">
<code class="sig-name descname">Y2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Y2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.Y2" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the ‘Y2’ statistic between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is computed exactly as
<code class="docutils literal notranslate"><span class="pre">Y3</span></code>, except that the average across randomly chosen trios of samples
<code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b1,</span> <span class="pre">b2)</span></code>, where <code class="docutils literal notranslate"><span class="pre">a</span></code> is chosen from the first sample set, and
<code class="docutils literal notranslate"><span class="pre">b1,</span> <span class="pre">b2</span></code> are chosen (without replacement) from the second sample set.
See <a class="reference internal" href="#tskit.TreeSequence.Y3" title="tskit.TreeSequence.Y3"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Y3</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.Y3">
<code class="sig-name descname">Y3</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Y3"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.Y3" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the ‘Y’ statistic between triples of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">3</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is an average across
randomly chosen trios of samples <code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b,</span> <span class="pre">c)</span></code>, one from each sample set:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>The average density of sites at which <code class="docutils literal notranslate"><span class="pre">a</span></code> differs from <code class="docutils literal notranslate"><span class="pre">b</span></code> and
<code class="docutils literal notranslate"><span class="pre">c</span></code>, per unit of chromosome length.</p>
</dd>
<dt>“branch”</dt><dd><p>The average length of all branches that separate <code class="docutils literal notranslate"><span class="pre">a</span></code> from <code class="docutils literal notranslate"><span class="pre">b</span></code>
and <code class="docutils literal notranslate"><span class="pre">c</span></code> (in units of time).</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the average proportion of the window on which <code class="docutils literal notranslate"><span class="pre">a</span></code>
inherits from that node but <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> do not, or vice-versa.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 3-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.allele_frequency_spectrum">
<code class="sig-name descname">allele_frequency_spectrum</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">polarised</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.allele_frequency_spectrum"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.allele_frequency_spectrum" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the allele frequency spectrum (AFS) in windows across the genome for
with respect to the specified <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-sample-sets"><span class="std std-ref">sample sets</span></a>,
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
<a class="reference internal" href="stats.html#sec-stats-polarisation"><span class="std std-ref">polarised</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
and see <a class="reference internal" href="tutorial.html#sec-tutorial-afs"><span class="std std-ref">Allele frequency spectra</span></a> for examples of how to use this method.</p>
<p>Similar to other windowed stats, the first dimension in the returned array
corresponds to windows, such that <code class="docutils literal notranslate"><span class="pre">result[i]</span></code> is the AFS in the ith
window. The AFS in each window is a k-dimensional numpy array, where k is
the number of input sample sets, such that <code class="docutils literal notranslate"><span class="pre">result[i,</span> <span class="pre">j0,</span> <span class="pre">j1,</span> <span class="pre">...]</span></code> is the
value associated with frequency <code class="docutils literal notranslate"><span class="pre">j0</span></code> in <code class="docutils literal notranslate"><span class="pre">sample_sets[0]</span></code>, <code class="docutils literal notranslate"><span class="pre">j1</span></code> in
<code class="docutils literal notranslate"><span class="pre">sample_sets[1]</span></code>, etc, in window <code class="docutils literal notranslate"><span class="pre">i</span></code>. From here, we will assume that
<code class="docutils literal notranslate"><span class="pre">afs</span></code> corresponds to the result in a single window, i.e.,
<code class="docutils literal notranslate"><span class="pre">afs</span> <span class="pre">=</span> <span class="pre">result[i]</span></code>.</p>
<p>If a single sample set is specified, the allele frequency spectrum within
this set is returned, such that <code class="docutils literal notranslate"><span class="pre">afs[j]</span></code> is the value associated with
frequency <code class="docutils literal notranslate"><span class="pre">j</span></code>. Thus, singletons are counted in <code class="docutils literal notranslate"><span class="pre">afs[1]</span></code>, doubletons in
<code class="docutils literal notranslate"><span class="pre">afs[2]</span></code>, and so on. The zeroth entry counts alleles or branches not
seen in the samples but that are polymorphic among the rest of the samples
of the tree sequence; likewise, the last entry counts alleles fixed in
the sample set but polymorphic in the entire set of samples. Please see
the <a class="reference internal" href="tutorial.html#sec-tutorial-afs-zeroth-entry"><span class="std std-ref">Zeroth and final entries in the AFS</span></a> for an illustration.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Please note that singletons are <strong>not</strong> counted in the initial
entry in each AFS array (i.e., <code class="docutils literal notranslate"><span class="pre">afs[0]</span></code>), but in <code class="docutils literal notranslate"><span class="pre">afs[1]</span></code>.</p>
</div>
<p>If <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> is None (the default), the allele frequency spectrum
for all samples in the tree sequence is returned.</p>
<p>If more than one sample set is specified, the <strong>joint</strong> allele frequency
spectrum within windows is returned. For example, if we set
<code class="docutils literal notranslate"><span class="pre">sample_sets</span> <span class="pre">=</span> <span class="pre">[S0,</span> <span class="pre">S1]</span></code>, then afs[1, 2] counts the number of sites that
are singletons within S0 and doubletons in S1. The dimensions of the
output array will be <code class="docutils literal notranslate"><span class="pre">[num_windows]</span> <span class="pre">+</span> <span class="pre">[1</span> <span class="pre">+</span> <span class="pre">len(S)</span> <span class="pre">for</span> <span class="pre">S</span> <span class="pre">in</span> <span class="pre">sample_sets]</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">polarised</span></code> is False (the default) the AFS will be <em>folded</em>, so that
the counts do not depend on knowing which allele is ancestral. If folded,
the frequency spectrum for a single sample set <code class="docutils literal notranslate"><span class="pre">S</span></code> has <code class="docutils literal notranslate"><span class="pre">afs[j]</span> <span class="pre">=</span> <span class="pre">0</span></code> for
all <code class="docutils literal notranslate"><span class="pre">j</span> <span class="pre">&gt;</span> <span class="pre">len(S)</span> <span class="pre">/</span> <span class="pre">2</span></code>, so that alleles at frequency <code class="docutils literal notranslate"><span class="pre">j</span></code> and <code class="docutils literal notranslate"><span class="pre">len(S)</span> <span class="pre">-</span> <span class="pre">j</span></code>
both add to the same entry. If there is more than one sample set, the
returned array is “lower triangular” in a similar way. For more details,
especially about handling of multiallelic sites, see <a class="reference internal" href="stats.html#sec-stats-afs"><span class="std std-ref">Allele frequency spectrum</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>The number of alleles at a given frequency within the specified sample
sets for each window, per unit of sequence length. To obtain the total
number of alleles, set <code class="docutils literal notranslate"><span class="pre">span_normalise</span></code> to False.</p>
</dd>
<dt>“branch”</dt><dd><p>The total length of branches in the trees subtended by subsets of the
specified sample sets, per unit of sequence length. To obtain the
total, set <code class="docutils literal notranslate"><span class="pre">span_normalise</span></code> to False.</p>
</dd>
<dt>“node”</dt><dd><p>Not supported for this method (raises a ValueError).</p>
</dd>
</dl>
<p>For example, suppose that <cite>S0</cite> is a list of 5 sample IDs, and <cite>S1</cite> is
a list of 3 other sample IDs. Then <cite>afs = ts.allele_frequency_spectrum([S0, S1],
mode=”site”, span_normalise=False)</cite> will be a 5x3 numpy array, and if
there are six alleles that are present in only one sample of <cite>S0</cite> but
two samples of <cite>S1</cite>, then <cite>afs[1,2]</cite> will be equal to 6.  Similarly,
<cite>branch_afs = ts.allele_frequency_spectrum([S0, S1], mode=”branch”,
span_normalise=False)</cite> will also be a 5x3 array, and <cite>branch_afs[1,2]</cite>
will be the total area (i.e., length times span) of all branches that
are above exactly one sample of <cite>S0</cite> and two samples of <cite>S1</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of samples to compute the joint allele frequency</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between windows
along the genome.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A (k + 1) dimensional numpy array, where k is the number of sample
sets specified.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.aslist">
<code class="sig-name descname">aslist</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.aslist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.aslist" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the trees in this tree sequence as a list. Each tree is
represented by a different instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. As such, this
method is inefficient and may use a large amount of memory, and should
not be used when performance is a consideration. The <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a>
method is the recommended way to efficiently iterate over the trees
in a tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>**kwargs</strong> – Further arguments used as parameters when constructing the
returned trees. For example <code class="docutils literal notranslate"><span class="pre">ts.aslist(sample_lists=True)</span></code> will result
in a list of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instances created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of the trees in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.at">
<code class="sig-name descname">at</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.at"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.at" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the tree covering the specified genomic location. The returned tree
will have <code class="docutils literal notranslate"><span class="pre">tree.interval.left</span></code> &lt;= <code class="docutils literal notranslate"><span class="pre">position</span></code> &lt; <code class="docutils literal notranslate"><span class="pre">tree.interval.right</span></code>.
See also <a class="reference internal" href="#tskit.Tree.seek" title="tskit.Tree.seek"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – A genomic location.</p></li>
<li><p><strong>**kwargs</strong> – Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.at(2.5,</span> <span class="pre">sample_lists=True)</span></code> will
result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> positioned to cover the specified
genomic location.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.at_index">
<code class="sig-name descname">at_index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.at_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.at_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the tree at the specified index. See also <a class="reference internal" href="#tskit.Tree.seek_index" title="tskit.Tree.seek_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek_index()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The index of the required tree.</p></li>
<li><p><strong>**kwargs</strong> – Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.at_index(4,</span> <span class="pre">sample_lists=True)</span></code>
will result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> positioned at the specified index.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.breakpoints">
<code class="sig-name descname">breakpoints</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">as_array</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.breakpoints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.breakpoints" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the breakpoints along the chromosome, including the two extreme points
0 and L. This is equivalent to</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">iter</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">t</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">right</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">trees</span><span class="p">()])</span>
</pre></div>
</div>
<p>By default we return an iterator over the breakpoints as Python float objects;
if <code class="docutils literal notranslate"><span class="pre">as_array</span></code> is True we return them as a numpy array.</p>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">as_array</span></code> form will be more efficient and convenient in most
cases; the default iterator behaviour is mainly kept to ensure compatability
with existing code.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>as_array</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, return the breakpoints as a numpy array.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The breakpoints defined by the tree intervals along the sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.9)">collections.abc.Iterable</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.coiterate">
<code class="sig-name descname">coiterate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.coiterate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.coiterate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over the pairs of trees for each distinct
interval in the specified pair of tree sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>) – The other tree sequence from which to take trees. The
sequence length must be the same as the current tree sequence.</p></li>
<li><p><strong>**kwargs</strong> – Further named arguments that will be passed to the
<a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method when constructing the returned trees.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An iterator returning successive tuples of the form
<code class="docutils literal notranslate"><span class="pre">(interval,</span> <span class="pre">tree_self,</span> <span class="pre">tree_other)</span></code>. For example, the first item returned
will consist of an tuple of the initial interval, the first tree of the
current tree sequence, and the first tree of the <code class="docutils literal notranslate"><span class="pre">other</span></code> tree sequence;
the <code class="docutils literal notranslate"><span class="pre">.left</span></code> attribute of the initial interval will be 0 and the <code class="docutils literal notranslate"><span class="pre">.right</span></code>
attribute will be the smallest non-zero breakpoint of the 2 tree sequences.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>iter(<a class="reference internal" href="#tskit.Interval" title="tskit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>, <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>, <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.count_topologies">
<code class="sig-name descname">count_topologies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.count_topologies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.count_topologies" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a generator that produces the same distribution of topologies as
<a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a> but sequentially for every tree in a tree
sequence. For use on a tree sequence this method is much faster than
computing the result independently per tree.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>iter(<a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TopologyCounter</span></code></a>)</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If nodes in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> are invalid or are
internal samples.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.delete_intervals">
<code class="sig-name descname">delete_intervals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">intervals</span></em>, <em class="sig-param"><span class="n">simplify</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.delete_intervals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.delete_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of this tree sequence for which information in the
specified list of genomic intervals has been deleted. Edges spanning these
intervals are truncated or deleted, and sites and mutations falling within
them are discarded. Note that it is the information in the intervals that
is deleted, not the intervals themselves, so in particular, all samples
will be isolated in the deleted intervals.</p>
<p>Note that node IDs may change as a result of this operation,
as by default <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> is called on the returned tree sequence
to remove redundant nodes. If you wish to map node IDs onto the same
nodes before and after this method has been called, specify <code class="docutils literal notranslate"><span class="pre">simplify=False</span></code>.</p>
<p>See also <a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">keep_intervals()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ltrim()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rtrim()</span></code></a>, and
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>) – A list (start, end) pairs describing the
genomic intervals to delete. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, return a simplified tree sequence where nodes
no longer used are discarded. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation to the
provenance information of the returned tree sequence. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.delete_sites">
<code class="sig-name descname">delete_sites</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site_ids</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.delete_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.delete_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of this tree sequence with the specified sites (and their
associated mutations) entirely removed. The site IDs do not need to be in any
particular order, and specifying the same ID multiple times does not have any
effect (i.e., calling <code class="docutils literal notranslate"><span class="pre">tree_sequence.delete_sites([0,</span> <span class="pre">1,</span> <span class="pre">1])</span></code> has the same
effect as calling <code class="docutils literal notranslate"><span class="pre">tree_sequence.delete_sites([0,</span> <span class="pre">1])</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>]</em>) – A list of site IDs specifying the sites to remove.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation to the
provenance information of the returned tree sequence. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.divergence">
<code class="sig-name descname">divergence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.divergence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.divergence" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes mean genetic divergence between (and within) pairs of
sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>As a special case, an index <code class="docutils literal notranslate"><span class="pre">(j,</span> <span class="pre">j)</span></code> will compute the
<a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity</span></code></a> of <code class="docutils literal notranslate"><span class="pre">sample_set[j]</span></code>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>Mean pairwise genetic divergence: the average across distinct,
randomly chosen pairs of chromosomes (one from each sample set), of
the density of sites at which the two carry different alleles, per
unit of chromosome length.</p>
</dd>
<dt>“branch”</dt><dd><p>Mean distance in the tree: the average across distinct, randomly
chosen pairs of chromsomes (one from each sample set) and locations
in the window, of the mean distance in the tree between the two
samples (in units of time).</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the proportion of genome on which the node is an ancestor to
only one of a random pair (one from each sample set), averaged over
choices of pair.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.diversity">
<code class="sig-name descname">diversity</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.diversity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.diversity" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes mean genetic diversity (also knowns as “Tajima’s pi”) in each of the
sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>Note that this quantity can also be computed by the
<a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence</span></code></a> method.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>Mean pairwise genetic diversity: the average across distinct,
randomly chosen pairs of chromosomes, of the density of sites at
which the two carry different alleles, per unit of chromosome length.</p>
</dd>
<dt>“branch”</dt><dd><p>Mean distance in the tree: the average across distinct, randomly chosen pairs
of chromsomes and locations in the window, of the mean distance in the tree
between the two samples (in units of time).</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the proportion of genome on which the node is an ancestor to
only one of a random pair from the sample set, averaged over choices of pair.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A numpy array.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.draw_svg">
<code class="sig-name descname">draw_svg</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">x_scale</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">tree_height_scale</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mutation_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">root_svg_attributes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">style</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">force_root_branch</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.draw_svg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.draw_svg" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an SVG representation of a tree sequence.</p>
<p>When working in a Jupyter notebook, use the <code class="docutils literal notranslate"><span class="pre">IPython.display.SVG</span></code> function
to display the SVG output from this function inline in the notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">SVG</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
<p>The visual elements in the svg are
<a class="reference external" href="https://www.w3.org/TR/SVG2/struct.html#Groups">grouped</a>
for easy styling and manipulation. The entire visualization with trees and X
axis is contained within a group of class <code class="docutils literal notranslate"><span class="pre">tree-sequence</span></code>. Each tree in
the displayed tree sequence is contained in a group of class <code class="docutils literal notranslate"><span class="pre">tree</span></code>, as
described in <a class="reference internal" href="#tskit.Tree.draw_svg" title="tskit.Tree.draw_svg"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.draw_svg()</span></code></a>, so that visual elements pertaining to one
or more trees targetted as documented in that method. For instance, the
following style will change the colour of all the edges of the <em>initial</em>
tree in the sequence and hide the non-sample node labels in <em>all</em> the trees</p>
<div class="highlight-css notranslate"><div class="highlight"><pre><span></span><span class="p">.</span><span class="nc">tree</span><span class="p">.</span><span class="nc">t0</span> <span class="p">.</span><span class="nc">edge</span> <span class="p">{</span><span class="n">stroke</span><span class="p">:</span> <span class="kc">blue</span><span class="p">}</span>
<span class="p">.</span><span class="nc">tree</span> <span class="p">.</span><span class="nc">node</span><span class="p">:</span><span class="nd">not</span><span class="o">(</span><span class="p">.</span><span class="nc">sample</span><span class="o">)</span> <span class="o">&gt;</span> <span class="nt">text</span> <span class="p">{</span><span class="k">visibility</span><span class="p">:</span> <span class="kc">hidden</span><span class="p">}</span>
</pre></div>
</div>
<p>See <a class="reference internal" href="#tskit.Tree.draw_svg" title="tskit.Tree.draw_svg"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.draw_svg()</span></code></a> for further details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The path to the file to write the output. If None, do not write
to file.</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>)</em>) – A tuple of (width, height) giving the width and height of the
produced SVG drawing in abstract user units (usually interpreted as pixels on
display).</p></li>
<li><p><strong>x_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Control how the X axis is drawn. If “physical” (the default)
the axis scales linearly with physical distance along the sequence, and
background shading is used to indicate the position of the trees along the
sequence. If “treewise”, each axis tick corresponds to a tree boundary, which
are positioned evenly along the axis, so that the X axis is of variable scale
and no background scaling is required.</p></li>
<li><p><strong>tree_height_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Control how height values for nodes are computed.
If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;time&quot;</span></code>, node heights are proportional to their time
values (this is the default). If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;log_time&quot;</span></code>, node
heights are proportional to their log(time) values. If it is equal to
<code class="docutils literal notranslate"><span class="pre">&quot;rank&quot;</span></code>, node heights are spaced equally according to their ranked times.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – If specified, show custom labels for the nodes
(specified by ID) that are present in this map; any nodes not present will
not have a label.</p></li>
<li><p><strong>mutation_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – If specified, show custom labels for the
mutations (specified by ID) that are present in the map; any mutations
not present will not have a label.</p></li>
<li><p><strong>root_svg_attributes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – Additional attributes, such as an id, that will
be embedded in the root <code class="docutils literal notranslate"><span class="pre">&lt;svg&gt;</span></code> tag of the generated drawing.</p></li>
<li><p><strong>style</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A <a class="reference external" href="https://www.w3.org/TR/CSS21/syndata.htm">css string</a>
that will be included in the <code class="docutils literal notranslate"><span class="pre">&lt;style&gt;</span></code> tag of the generated svg.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string specifying the traversal type used to order the tips
in each tree, as detailed in <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a>. If None (default), use
the default order as described in that method.</p></li>
<li><p><strong>force_root_branch</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> plot a branch (edge) above every tree
root in the tree sequence. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) then only plot such
root branches if any root in the tree sequence has a mutation above it.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An SVG representation of a tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.dump">
<code class="sig-name descname">dump</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_or_path</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.dump"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.dump" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the tree sequence to the specified path or file object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_or_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The file object or path to write the TreeSequence to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.dump_tables">
<code class="sig-name descname">dump_tables</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.dump_tables"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.dump_tables" title="Permalink to this definition">¶</a></dt>
<dd><p>A copy of the tables defining this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> containing all tables underlying
the tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection">TableCollection</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.dump_text">
<code class="sig-name descname">dump_text</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nodes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">edges</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sites</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mutations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">individuals</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">populations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">provenances</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">precision</span><span class="o">=</span><span class="default_value">6</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em>, <em class="sig-param"><span class="n">base64_metadata</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.dump_text"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.dump_text" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a text representation of the tables underlying the tree sequence
to the specified connections.</p>
<p>If Base64 encoding is not used, then metadata will be saved directly, possibly
resulting in errors reading the tables back in if metadata includes whitespace.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object (having a .write() method) to
write the NodeTable to.</p></li>
<li><p><strong>edges</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object to write the EdgeTable to.</p></li>
<li><p><strong>sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object to write the SiteTable to.</p></li>
<li><p><strong>mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object to write the
MutationTable to.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object to write the
IndividualTable to.</p></li>
<li><p><strong>populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object to write the
PopulationTable to.</p></li>
<li><p><strong>provenances</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object to write the
ProvenanceTable to.</p></li>
<li><p><strong>precision</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of digits of precision.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.edge">
<code class="sig-name descname">edge</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge</span></a> in this tree sequence
with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Edge" title="tskit.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.edge_diffs">
<code class="sig-name descname">edge_diffs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_terminal</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.edge_diffs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.edge_diffs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over all the edges that are inserted and removed to
build the trees as we move from left-to-right along the tree sequence.
The iterator yields a sequence of 3-tuples, <code class="docutils literal notranslate"><span class="pre">(interval,</span> <span class="pre">edges_out,</span>
<span class="pre">edges_in)</span></code>. The <code class="docutils literal notranslate"><span class="pre">interval</span></code> is a pair <code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right)</span></code> representing
the genomic interval (see <a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Tree.interval</span></code></a>). The <code class="docutils literal notranslate"><span class="pre">edges_out</span></code>
value is a list of the edges that were just-removed to create the tree
covering the interval (hence <code class="docutils literal notranslate"><span class="pre">edges_out</span></code> will always be empty for the
first tree). The <code class="docutils literal notranslate"><span class="pre">edges_in</span></code> value is a list of edges that were just
inserted to construct the tree covering the current interval.</p>
<p>The edges returned within each <code class="docutils literal notranslate"><span class="pre">edges_in</span></code> list are ordered by ascending
time of the parent node, then ascending parent id, then ascending child id.
The edges within each <code class="docutils literal notranslate"><span class="pre">edges_out</span></code> list are the reverse order (e.g.
descending parent time, parent id, then child_id). This means that within
each list, edges with the same parent appear consecutively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>include_terminal</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If False (default), the iterator terminates
after the final interval in the tree sequence (i.e. it does not
report a final removal of all remaining edges), and the number
of iterations will be equal to the number of trees in the tree
sequence. If True, an additional iteration takes place, with the last
<code class="docutils literal notranslate"><span class="pre">edges_out</span></code> value reporting all the edges contained in the final
tree (with both <code class="docutils literal notranslate"><span class="pre">left</span></code> and <code class="docutils literal notranslate"><span class="pre">right</span></code> equal to the sequence length).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An iterator over the (interval, edges_out, edges_in) tuples.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">collections.abc.Iterable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.edges">
<code class="sig-name descname">edges</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a>
in this tree sequence. Edges are returned in the order required
for a <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">valid tree sequence</span></a>. So,
edges are guaranteed to be ordered such that (a) all parents with a
given ID are contiguous; (b) edges are returned in non-descreasing
order of parent time ago; (c) within the edges for a given parent, edges
are sorted first by child ID and then by left coordinate.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterable sequence of all edges.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Edge" title="tskit.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_ts_metadata</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_provenance</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_timestamps</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.equals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. Uses the underlying table equlity,
see <a class="reference internal" href="#tskit.TableCollection.equals" title="tskit.TableCollection.equals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.equals()</span></code></a> for details and options.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.f2">
<code class="sig-name descname">f2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.f2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.f2" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes Patterson’s f3 statistic between two groups of nodes from
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each works exactly as
<a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a>, except the average is across randomly
chosen set of four samples <code class="docutils literal notranslate"><span class="pre">(a1,</span> <span class="pre">b1;</span> <span class="pre">a2,</span> <span class="pre">b2)</span></code>, with <cite>a1</cite> and <cite>a2</cite>
both chosen (without replacement) from the first sample set and <code class="docutils literal notranslate"><span class="pre">b1</span></code>
and <code class="docutils literal notranslate"><span class="pre">b2</span></code> chosen randomly without replacement from the second sample
set. See <a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.f3">
<code class="sig-name descname">f3</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.f3"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.f3" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes Patterson’s f3 statistic between three groups of nodes from
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">3</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each works exactly as
<a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a>, except the average is across randomly
chosen set of four samples <code class="docutils literal notranslate"><span class="pre">(a1,</span> <span class="pre">b;</span> <span class="pre">a2,</span> <span class="pre">c)</span></code>, with <cite>a1</cite> and <cite>a2</cite> both
chosen (without replacement) from the first sample set. See
<a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 3-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.f4">
<code class="sig-name descname">f4</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.f4"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.f4" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes Patterson’s f4 statistic between four groups of nodes from
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">4</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is an average across
randomly chosen set of four samples <code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b;</span> <span class="pre">c,</span> <span class="pre">d)</span></code>, one from each sample set:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>The average density of sites at which <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> agree but
differs from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code>, minus the average density of sites at
which <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code> agree but differs from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code>, per
unit of chromosome length.</p>
</dd>
<dt>“branch”</dt><dd><p>The average length of all branches that separate <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code>
from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code>, minus the average length of all branches that
separate <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code> from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> (in units of time).</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the average proportion of the window on which <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code>
inherit from that node but <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code> do not, or vice-versa,
minus the average proportion of the window on which <code class="docutils literal notranslate"><span class="pre">a</span></code> anc <code class="docutils literal notranslate"><span class="pre">d</span></code>
inherit from that node but <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> do not, or vice-versa.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 4-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.first">
<code class="sig-name descname">first</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.first"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.first" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the first tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. To iterate over all
trees in the sequence, use the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>**kwargs</strong> – Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.first(sample_lists=True)</span></code> will
result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The first tree in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.genealogical_nearest_neighbours">
<code class="sig-name descname">genealogical_nearest_neighbours</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">focal</span></em>, <em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">num_threads</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genealogical_nearest_neighbours"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.genealogical_nearest_neighbours" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the genealogical nearest neighbours (GNN) proportions for the given
focal nodes, with reference to two or more sets of interest, averaged over all
trees in the tree sequence.</p>
<p>The GNN proportions for a focal node in a single tree are given by first finding
the most recent common ancestral node <span class="math notranslate nohighlight">\(a\)</span> between the focal node and any
other node present in the reference sets. The GNN proportion for a specific
reference set, <span class="math notranslate nohighlight">\(S\)</span> is the number of nodes in <span class="math notranslate nohighlight">\(S\)</span> that descend from
<span class="math notranslate nohighlight">\(a\)</span>, as a proportion of the total number of descendant nodes in any of the
reference sets.</p>
<p>For example, consider a case with 2 sample sets, <span class="math notranslate nohighlight">\(S_1\)</span> and <span class="math notranslate nohighlight">\(S_2\)</span>.
For a given tree, <span class="math notranslate nohighlight">\(a\)</span> is the node that includes at least one descendant in
<span class="math notranslate nohighlight">\(S_1\)</span> or <span class="math notranslate nohighlight">\(S_2\)</span> (not including the focal node). If the descendants of
<span class="math notranslate nohighlight">\(a\)</span> include some nodes in <span class="math notranslate nohighlight">\(S_1\)</span> but no nodes in <span class="math notranslate nohighlight">\(S_2\)</span>, then the
GNN proportions for that tree will be 100% <span class="math notranslate nohighlight">\(S_1\)</span> and 0% <span class="math notranslate nohighlight">\(S_2\)</span>, or
<span class="math notranslate nohighlight">\([1.0, 0.0]\)</span>.</p>
<p>For a given focal node, the GNN proportions returned by this function are an
average of the GNNs for each tree, weighted by the genomic distance spanned by
that tree.</p>
<p>For an precise mathematical definition of GNN, see <a class="reference external" href="https://doi.org/10.1101/458067">https://doi.org/10.1101/458067</a></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The reference sets need not include all the samples, hence the most
recent common ancestral node of the reference sets, <span class="math notranslate nohighlight">\(a\)</span>, need not be
the immediate ancestor of the focal node. If the reference sets only comprise
sequences from relatively distant individuals, the GNN statistic may end up
as a measure of comparatively distant ancestry, even for tree sequences that
contain many closely related individuals.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future. The long-term stable
API for this method will be consistent with other <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>focal</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of <span class="math notranslate nohighlight">\(n\)</span> nodes whose GNNs should be calculated.</p></li>
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of <span class="math notranslate nohighlight">\(m\)</span> lists of node IDs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An <span class="math notranslate nohighlight">\(n\)</span>  by <span class="math notranslate nohighlight">\(m\)</span> array of focal nodes by GNN proportions.
Every focal node corresponds to a row. The numbers in each
row corresponding to the GNN proportion for each of the passed-in reference
sets. Rows therefore sum to one.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.general_stat">
<code class="sig-name descname">general_stat</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">W</span></em>, <em class="sig-param"><span class="n">f</span></em>, <em class="sig-param"><span class="n">output_dim</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">polarised</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.general_stat"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.general_stat" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a windowed statistic from weights and a summary function.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
On each tree, this
propagates the weights <code class="docutils literal notranslate"><span class="pre">W</span></code> up the tree, so that the “weight” of each
node is the sum of the weights of all samples at or below the node.
Then the summary function <code class="docutils literal notranslate"><span class="pre">f</span></code> is applied to the weights, giving a
summary for each node in each tree. How this is then aggregated depends
on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>Adds together the total summary value across all alleles in each window.</p>
</dd>
<dt>“branch”</dt><dd><p>Adds together the summary value for each node, multiplied by the
length of the branch above the node and the span of the tree.</p>
</dd>
<dt>“node”</dt><dd><p>Returns each node’s summary value added across trees and multiplied
by the span of the tree.</p>
</dd>
</dl>
<p>Both the weights and the summary can be multidimensional: if <code class="docutils literal notranslate"><span class="pre">W</span></code> has <code class="docutils literal notranslate"><span class="pre">k</span></code>
columns, and <code class="docutils literal notranslate"><span class="pre">f</span></code> takes a <code class="docutils literal notranslate"><span class="pre">k</span></code>-vector and returns an <code class="docutils literal notranslate"><span class="pre">m</span></code>-vector,
then the output will be <code class="docutils literal notranslate"><span class="pre">m</span></code>-dimensional for each node or window (depending
on “mode”).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The summary function <code class="docutils literal notranslate"><span class="pre">f</span></code> should return zero when given both 0 and
the total weight (i.e., <code class="docutils literal notranslate"><span class="pre">f(0)</span> <span class="pre">=</span> <span class="pre">0</span></code> and <code class="docutils literal notranslate"><span class="pre">f(np.sum(W,</span> <span class="pre">axis=0))</span> <span class="pre">=</span> <span class="pre">0</span></code>),
unless <code class="docutils literal notranslate"><span class="pre">strict=False</span></code>.  This is necessary for the statistic to be
unaffected by parts of the tree sequence ancestral to none or all
of the samples, respectively.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – An array of values with one row for each sample and one
column for each weight.</p></li>
<li><p><strong>f</strong> – A function that takes a one-dimensional array of length
equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">W</span></code> and returns a one-dimensional
array.</p></li>
<li><p><strong>output_dim</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The length of <code class="docutils literal notranslate"><span class="pre">f</span></code>’s return value.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>polarised</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to leave the ancestral state out of computations:
see <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a> for more details.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to check that f(0) and f(total weight) are zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.genetic_relatedness">
<code class="sig-name descname">genetic_relatedness</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">indexes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">polarised</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">proportion</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genetic_relatedness"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.genetic_relatedness" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes genetic relatedness between (and within) pairs of
sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
<a class="reference internal" href="stats.html#sec-stats-polarisation"><span class="std std-ref">polarised</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>Number of pairwise allelic matches in the window between two
sample sets relative to the rest of the sample sets. To be precise,
let <cite>m(u,v)</cite> denote the total number of alleles shared between
nodes <cite>u</cite> and <cite>v</cite>, and let <cite>m(I,J)</cite> be the sum of <cite>m(u,v)</cite> over all
nodes <cite>u</cite> in sample set <cite>I</cite> and <cite>v</cite> in sample set <cite>J</cite>. Let <cite>S</cite> and
<cite>T</cite> be independently chosen sample sets. Then, for sample sets <cite>I</cite>
and <cite>J</cite>, this computes <cite>E[m(I,J) - m(I,S) - m(J,T) + m(S,T)]</cite>.
This can also be seen as the covariance of a quantitative trait
determined by additive contributions from the genomes in each
sample set. Let each allele be associated with an effect drawn from
a <cite>N(0,1/2)</cite> distribution, and let the trait value of a sample set
be the sum of its allele effects. Then, this computes the covariance
between the trait values of two sample sets. For example, to
compute covariance between the traits of diploid individuals, each
sample set would be the pair of genomes of each individual; if
<code class="docutils literal notranslate"><span class="pre">proportion=True</span></code>, this then corresponds to <span class="math notranslate nohighlight">\(K_{c0}\)</span> in
<a class="reference external" href="https://www.nature.com/articles/nrg3821">Speed &amp; Balding (2014)</a>.</p>
</dd>
<dt>“branch”</dt><dd><p>Total area of branches in the window ancestral to pairs of samples
in two sample sets relative to the rest of the sample sets. To be
precise, let <cite>B(u,v)</cite> denote the total area of all branches
ancestral to nodes <cite>u</cite> and <cite>v</cite>, and let <cite>B(I,J)</cite> be the sum of
<cite>B(u,v)</cite> over all nodes <cite>u</cite> in sample set <cite>I</cite> and <cite>v</cite> in sample set
<cite>J</cite>. Let <cite>S</cite> and <cite>T</cite> be two independently chosen sample sets. Then
for sample sets <cite>I</cite> and <cite>J</cite>, this computes
<cite>E[B(I,J) - B(I,S) - B(J,T) + B(S,T)]</cite>.</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the proportion of the window over which pairs of
samples in two sample sets are descendants, relative to the rest of
the sample sets. To be precise, for each node <cite>n</cite>, let <cite>N(u,v)</cite>
denote the proportion of the window over which samples <cite>u</cite> and <cite>v</cite>
are descendants of <cite>n</cite>, and let and let <cite>N(I,J)</cite> be the sum of
<cite>N(u,v)</cite> over all nodes <cite>u</cite> in sample set <cite>I</cite> and <cite>v</cite> in sample set
<cite>J</cite>. Let <cite>S</cite> and <cite>T</cite> be two independently chosen sample sets. Then
for sample sets <cite>I</cite> and <cite>J</cite>, this computes
<cite>E[N(I,J) - N(I,S) - N(J,T) + N(S,T)]</cite>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
<li><p><strong>proportion</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by
<a class="reference internal" href="#tskit.TreeSequence.segregating_sites" title="tskit.TreeSequence.segregating_sites"><code class="xref py py-meth docutils literal notranslate"><span class="pre">segregating_sites()</span></code></a>, called with the same <code class="docutils literal notranslate"><span class="pre">windows</span></code> and
<code class="docutils literal notranslate"><span class="pre">mode</span></code> (defaults to True). Note that this counts sites
that are segregating between <em>any</em> of the samples of <em>any</em> of the
sample sets (rather than segregating between all of the samples of
the tree sequence).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.genotype_matrix">
<code class="sig-name descname">genotype_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">isolated_as_missing</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alleles</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">impute_missing_data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genotype_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.genotype_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an <span class="math notranslate nohighlight">\(m \times n\)</span> numpy array of the genotypes in this
tree sequence, where <span class="math notranslate nohighlight">\(m\)</span> is the number of sites and <span class="math notranslate nohighlight">\(n\)</span>
the number of samples. The genotypes are the indexes into the array
of <code class="docutils literal notranslate"><span class="pre">alleles</span></code>, as described for the <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class.</p>
<p>If isolated samples are present at a given site without mutations above them,
they will be interpreted as <a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>
the genotypes array will contain a special value <a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a>
(-1) to identify these missing samples.</p>
<p>Such samples are treated as missing data by default, but if
<code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is set to to False, they will not be treated as missing,
and so assigned the ancestral state. This was the default behaviour in
versions prior to 0.2.0. Prior to 0.3.0 the <cite>impute_missing_data</cite>
argument controlled this behaviour.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method can consume a <strong>very large</strong> amount of memory! If
all genotypes are not needed at once, it is usually better to
access them sequentially using the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> iterator.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, the allele assigned to
missing samples (i.e., isolated samples without mutations) is
the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code>. If False, missing samples will be
assigned the ancestral state.
Default: True.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a>) – A tuple of strings describing the encoding of
alleles to genotype values. At least one allele must be provided.
If duplicate alleles are provided, output genotypes will always be
encoded as the first occurance of the allele. If None (the default),
the alleles are encoded as they are encountered during genotype
generation.</p></li>
<li><p><strong>impute_missing_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – <em>Deprecated in 0.3.0. Use ``isolated_as_missing``, but inverting value.
Will be removed in a future version</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The full matrix of genotypes.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a> (dtype=np.int8)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.haplotypes">
<code class="sig-name descname">haplotypes</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">isolated_as_missing</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">missing_data_character</span><span class="o">=</span><span class="default_value">'-'</span></em>, <em class="sig-param"><span class="n">impute_missing_data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.haplotypes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.haplotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over the strings of haplotypes that result from
the trees and mutations in this tree sequence. Each haplotype string
is guaranteed to be of the same length. A tree sequence with
<span class="math notranslate nohighlight">\(n\)</span> samples and <span class="math notranslate nohighlight">\(s\)</span> sites will return a total of <span class="math notranslate nohighlight">\(n\)</span>
strings of <span class="math notranslate nohighlight">\(s\)</span> alleles concatenated together, where an allele
consists of a single ascii character (tree sequences that include alleles
which are not a single character in length, or where the character is
non-ascii, will raise an error). The first string returned is the
haplotype for sample <code class="docutils literal notranslate"><span class="pre">0</span></code>, and so on.</p>
<p>The alleles at each site must be represented by single byte characters,
(i.e. variants must be single nucleotide polymorphisms, or SNPs), hence
the strings returned will all be of length <span class="math notranslate nohighlight">\(s\)</span>, and for a haplotype
<code class="docutils literal notranslate"><span class="pre">h</span></code>, the value of <code class="docutils literal notranslate"><span class="pre">h[j]</span></code> will be the observed allelic state
at site <code class="docutils literal notranslate"><span class="pre">j</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is True (the default), isolated samples without
mutations directly above them will be treated as
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a> and will be
represented in the string by the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code>. If
instead it is set to False, missing data will be assigned the ancestral state
(unless they have mutations directly above them, in which case they will take
the most recent derived mutational state for that node). This was the default
behaviour in versions prior to 0.2.0. Prior to 0.3.0 the <cite>impute_missing_data</cite>
argument controlled this behaviour.</p>
<p>See also the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> iterator for site-centric access
to sample genotypes.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>For large datasets, this method can consume a <strong>very large</strong> amount of
memory! To output all the sample data, it is more efficient to iterate
over sites rather than over samples. If you have a large dataset but only
want to output the haplotypes for a subset of samples, it may be worth
calling <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> to reduce tree sequence down to the required
samples before outputting haplotypes.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterator over the haplotype strings for the samples in
this tree sequence.</p>
</dd>
<dt class="field-even">Parameters</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, the allele assigned to
missing samples (i.e., isolated samples without mutations) is
the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code>. If False,
missing samples will be assigned the ancestral state.
Default: True.</p></li>
<li><p><strong>missing_data_character</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A single ascii character that will
be used to represent missing data.
If any normal allele contains this character, an error is raised.
Default: ‘-‘.</p></li>
<li><p><strong>impute_missing_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – <em>Deprecated in 0.3.0. Use ``isolated_as_missing``, but inverting value.
Will be removed in a future version</em></p></li>
</ul>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.9)">collections.abc.Iterable</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p>TypeError if the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code> or any of the alleles
at a site or the are not a single ascii character.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p>ValueError
if the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code> exists in one of the alleles</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.individual">
<code class="sig-name descname">individual</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.individual"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.individual" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual</span></a>
in this tree sequence with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Individual" title="tskit.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.individuals">
<code class="sig-name descname">individuals</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.individuals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.individuals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterable sequence of all individuals.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Individual" title="tskit.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.kc_distance">
<code class="sig-name descname">kc_distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">lambda_</span><span class="o">=</span><span class="default_value">0.0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.kc_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.kc_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average <a class="reference internal" href="#tskit.Tree.kc_distance" title="tskit.Tree.kc_distance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.kc_distance()</span></code></a> between pairs of trees along
the sequence whose intervals overlap. The average is weighted by the
fraction of the sequence on which each pair of trees overlap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>) – The other tree sequence to compare to.</p></li>
<li><p><strong>lambda</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The KC metric lambda parameter determining the
relative weight of topology and branch length.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The computed KC distance between this tree sequence and other.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.keep_intervals">
<code class="sig-name descname">keep_intervals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">intervals</span></em>, <em class="sig-param"><span class="n">simplify</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.keep_intervals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.keep_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of this tree sequence which includes only information in
the specified list of genomic intervals. Edges are truncated to lie within
these intervals, and sites and mutations falling outside these intervals
are discarded.  Note that it is the information outside the intervals that
is deleted, not the intervals themselves, so in particular, all samples
will be isolated outside of the retained intervals.</p>
<p>Note that node IDs may change as a result of this operation,
as by default <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> is called on the returned tree sequence
to remove redundant nodes. If you wish to map node IDs onto the same
nodes before and after this method has been called, specify <code class="docutils literal notranslate"><span class="pre">simplify=False</span></code>.</p>
<p>See also <a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">keep_intervals()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ltrim()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rtrim()</span></code></a>, and
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>) – A list (start, end) pairs describing the
genomic intervals to keep. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, return a simplified tree sequence where nodes
no longer used are discarded. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, add details of this operation to the
provenance information of the returned tree sequence.
(Default: True).</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.last">
<code class="sig-name descname">last</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.last"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.last" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the last tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. To iterate over all
trees in the sequence, use the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>**kwargs</strong> – Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.first(sample_lists=True)</span></code> will
result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The last tree in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.ltrim">
<code class="sig-name descname">ltrim</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.ltrim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.ltrim" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of this tree sequence with a potentially changed coordinate
system, such that empty regions (i.e. those not covered by any edge) at the start
of the tree sequence are trimmed away, and the leftmost edge starts at position
0. This affects the reported position of sites and edges. Additionally, sites and
their associated mutations to the left of the new zero point are thrown away.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.max_root_time">
<em class="property">property </em><code class="sig-name descname">max_root_time</code><a class="headerlink" href="#tskit.TreeSequence.max_root_time" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns time of the oldest root in any of the trees in this tree sequence.
This is usually equal to <code class="docutils literal notranslate"><span class="pre">np.max(ts.tables.nodes.time)</span></code> but may not be
since there can be nodes that are not present in any tree. Consistent
with the definition of tree roots, if there are no edges in the tree
sequence we return the time of the oldest sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The maximum time of a root in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.mean_descendants">
<code class="sig-name descname">mean_descendants</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.mean_descendants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.mean_descendants" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes for every node the mean number of samples in each of the
<cite>sample_sets</cite> that descend from that node, averaged over the
portions of the genome for which the node is ancestral to <em>any</em> sample.
The output is an array, <cite>C[node, j]</cite>, which reports the total span of
all genomes in <cite>sample_sets[j]</cite> that inherit from <cite>node</cite>, divided by
the total span of the genome on which <cite>node</cite> is an ancestor to any
sample in the tree sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future. The long-term stable
API for this method will be consistent with other <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a>.
In particular, the normalization by proportion of the genome that <cite>node</cite>
is an ancestor to anyone may not be the default behaviour in the future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of node IDs.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An array with dimensions (number of nodes in the tree sequence,
number of reference sets)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.metadata">
<em class="property">property </em><code class="sig-name descname">metadata</code><a class="headerlink" href="#tskit.TreeSequence.metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>The decoded metadata for this TreeSequence.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.TreeSequence.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this TreeSequence.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.migration">
<code class="sig-name descname">migration</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.migration"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.migration" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration</span></a> in this tree
sequence with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Migration" title="tskit.Migration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.migrations">
<code class="sig-name descname">migrations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.migrations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.migrations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a> in this tree sequence.</p>
<p>Migrations are returned in nondecreasing order of the <code class="docutils literal notranslate"><span class="pre">time</span></code> value.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterable sequence of all migrations.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Migration" title="tskit.Migration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.mutation">
<code class="sig-name descname">mutation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.mutation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.mutation" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation</span></a> in this tree sequence
with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.mutations">
<code class="sig-name descname">mutations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over all the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree sequence.
Mutations are returned in order of nondecreasing site ID.
See the <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> class for details on the available fields for
each mutation.</p>
<p>The returned iterator is equivalent to iterating over all sites
and all mutations in each site, i.e.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">tree_sequence</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="k">yield</span> <span class="n">mutation</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterator over all mutations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>iter(<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.TreeSequence.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this tree sequence. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.node">
<code class="sig-name descname">node</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.node"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.node" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node</span></a> in this tree sequence
with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.nodes">
<code class="sig-name descname">nodes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.nodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterable sequence of all nodes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_edges">
<em class="property">property </em><code class="sig-name descname">num_edges</code><a class="headerlink" href="#tskit.TreeSequence.num_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a> in this
tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of edges in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_individuals">
<em class="property">property </em><code class="sig-name descname">num_individuals</code><a class="headerlink" href="#tskit.TreeSequence.num_individuals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in
this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of individuals in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_migrations">
<em class="property">property </em><code class="sig-name descname">num_migrations</code><a class="headerlink" href="#tskit.TreeSequence.num_migrations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of migrations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_mutations">
<em class="property">property </em><code class="sig-name descname">num_mutations</code><a class="headerlink" href="#tskit.TreeSequence.num_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of mutations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_nodes">
<em class="property">property </em><code class="sig-name descname">num_nodes</code><a class="headerlink" href="#tskit.TreeSequence.num_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a> in
this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of nodes in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_populations">
<em class="property">property </em><code class="sig-name descname">num_populations</code><a class="headerlink" href="#tskit.TreeSequence.num_populations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of populations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_provenances">
<em class="property">property </em><code class="sig-name descname">num_provenances</code><a class="headerlink" href="#tskit.TreeSequence.num_provenances" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of provenances in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_samples">
<em class="property">property </em><code class="sig-name descname">num_samples</code><a class="headerlink" href="#tskit.TreeSequence.num_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of samples in this tree sequence. This is the number
of sample nodes in each tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of sample nodes in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_sites">
<em class="property">property </em><code class="sig-name descname">num_sites</code><a class="headerlink" href="#tskit.TreeSequence.num_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in
this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of sites in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.num_trees">
<em class="property">property </em><code class="sig-name descname">num_trees</code><a class="headerlink" href="#tskit.TreeSequence.num_trees" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of distinct trees in this tree sequence. This
is equal to the number of trees returned by the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a>
method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of trees in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.pairwise_diversity">
<code class="sig-name descname">pairwise_diversity</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.pairwise_diversity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.pairwise_diversity" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the pairwise nucleotide site diversity, the average number of sites
that differ between a randomly chosen pair of samples.  If <cite>samples</cite> is
specified, calculate the diversity within this set.</p>
<blockquote>
<div><div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.2.0: </span>please use <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity()</span></code></a> instead. Since version 0.2.0 the error
semantics have also changed slightly. It is no longer an error
when there is one sample and a tskit.LibraryError is raised
when non-sample IDs are provided rather than a ValueError. It is
also no longer an error to compute pairwise diversity at sites
with multiple mutations.</p>
</div>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The set of samples within which we calculate
the diversity. If None, calculate diversity within the entire sample.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The pairwise nucleotide site diversity.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.population">
<code class="sig-name descname">population</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.population"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.population" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population</span></a>
in this tree sequence with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Population" title="tskit.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.populations">
<code class="sig-name descname">populations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.populations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.populations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a> in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterable sequence of all populations.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Population" title="tskit.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.provenances">
<code class="sig-name descname">provenances</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.provenances"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.provenances" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a> in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterable sequence of all provenances.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Provenance" title="tskit.Provenance"><code class="xref py py-class docutils literal notranslate"><span class="pre">Provenance</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.rtrim">
<code class="sig-name descname">rtrim</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.rtrim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.rtrim" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of this tree sequence with the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property reset
so that the sequence ends at the end of the rightmost edge. Additionally, sites
and their associated mutations at positions greater than the new
<code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> are thrown away.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.sample_count_stat">
<code class="sig-name descname">sample_count_stat</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span></em>, <em class="sig-param"><span class="n">f</span></em>, <em class="sig-param"><span class="n">output_dim</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">polarised</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.sample_count_stat"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.sample_count_stat" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a windowed statistic from sample counts and a summary function.
This is a wrapper around <a class="reference internal" href="#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">general_stat()</span></code></a> for the common case in
which the weights are all either 1 or 0, i.e., functions of the joint
allele frequency spectrum.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-sample-sets"><span class="std std-ref">sample sets</span></a>,
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
If <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> is a list of <code class="docutils literal notranslate"><span class="pre">k</span></code> sets of samples, then
<code class="docutils literal notranslate"><span class="pre">f</span></code> should be a function that takes an argument of length <code class="docutils literal notranslate"><span class="pre">k</span></code> and
returns a one-dimensional array. The <code class="docutils literal notranslate"><span class="pre">j</span></code>-th element of the argument
to <code class="docutils literal notranslate"><span class="pre">f</span></code> will be the number of samples in <code class="docutils literal notranslate"><span class="pre">sample_sets[j]</span></code> that lie
below the node that <code class="docutils literal notranslate"><span class="pre">f</span></code> is being evaluated for. See
<a class="reference internal" href="#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">general_stat()</span></code></a>  for more details.</p>
<p>Here is a contrived example: suppose that <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> are two sets
of samples with <code class="docutils literal notranslate"><span class="pre">nA</span></code> and <code class="docutils literal notranslate"><span class="pre">nB</span></code> elements, respectively. Passing these
as sample sets will give <code class="docutils literal notranslate"><span class="pre">f</span></code> an argument of length two, giving the number
of samples in <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> below the node in question. So, if we define</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">f</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="n">pA</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">nA</span>
    <span class="n">pB</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">nB</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">pA</span> <span class="o">*</span> <span class="n">pB</span><span class="p">])</span>
</pre></div>
</div>
<p>then if all sites are biallelic,</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">sample_count_stat</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">],</span> <span class="n">f</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="s2">&quot;site&quot;</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;site&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>would compute, for each site, the product of the derived allele
frequencies in the two sample sets, in a (num sites, 1) array.  If
instead <code class="docutils literal notranslate"><span class="pre">f</span></code> returns <code class="docutils literal notranslate"><span class="pre">np.array([pA,</span> <span class="pre">pB,</span> <span class="pre">pA</span> <span class="pre">*</span> <span class="pre">pB])</span></code>, then the
output would be a (num sites, 3) array, with the first two columns
giving the allele frequencies in <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code>, respectively.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The summary function <code class="docutils literal notranslate"><span class="pre">f</span></code> should return zero when given both 0 and
the sample size (i.e., <code class="docutils literal notranslate"><span class="pre">f(0)</span> <span class="pre">=</span> <span class="pre">0</span></code> and
<code class="docutils literal notranslate"><span class="pre">f(np.array([len(x)</span> <span class="pre">for</span> <span class="pre">x</span> <span class="pre">in</span> <span class="pre">sample_sets])</span> <span class="pre">=</span> <span class="pre">0</span></code>).  This is
necessary for the statistic to be unaffected by parts of the tree
sequence ancestral to none or all of the samples, respectively.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>f</strong> – A function that takes a one-dimensional array of length
equal to the number of sample sets and returns a one-dimensional array.</p></li>
<li><p><strong>output_dim</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The length of <code class="docutils literal notranslate"><span class="pre">f</span></code>’s return value.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>polarised</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to leave the ancestral state out of computations:
see <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a> for more details.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to check that f(0) and f(total weight) are zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.samples">
<code class="sig-name descname">samples</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">population</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population_id</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array of the sample node IDs in this tree sequence. If the
<code class="docutils literal notranslate"><span class="pre">population</span></code> parameter is specified, only return sample IDs from this
population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The population of interest. If None,
return all samples.</p></li>
<li><p><strong>population_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – Deprecated alias for <code class="docutils literal notranslate"><span class="pre">population</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A numpy array of the node IDs for the samples of interest.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.segregating_sites">
<code class="sig-name descname">segregating_sites</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.segregating_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.segregating_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the density of segregating sites for each of the sets of nodes
from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>, and related quantities.
Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>, <a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. For a sample set <code class="docutils literal notranslate"><span class="pre">A</span></code>, computes:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>The sum over sites of the number of alleles found in <code class="docutils literal notranslate"><span class="pre">A</span></code> at each site
minus one, per unit of chromosome length.
If all sites have at most two alleles, this is the density of sites
that are polymorphic in <code class="docutils literal notranslate"><span class="pre">A</span></code>. To get the <strong>number</strong> of segregating minor
alleles per window, pass <code class="docutils literal notranslate"><span class="pre">span_normalise=False</span></code>.</p>
</dd>
<dt>“branch”</dt><dd><p>The total length of all branches in the tree subtended by the samples in
<code class="docutils literal notranslate"><span class="pre">A</span></code>, averaged across the window.</p>
</dd>
<dt>“node”</dt><dd><p>The proportion of the window on which the node is ancestral to some,
but not all, of the samples in <code class="docutils literal notranslate"><span class="pre">A</span></code>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.sequence_length">
<em class="property">property </em><code class="sig-name descname">sequence_length</code><a class="headerlink" href="#tskit.TreeSequence.sequence_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sequence length in this tree sequence. This defines the
genomic scale over which tree coordinates are defined. Given a
tree sequence with a sequence length <span class="math notranslate nohighlight">\(L\)</span>, the constituent
trees will be defined over the half-closed interval
<span class="math notranslate nohighlight">\([0, L)\)</span>. Each tree then covers some subset of this
interval — see <a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tskit.Tree.interval</span></code></a> for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The length of the sequence in this tree sequence in bases.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.simplify">
<code class="sig-name descname">simplify</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">map_nodes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">reduce_to_site_topology</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">filter_populations</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_individuals</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_sites</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">keep_unary</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">keep_input_roots</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_zero_mutation_sites</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.simplify"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.simplify" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a simplified tree sequence that retains only the history of
the nodes given in the list <code class="docutils literal notranslate"><span class="pre">samples</span></code>. If <code class="docutils literal notranslate"><span class="pre">map_nodes</span></code> is true,
also return a numpy array whose <code class="docutils literal notranslate"><span class="pre">u``th</span> <span class="pre">element</span> <span class="pre">is</span> <span class="pre">the</span> <span class="pre">ID</span> <span class="pre">of</span> <span class="pre">the</span> <span class="pre">node</span>
<span class="pre">in</span> <span class="pre">the</span> <span class="pre">simplified</span> <span class="pre">tree</span> <span class="pre">sequence</span> <span class="pre">that</span> <span class="pre">corresponds</span> <span class="pre">to</span> <span class="pre">node</span> <span class="pre">``u</span></code> in the
original tree sequence, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> (-1) if <code class="docutils literal notranslate"><span class="pre">u</span></code> is no longer
present in the simplified tree sequence.</p>
<p>In the returned tree sequence, the node with ID <code class="docutils literal notranslate"><span class="pre">0</span></code> corresponds to
<code class="docutils literal notranslate"><span class="pre">samples[0]</span></code>, node <code class="docutils literal notranslate"><span class="pre">1</span></code> corresponds to <code class="docutils literal notranslate"><span class="pre">samples[1]</span></code>, and so on.
Besides the samples, node IDs in the returned tree sequence are then
allocated sequentially in time order.</p>
<p>If you wish to simplify a set of tables that do not satisfy all
requirements for building a TreeSequence, then use
<a class="reference internal" href="#tskit.TableCollection.simplify" title="tskit.TableCollection.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.simplify()</span></code></a>.</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">reduce_to_site_topology</span></code> parameter is True, the returned tree
sequence will contain only topological information that is necessary to
represent the trees that contain sites. If there are zero sites in this
tree sequence, this will result in an output tree sequence with zero edges.
When the number of sites is greater than zero, every tree in the output
tree sequence will contain at least one site. For a given site, the
topology of the tree containing that site will be identical
(up to node ID remapping) to the topology of the corresponding tree
in the input tree sequence.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">filter_populations</span></code>, <code class="docutils literal notranslate"><span class="pre">filter_individuals</span></code> or <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code> is
True, any of the corresponding objects that are not referenced elsewhere
are filtered out. As this is the default behaviour, it is important to
realise IDs for these objects may change through simplification. By setting
these parameters to False, however, the corresponding tables can be preserved
without changes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The list of nodes for which to retain information. This
may be a numpy array (or array-like) object (dtype=np.int32).</p></li>
<li><p><strong>map_nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, return a tuple containing the resulting
tree sequence and a numpy array mapping node IDs in the current tree
sequence to their corresponding node IDs in the returned tree sequence.
If False (the default), return only the tree sequence object itself.</p></li>
<li><p><strong>reduce_to_site_topology</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to reduce the topology down
to the trees that are present at sites. (Default: False)</p></li>
<li><p><strong>filter_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, remove any populations that are
not referenced by nodes after simplification; new population IDs are
allocated sequentially from zero. If False, the population table will
not be altered in any way. (Default: True)</p></li>
<li><p><strong>filter_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, remove any individuals that are
not referenced by nodes after simplification; new individual IDs are
allocated sequentially from zero. If False, the individual table will
not be altered in any way. (Default: True)</p></li>
<li><p><strong>filter_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, remove any sites that are
not referenced by mutations after simplification; new site IDs are
allocated sequentially from zero. If False, the site table will not
be altered in any way. (Default: True)</p></li>
<li><p><strong>keep_unary</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, any unary nodes (i.e. nodes with exactly
one child) that exist on the path from samples to root will be preserved
in the output. (Default: False)</p></li>
<li><p><strong>keep_input_roots</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, insert edges from the MRCAs of the
samples to the roots in the input trees. If False, no topology older
than the MRCAs of the samples will be included. (Default: False)</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, record details of this call to
simplify in the returned tree sequence’s provenance information
(Default: True).</p></li>
<li><p><strong>filter_zero_mutation_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Deprecated alias for <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The simplified tree sequence, or (if <code class="docutils literal notranslate"><span class="pre">map_nodes</span></code> is True)
a tuple consisting of the simplified tree sequence and a numpy array
mapping source node IDs to their corresponding IDs in the new tree
sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">TreeSequence</a> or (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">TreeSequence</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.site">
<code class="sig-name descname">site</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">id_</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.site"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.site" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> in this tree sequence
with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.sites">
<code class="sig-name descname">sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a>
in this tree sequence. Sites are returned in order of increasing ID
(and also position). See the <a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> class for details on
the available fields for each site.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterable sequence of all sites.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.subset">
<code class="sig-name descname">subset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nodes</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.subset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tree sequence modified to contain only the entries referring to
the provided list of nodes, with nodes reordered according to the order
they appear in the <code class="docutils literal notranslate"><span class="pre">nodes</span></code> argument. Specifically, this subsets and reorders
each of the tables as follows:</p>
<ol class="arabic simple">
<li><p>Nodes: if in the list of nodes, and in the order provided.</p></li>
<li><p>Individuals and Populations: if referred to by a retained node,
and in the order first seen when traversing the list of retained nodes.</p></li>
<li><p>Edges: if both parent and child are retained nodes.</p></li>
<li><p>Mutations: if the mutation’s node is a retained node.</p></li>
<li><p>Sites: if any mutations remain at the site after removing mutations.</p></li>
</ol>
<p>Retained edges, mutations, and sites appear in the same
order as in the original tables.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">nodes</span></code> is the entire list of nodes in the tables, then the
resulting tables will be identical to the original tables, but with
nodes (and individuals and populations) reordered.</p>
<p>To instead subset the tables to a given portion of the <em>genome</em>, see
<a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">keep_intervals()</span></code></a>.</p>
<p><strong>Note:</strong> This is quite different from <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a>: the resulting
tables contain only the nodes given, not ancestral ones as well, and
does not simplify the relationships in any way.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The list of nodes for which to retain information. This
may be a numpy array (or array-like) object (dtype=np.int32).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.table_metadata_schemas">
<em class="property">property </em><code class="sig-name descname">table_metadata_schemas</code><a class="headerlink" href="#tskit.TreeSequence.table_metadata_schemas" title="Permalink to this definition">¶</a></dt>
<dd><p>The set of metadata schemas for the tables in this tree sequence.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.tables">
<em class="property">property </em><code class="sig-name descname">tables</code><a class="headerlink" href="#tskit.TreeSequence.tables" title="Permalink to this definition">¶</a></dt>
<dd><p>A copy of the tables underlying this tree sequence. See also
<a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dump_tables()</span></code></a>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This propery currently returns a copy of the tables
underlying a tree sequence but it may return a read-only
<strong>view</strong> in the future. Thus, if the tables will subsequently be
updated, please use the <a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dump_tables()</span></code></a> method instead as
this will always return a new copy of the TableCollection.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> containing all a copy of the
tables underlying this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection">TableCollection</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.tables_dict">
<em class="property">property </em><code class="sig-name descname">tables_dict</code><a class="headerlink" href="#tskit.TreeSequence.tables_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping names to tables in the
underlying <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>. Equivalent to calling
<code class="docutils literal notranslate"><span class="pre">ts.tables.name_map</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.to_macs">
<code class="sig-name descname">to_macs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.to_macs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.to_macs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a <a class="reference external" href="https://github.com/gchen98/macs">macs encoding</a>
of this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The macs representation of this TreeSequence as a string.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.to_nexus">
<code class="sig-name descname">to_nexus</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">precision</span><span class="o">=</span><span class="default_value">14</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.to_nexus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.to_nexus" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Nexus_file">nexus encoding</a>
of this tree sequence. Trees along the sequence are listed sequentially in
the TREES block. The tree spanning the interval <span class="math notranslate nohighlight">\([x, y)\)</span> is
given the name “tree_x_y”. Spatial positions are written at the
specified precision.</p>
<p>Nodes in the tree sequence are identified by the taxon labels of the
form <code class="docutils literal notranslate"><span class="pre">f&quot;tsk_{node.id}_{node.flags}&quot;</span></code>, such that a node with <code class="docutils literal notranslate"><span class="pre">id=5</span></code>
and <code class="docutils literal notranslate"><span class="pre">flags=1</span></code> will have the label <code class="docutils literal notranslate"><span class="pre">&quot;tsk_5_1&quot;</span></code> (please see the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">data model</span></a> section for details
on the interpretation of node ID and flags values). These labels are
listed for all nodes in the tree sequence in the <code class="docutils literal notranslate"><span class="pre">TAXLABELS</span></code> block.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>precision</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The numerical precision with which branch lengths
and tree positions are printed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A nexus representation of this TreeSequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.trait_correlation">
<code class="sig-name descname">trait_correlation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">W</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.trait_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the mean squared correlations between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code>
(the “phenotypes”) and inheritance along the tree sequence.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on all samples in the tree sequence.</p>
<p>This is computed as squared covariance in
<a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_covariance</span></code></a>,
but divided by <span class="math notranslate nohighlight">\(p (1-p)\)</span>, where <cite>p</cite> is the proportion of samples
inheriting from the allele, branch, or node in question.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>The sum of squared correlations between presence/absence of each allele and
phenotypes, divided by length of the window (if <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>).
This is computed as the
<a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_covariance</span></code></a>
divided by the variance of the relevant column of W
and by ;math:<cite>p * (1 - p)</cite>, where <span class="math notranslate nohighlight">\(p\)</span> is the allele frequency.</p>
</dd>
<dt>“branch”</dt><dd><p>The sum of squared correlations between the split induced by each branch and
phenotypes, multiplied by branch length, averaged across trees in
the window. This is computed as the
<a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_covariance</span></code></a>,
divided by the variance of the column of w
and by <span class="math notranslate nohighlight">\(p * (1 - p)\)</span>, where <span class="math notranslate nohighlight">\(p\)</span> is the proportion of
the samples lying below the branch.</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the squared correlation between the property of
inheriting from this node and phenotypes, computed as in “branch”.</p>
</dd>
</dl>
<p>Note that above we divide by the <strong>sample</strong> variance, which for a
vector x of length n is <code class="docutils literal notranslate"><span class="pre">np.var(x)</span> <span class="pre">*</span> <span class="pre">n</span> <span class="pre">/</span> <span class="pre">(n-1)</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – An array of values with one row for each sample and one
column for each “phenotype”. Each column must have positive standard
deviation.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.trait_covariance">
<code class="sig-name descname">trait_covariance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">W</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_covariance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.trait_covariance" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the mean squared covariances between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code>
(the “phenotypes”) and inheritance along the tree sequence.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on all samples in the tree sequence.</p>
<p>Concretely, if <cite>g</cite> is a binary vector that indicates inheritance from an allele,
branch, or node and <cite>w</cite> is a column of W, normalised to have mean zero,
then the covariance of <cite>g</cite> and <cite>w</cite> is <span class="math notranslate nohighlight">\(\sum_i g_i w_i\)</span>, the sum of the
weights corresponding to entries of <cite>g</cite> that are <cite>1</cite>. Since weights sum to
zero, this is also equal to the sum of weights whose entries of <cite>g</cite> are 0.
So, <span class="math notranslate nohighlight">\(cov(g,w)^2 = ((\sum_i g_i w_i)^2 + (\sum_i (1-g_i) w_i)^2)/2\)</span>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>The sum of squared covariances between presence/absence of each allele and
phenotypes, divided by length of the window (if <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>).
This is computed as sum_a (sum(w[a])^2 / 2), where
w is a column of W with the average subtracted off,
and w[a] is the sum of all entries of w corresponding to samples
carrying allele “a”, and the first sum is over all alleles.</p>
</dd>
<dt>“branch”</dt><dd><p>The sum of squared covariances between the split induced by each branch and
phenotypes, multiplied by branch length, averaged across trees in
the window. This is computed as above: a branch with total weight
w[b] below b contributes (branch length) * w[b]^2 to the total
value for a tree. (Since the sum of w is zero, the total weight
below b and not below b are equal, canceling the factor of 2
above.)</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the squared covariance between the property of
inheriting from this node and phenotypes, computed as in “branch”.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – An array of values with one row for each sample and one
column for each “phenotype”.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.trait_linear_model">
<code class="sig-name descname">trait_linear_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">W</span></em>, <em class="sig-param"><span class="n">Z</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">windows</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'site'</span></em>, <em class="sig-param"><span class="n">span_normalise</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_linear_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.trait_linear_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the relationship between trait and genotype after accounting for
covariates.  Concretely, for each trait w (i.e., each column of W),
this does a least-squares fit of the linear model <span class="math notranslate nohighlight">\(w \sim g + Z\)</span>,
where <span class="math notranslate nohighlight">\(g\)</span> is inheritance in the tree sequence (e.g., genotype)
and the columns of <span class="math notranslate nohighlight">\(Z\)</span> are covariates, and returns the squared
coefficient of <span class="math notranslate nohighlight">\(g\)</span> in this linear model.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on all samples in the tree sequence.</p>
<p>To do this, if <cite>g</cite> is a binary vector that indicates inheritance from an allele,
branch, or node and <cite>w</cite> is a column of W, there are <span class="math notranslate nohighlight">\(k\)</span> columns of
<span class="math notranslate nohighlight">\(Z\)</span>, and the <span class="math notranslate nohighlight">\(k+2\)</span>-vector <span class="math notranslate nohighlight">\(b\)</span> minimises
<span class="math notranslate nohighlight">\(\sum_i (w_i - b_0 - b_1 g_i - b_2 z_{2,i} - ... b_{k+2} z_{k+2,i})^2\)</span>
then this returns the number <span class="math notranslate nohighlight">\(b_1^2\)</span>. If <span class="math notranslate nohighlight">\(g\)</span> lies in the linear span
of the columns of <span class="math notranslate nohighlight">\(Z\)</span>, then <span class="math notranslate nohighlight">\(b_1\)</span> is set to 0. To fit the
linear model without covariates (only the intercept), set <cite>Z = None</cite>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>“site”</dt><dd><p>Computes the sum of <span class="math notranslate nohighlight">\(b_1^2/2\)</span> for each allele in the window,
as above with <span class="math notranslate nohighlight">\(g\)</span> indicating presence/absence of the allele,
then divided by the length of the window if <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>.
(For biallelic loci, this number is the same for both alleles, and so summing
over each cancels the factor of two.)</p>
</dd>
<dt>“branch”</dt><dd><p>The squared coefficient <cite>b_1^2</cite>, computed for the split induced by each
branch (i.e., with <span class="math notranslate nohighlight">\(g\)</span> indicating inheritance from that branch),
multiplied by branch length and tree span, summed over all trees
in the window, and divided by the length of the window if
<code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>.</p>
</dd>
<dt>“node”</dt><dd><p>For each node, the squared coefficient <cite>b_1^2</cite>, computed for the property of
inheriting from this node, as in “branch”.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – An array of values with one row for each sample and one
column for each “phenotype”.</p></li>
<li><p><strong>Z</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – An array of values with one row for each sample and one
column for each “covariate”, or <cite>None</cite>. Columns of <cite>Z</cite> must be linearly
independent.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string giving the “type” of the statistic to be computed
(defaults to “site”).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.trait_regression">
<code class="sig-name descname">trait_regression</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_regression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.trait_regression" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated synonym for
<a class="reference internal" href="#tskit.TreeSequence.trait_linear_model" title="tskit.TreeSequence.trait_linear_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_linear_model</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.trees">
<code class="sig-name descname">trees</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tracked_samples</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">sample_lists</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">root_threshold</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">sample_counts</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">tracked_leaves</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">leaf_counts</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">leaf_lists</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trees"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.trees" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over the trees in this tree sequence. Each value
returned in this iterator is an instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. Upon
successful termination of the iterator, the tree will be in the
“cleared” null state.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sample_lists</span></code> and <code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameters are passed
to the <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> constructor, and control
the options that are set in the returned tree instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>Do not store the results of this iterator in a list!
For performance reasons, the same underlying object is used
for every tree returned which will most likely lead to unexpected
behaviour. If you wish to obtain a list of trees in a tree sequence
please use <code class="docutils literal notranslate"><span class="pre">ts.aslist()</span></code> instead.</p>
</dd>
<dt class="field-even">Parameters</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>tracked_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The list of samples to be tracked and
counted using the <a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a> method.</p></li>
<li><p><strong>sample_lists</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, provide more efficient access
to the samples beneath a give node using the
<a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a> method.</p></li>
<li><p><strong>root_threshold</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The minimum number of samples that a node
must be ancestral to for it to be in the list of roots. By default
this is 1, so that isolated samples (representing missing data)
are roots. To efficiently restrict the roots of the tree to
those subtending meaningful topology, set this to 2. This value
is only relevant when trees have multiple roots.</p></li>
<li><p><strong>sample_counts</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Deprecated since 0.2.4.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterator over the Trees in this tree sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>collections.abc.Iterable, <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.trim">
<code class="sig-name descname">trim</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.trim" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of this tree sequence with any empty regions (i.e. those not
covered by any edge) on the right and left trimmed away. This may reset both the
coordinate system and the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property. It is functionally
equivalent to <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rtrim()</span></code></a> followed by <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ltrim()</span></code></a>. Sites and their
associated mutations in the empty regions are thrown away.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.union">
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">node_mapping</span></em>, <em class="sig-param"><span class="n">check_shared_equality</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">add_populations</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.union"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.union" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an expanded tree sequence which contains the node-wise union of
<code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>, obtained by adding the non-shared portions of
<code class="docutils literal notranslate"><span class="pre">other</span></code> onto <code class="docutils literal notranslate"><span class="pre">self</span></code>. The “shared” portions are specified using a
map that specifies which nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code> are equivalent to those in
<code class="docutils literal notranslate"><span class="pre">self</span></code>: the <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code> argument should be an array of length
equal to the number of nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code> and whose entries are the ID
of the matching node in <code class="docutils literal notranslate"><span class="pre">self</span></code>, or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if there is no
matching node. Those nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code> that map to <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> will
be added to <code class="docutils literal notranslate"><span class="pre">self</span></code>, along with:</p>
<ol class="arabic simple">
<li><p>Individuals whose nodes are new to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p>Edges whose parent or child are new to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p>Mutations whose nodes are new to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p>Sites which were not present in <code class="docutils literal notranslate"><span class="pre">self</span></code>, if the site contains a newly
added mutation.</p></li>
</ol>
<p>By default, populations of newly added nodes are assumed to be new
populations, and added to the population table as well.</p>
<p>Note that this operation also sorts the resulting tables, so the
resulting tree sequence may not be equal to <code class="docutils literal notranslate"><span class="pre">self</span></code> even if nothing
new was added (although it would differ only in ordering of the tables).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><em>TableCollection</em></a>) – Another table collection.</p></li>
<li><p><strong>node_mapping</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An array of node IDs that relate nodes in
<code class="docutils literal notranslate"><span class="pre">other</span></code> to nodes in <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p><strong>check_shared_equality</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, the shared portions of the
tree sequences will be checked for equality. It does so by
subsetting both <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code> on the equivalent nodes
specified in <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code>, and then checking for equality of
the subsets.</p></li>
<li><p><strong>add_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, nodes new to <code class="docutils literal notranslate"><span class="pre">self</span></code> will be
assigned new population IDs.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to record a provenance entry
in the provenance table for this operation.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.variants">
<code class="sig-name descname">variants</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">as_bytes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">isolated_as_missing</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alleles</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">impute_missing_data</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over the variants in this tree sequence. See the
<a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class for details on the fields of each returned
object. The <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> for the variants are numpy arrays,
corresponding to indexes into the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> attribute in the
<a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> object. By default, the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> for each
site are generated automatically, such that the ancestral state
is at the zeroth index and subsequent alleles are listed in no
particular order. This means that the encoding of alleles in
terms of genotype values can vary from site-to-site, which is
sometimes inconvenient. It is possible to specify a fixed mapping
from allele strings to genotype values using the <code class="docutils literal notranslate"><span class="pre">alleles</span></code>
parameter. For example, if we set <code class="docutils literal notranslate"><span class="pre">alleles=(&quot;A&quot;,</span> <span class="pre">&quot;C&quot;,</span> <span class="pre">&quot;G&quot;,</span> <span class="pre">&quot;T&quot;)</span></code>,
this will map allele “A” to 0, “C” to 1 and so on (the
<a class="reference internal" href="#tskit.ALLELES_ACGT" title="tskit.ALLELES_ACGT"><code class="xref py py-data docutils literal notranslate"><span class="pre">ALLELES_ACGT</span></code></a> constant provides a shortcut for this
common mapping).</p>
<p>By default, genotypes are generated for all samples. The <code class="docutils literal notranslate"><span class="pre">samples</span></code>
parameter allows us to specify the nodes for which genotypes are
generated; output order of genotypes in the returned variants
corresponds to the order of the samples in this list. It is also
possible to provide <strong>non-sample</strong> nodes as an argument here, if you
wish to generate genotypes for (e.g.) internal nodes. However,
<code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> must be False in this case, as it is not
possible to detect missing data for non-sample nodes.</p>
<p>If isolated samples are present at a given site without mutations above them,
they will be interpreted as <a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>
the genotypes array will contain a special value <a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a>
(-1) to identify these missing samples, and the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> tuple will
end with the value <code class="docutils literal notranslate"><span class="pre">None</span></code> (note that this is true whether we specify
a fixed mapping using the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> parameter or not).
See the <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class for more details on how missing data is
reported.</p>
<p>Such samples are treated as missing data by default, but if
<code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is set to to False, they will not be treated as
missing, and so assigned the ancestral state.
This was the default behaviour in versions prior to 0.2.0. Prior to 0.3.0
the <cite>impute_missing_data</cite> argument controlled this behaviour.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">as_bytes</span></code> parameter is kept as a compatibility
option for older code. It is not the recommended way of
accessing variant data, and will be deprecated in a later
release.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>as_bytes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, the genotype values will be returned
as a Python bytes object. Legacy use only.</p></li>
<li><p><strong>samples</strong> (<em>array_like</em>) – An array of sample IDs for which to generate
genotypes, or None for all samples. Default: None.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, the allele assigned to
missing samples (i.e., isolated samples without mutations) is
the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code>. If False, missing samples will be
assigned the ancestral state.
Default: True.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a>) – A tuple of strings defining the encoding of
alleles as integer genotype values. At least one allele must be provided.
If duplicate alleles are provided, output genotypes will always be
encoded as the first occurance of the allele. If None (the default),
the alleles are encoded as they are encountered during genotype
generation.</p></li>
<li><p><strong>impute_missing_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – <em>Deprecated in 0.3.0. Use ``isolated_as_missing``, but inverting value.
Will be removed in a future version</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An iterator of all variants this tree sequence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>iter(<a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TreeSequence.write_vcf">
<code class="sig-name descname">write_vcf</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">output</span></em>, <em class="sig-param"><span class="n">ploidy</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">contig_id</span><span class="o">=</span><span class="default_value">'1'</span></em>, <em class="sig-param"><span class="n">individuals</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">individual_names</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">position_transform</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.write_vcf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TreeSequence.write_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a VCF formatted file to the specified file-like object.
If there is individual information present in the tree sequence
(see <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a>), the values for
sample nodes associated with these individuals are combined
into phased multiploid individuals and output.</p>
<p>If there is no individual data present in the tree sequence, synthetic
individuals are created by combining adjacent samples, and the number
of samples combined is equal to the specified ploidy value (1 by
default). For example, if we have a ploidy of 2 and a sample of size 6,
then we will have 3 diploid samples in the output, consisting of the
combined genotypes for samples [0, 1], [2, 3] and [4, 5]. If we had
genotypes 011110 at a particular variant, then we would output the
diploid genotypes 0|1, 1|1 and 1|0 in VCF.</p>
<p>Each individual in the output is identified by a string; these are the
VCF “sample” names. By default, these are of the form <code class="docutils literal notranslate"><span class="pre">tsk_0</span></code>,
<code class="docutils literal notranslate"><span class="pre">tsk_1</span></code> etc, up to the number of individuals, but can be manually
specified using the <code class="docutils literal notranslate"><span class="pre">individual_names</span></code> argument. We do not check
for duplicates in this array, or perform any checks to ensure that
the output VCF is well-formed.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Warning to <code class="docutils literal notranslate"><span class="pre">plink</span></code> users:</p>
<p>As the default first individual name is <code class="docutils literal notranslate"><span class="pre">tsk_0</span></code>, <code class="docutils literal notranslate"><span class="pre">plink</span></code> will
throw this error when loading the VCF:</p>
<p><code class="docutils literal notranslate"><span class="pre">Error:</span> <span class="pre">Sample</span> <span class="pre">ID</span> <span class="pre">ends</span> <span class="pre">with</span> <span class="pre">&quot;_0&quot;,</span> <span class="pre">which</span> <span class="pre">induces</span> <span class="pre">an</span> <span class="pre">invalid</span> <span class="pre">IID</span> <span class="pre">of</span> <span class="pre">'0'.</span></code></p>
<p>This can be fixed by using the <code class="docutils literal notranslate"><span class="pre">individual_names</span></code> argument
to set the names to anything where the first name doesn’t end with <code class="docutils literal notranslate"><span class="pre">_0</span></code>.
An example implementation for diploid individuals is:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">n_dip_indv</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_samples</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">indv_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;tsk_</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="si">}</span><span class="s2">indv&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_dip_indv</span><span class="p">)]</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;output.vcf&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">vcf_file</span><span class="p">:</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">individual_names</span><span class="o">=</span><span class="n">indv_names</span><span class="p">)</span>
</pre></div>
</div>
<p>Adding a second <code class="docutils literal notranslate"><span class="pre">_</span></code> (eg: <code class="docutils literal notranslate"><span class="pre">tsk_0_indv</span></code>) is not recommended as
<code class="docutils literal notranslate"><span class="pre">plink</span></code> uses <code class="docutils literal notranslate"><span class="pre">_</span></code> as the default separator for separating family
id and individual id, and two <code class="docutils literal notranslate"><span class="pre">_</span></code> will throw an error.</p>
</div>
<p>The REF value in the output VCF is the ancestral allele for a site
and ALT values are the remaining alleles. It is important to note,
therefore, that for real data this means that the REF value for a given
site <strong>may not</strong> be equal to the reference allele. We also do not
check that the alleles result in a valid VCF—for example, it is possible
to use the tab character as an allele, leading to a broken VCF.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">position_transform</span></code> argument provides a way to flexibly translate
the genomic location of sites in tskit to the appropriate value in VCF.
There are two fundamental differences in the way that tskit and VCF define
genomic coordinates. The first is that tskit uses floating point values
to encode positions, whereas VCF uses integers. Thus, if the tree sequence
contains positions at non-integral locations there is an information loss
incurred by translating to VCF. By default, we round the site positions
to the nearest integer, such that there may be several sites with the
same integer position in the output. The second difference between VCF
and tskit is that VCF is defined to be a 1-based coordinate system, whereas
tskit uses 0-based. However, how coordinates are transformed depends
on the VCF parser, and so we do <strong>not</strong> account for this change in
coordinate system by default.</p>
<p>Example usage:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;output.vcf&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">vcf_file</span><span class="p">:</span>
    <span class="n">tree_sequence</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>The VCF output can also be compressed using the <a class="reference external" href="https://docs.python.org/3/library/gzip.html#module-gzip" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">gzip</span></code></a> module, if you wish:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">gzip</span>

<span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;output.vcf.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;wt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</pre></div>
</div>
<p>However, this gzipped VCF may not be fully compatible with downstream tools
such as tabix, which may require the VCF use the specialised bgzip format.
A general way to convert VCF data to various formats is to pipe the text
produced by <code class="docutils literal notranslate"><span class="pre">tskit</span></code> into <code class="docutils literal notranslate"><span class="pre">bcftools</span></code>, as done here:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="n">read_fd</span><span class="p">,</span> <span class="n">write_fd</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">pipe</span><span class="p">()</span>
<span class="n">write_pipe</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">fdopen</span><span class="p">(</span><span class="n">write_fd</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;output.bcf&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">bcf_file</span><span class="p">:</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span>
        <span class="p">[</span><span class="s2">&quot;bcftools&quot;</span><span class="p">,</span> <span class="s2">&quot;view&quot;</span><span class="p">,</span> <span class="s2">&quot;-O&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">],</span> <span class="n">stdin</span><span class="o">=</span><span class="n">read_fd</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">bcf_file</span>
    <span class="p">)</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="n">write_pipe</span><span class="p">)</span>
    <span class="n">write_pipe</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">os</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">read_fd</span><span class="p">)</span>
    <span class="n">proc</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">proc</span><span class="o">.</span><span class="n">returncode</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;bcftools failed with status:&quot;</span><span class="p">,</span> <span class="n">proc</span><span class="o">.</span><span class="n">returncode</span><span class="p">)</span>
</pre></div>
</div>
<p>This can also be achieved on the command line use the <code class="docutils literal notranslate"><span class="pre">tskit</span> <span class="pre">vcf</span></code> command,
e.g.:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ tskit vcf example.trees <span class="p">|</span> bcftools view -O b &gt; example.bcf
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.IOBase" title="(in Python v3.9)"><em>io.IOBase</em></a>) – The file-like object to write the VCF output.</p></li>
<li><p><strong>ploidy</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ploidy of the individuals to be written to
VCF. This sample size must be evenly divisible by ploidy.</p></li>
<li><p><strong>contig_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The value of the CHROM column in the output VCF.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>)</em>) – A list containing the individual IDs to
write out to VCF. Defaults to all individuals in the tree sequence.</p></li>
<li><p><strong>individual_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – A list of string names to identify
individual columns in the VCF. In VCF nomenclature, these are the
sample IDs. If specified, this must be a list of strings of
length equal to the number of individuals to be output. Note that
we do not check the form of these strings in any way, so that is
is possible to output malformed VCF (for example, by embedding a
tab character within on of the names). The default is to output
<code class="docutils literal notranslate"><span class="pre">tsk_j</span></code> for the jth individual.</p></li>
<li><p><strong>position_transform</strong> – A callable that transforms the
site position values into integer valued coordinates suitable for
VCF. The function takes a single positional parameter x and must
return an integer numpy array the same dimension as x. By default,
this is set to <code class="docutils literal notranslate"><span class="pre">numpy.round()</span></code> which will round values to the
nearest integer. If the string “legacy” is provided here, the
pre 0.2.0 legacy behaviour of rounding values to the nearest integer
(starting from 1) and avoiding the output of identical positions
by incrementing is used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="the-tree-class">
<h3>The <code class="docutils literal notranslate"><span class="pre">Tree</span></code> class<a class="headerlink" href="#the-tree-class" title="Permalink to this headline">¶</a></h3>
<p>Trees in a tree sequence can be obtained by iterating over
<a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> and specific trees can be accessed using methods
such as <a class="reference internal" href="#tskit.TreeSequence.first" title="tskit.TreeSequence.first"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.first()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.at" title="tskit.TreeSequence.at"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.at()</span></code></a> and
<a class="reference internal" href="#tskit.TreeSequence.at_index" title="tskit.TreeSequence.at_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.at_index()</span></code></a>. Each tree is an instance of the following
class which provides methods, for example, to access information
about particular nodes in the tree.</p>
<dl class="py class">
<dt id="tskit.Tree">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Tree</code><a class="reference internal" href="_modules/tskit/trees.html#Tree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree" title="Permalink to this definition">¶</a></dt>
<dd><p>A single tree in a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. Please see the
<a class="reference internal" href="tutorial.html#sec-tutorial-moving-along-a-tree-sequence"><span class="std std-ref">Moving along a tree sequence</span></a> section for information
on how efficiently access trees sequentially or obtain a list
of individual trees in a tree sequence.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sample_lists</span></code> parameter controls the features that are enabled
for this tree. If <code class="docutils literal notranslate"><span class="pre">sample_lists</span></code> is True a more efficient algorithm is
used in the <a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a> method.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameter can be used to efficiently count the
number of samples in a given set that exist in a particular subtree
using the <a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a> method.</p>
<p>The <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> class is a state-machine which has a state
corresponding to each of the trees in the parent tree sequence. We
transition between these states by using the seek functions like
<a class="reference internal" href="#tskit.Tree.first" title="tskit.Tree.first"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.first()</span></code></a>, <a class="reference internal" href="#tskit.Tree.last" title="tskit.Tree.last"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.last()</span></code></a>, <a class="reference internal" href="#tskit.Tree.seek" title="tskit.Tree.seek"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek()</span></code></a> and
<a class="reference internal" href="#tskit.Tree.seek_index" title="tskit.Tree.seek_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek_index()</span></code></a>. There is one more state, the so-called “null”
or “cleared” state. This is the state that a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> is in
immediately after initialisation;  it has an index of -1, and no edges. We
can also enter the null state by calling <a class="reference internal" href="#tskit.Tree.next" title="tskit.Tree.next"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.next()</span></code></a> on the last
tree in a sequence, calling <a class="reference internal" href="#tskit.Tree.prev" title="tskit.Tree.prev"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.prev()</span></code></a> on the first tree in a
sequence or calling calling the <a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.clear()</span></code></a> method at any time.</p>
<p>The high-level TreeSequence seeking and iterations methods (e.g,
<a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a>) are built on these low-level state-machine
seek operations. We recommend these higher level operations for most
users.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>) – The parent tree sequence.</p></li>
<li><p><strong>tracked_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The list of samples to be tracked and
counted using the <a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a> method.</p></li>
<li><p><strong>sample_lists</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, provide more efficient access
to the samples beneath a give node using the
<a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a> method.</p></li>
<li><p><strong>root_threshold</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The minimum number of samples that a node
must be ancestral to for it to be in the list of roots. By default
this is 1, so that isolated samples (representing missing data)
are roots. To efficiently restrict the roots of the tree to
those subtending meaningful topology, set this to 2. This value
is only relevant when trees have multiple roots.</p></li>
<li><p><strong>sample_counts</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Deprecated since 0.2.4.</p></li>
</ul>
</dd>
</dl>
<p><strong>Methods</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.as_dict_of_dicts" title="tskit.Tree.as_dict_of_dicts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">as_dict_of_dicts</span></code></a>()</p></td>
<td><p><p>Convert tree to dict of dicts for conversion to a <a class="reference external" href="https://networkx.github.io/documentation/stable/reference/classes/digraph.html">networkx graph</a>.</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.branch_length" title="tskit.Tree.branch_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">branch_length</span></code></a>(u)</p></td>
<td><p>Returns the length of the branch (in generations) joining the specified node to its parent.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.children" title="tskit.Tree.children"><code class="xref py py-obj docutils literal notranslate"><span class="pre">children</span></code></a>(u)</p></td>
<td><p>Returns the children of the specified node <code class="docutils literal notranslate"><span class="pre">u</span></code> as a tuple of integer node IDs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code></a>()</p></td>
<td><p>Resets this tree back to the initial null state.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.copy" title="tskit.Tree.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Returns a deep copy of this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_topologies</span></code></a>([sample_sets])</p></td>
<td><p>Calculates the distribution of embedded topologies for every combination of the sample sets in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.depth" title="tskit.Tree.depth"><code class="xref py py-obj docutils literal notranslate"><span class="pre">depth</span></code></a>(u)</p></td>
<td><p>Returns the number of nodes on the path from <code class="docutils literal notranslate"><span class="pre">u</span></code> to a root, not including <code class="docutils literal notranslate"><span class="pre">u</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.draw" title="tskit.Tree.draw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">draw</span></code></a>([path, width, height, node_labels, …])</p></td>
<td><p>Returns a drawing of this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.draw_svg" title="tskit.Tree.draw_svg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">draw_svg</span></code></a>([path, size, tree_height_scale, …])</p></td>
<td><p>Return an SVG representation of a single tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.first" title="tskit.Tree.first"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first</span></code></a>()</p></td>
<td><p>Seeks to the first tree in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.generate_balanced" title="tskit.Tree.generate_balanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">generate_balanced</span></code></a>(num_leaves, *[, arity, …])</p></td>
<td><p>Generate a :class:&lt;Tree&gt; with the specified number of leaves that is maximally balanced.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.generate_comb" title="tskit.Tree.generate_comb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">generate_comb</span></code></a>(num_leaves, *[, span, …])</p></td>
<td><p>Generate a :class:&lt;Tree&gt; in which all internal nodes have two children and the left child is a leaf.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.generate_random" title="tskit.Tree.generate_random"><code class="xref py py-obj docutils literal notranslate"><span class="pre">generate_random</span></code></a>(num_leaves, *[, arity, …])</p></td>
<td><p>Generate a random :class:&lt;Tree&gt; with <span class="math notranslate nohighlight">\(n\)</span> = <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves with an equal probability of returning any valid topology.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.generate_star" title="tskit.Tree.generate_star"><code class="xref py py-obj docutils literal notranslate"><span class="pre">generate_star</span></code></a>(num_leaves, *[, span, …])</p></td>
<td><p>Generate a :class:&lt;Tree&gt; whose leaf nodes all have the same parent (i.e.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.is_descendant" title="tskit.Tree.is_descendant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_descendant</span></code></a>(u, v)</p></td>
<td><p>Returns True if the specified node u is a descendant of node v and False otherwise.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.is_internal" title="tskit.Tree.is_internal"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_internal</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is not a leaf.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.is_isolated" title="tskit.Tree.is_isolated"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_isolated</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is isolated in this tree: that is it has no parents and no children.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.is_leaf" title="tskit.Tree.is_leaf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_leaf</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is a leaf.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.is_sample" title="tskit.Tree.is_sample"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_sample</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is a sample.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.kc_distance" title="tskit.Tree.kc_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kc_distance</span></code></a>(other[, lambda_])</p></td>
<td><p>Returns the Kendall-Colijn distance between the specified pair of trees.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.last" title="tskit.Tree.last"><code class="xref py py-obj docutils literal notranslate"><span class="pre">last</span></code></a>()</p></td>
<td><p>Seeks to the last tree in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.leaves" title="tskit.Tree.leaves"><code class="xref py py-obj docutils literal notranslate"><span class="pre">leaves</span></code></a>([u])</p></td>
<td><p>Returns an iterator over all the leaves in this tree that are underneath the specified node.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.left_child" title="tskit.Tree.left_child"><code class="xref py py-obj docutils literal notranslate"><span class="pre">left_child</span></code></a>(u)</p></td>
<td><p>Returns the leftmost child of the specified node.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.left_sib" title="tskit.Tree.left_sib"><code class="xref py py-obj docutils literal notranslate"><span class="pre">left_sib</span></code></a>(u)</p></td>
<td><p>Returns the sibling node to the left of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u does not have a left sibling.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.map_mutations" title="tskit.Tree.map_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">map_mutations</span></code></a>(genotypes, alleles)</p></td>
<td><p>Given observations for the samples in this tree described by the specified set of genotypes and alleles, return a parsimonious set of state transitions explaining these observations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.mrca" title="tskit.Tree.mrca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mrca</span></code></a>(u, v)</p></td>
<td><p>Returns the most recent common ancestor of the specified nodes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.mutations" title="tskit.Tree.mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mutations</span></code></a>()</p></td>
<td><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.newick" title="tskit.Tree.newick"><code class="xref py py-obj docutils literal notranslate"><span class="pre">newick</span></code></a>([precision, root, node_labels, …])</p></td>
<td><p><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">newick encoding</a> of this tree.</p>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.next" title="tskit.Tree.next"><code class="xref py py-obj docutils literal notranslate"><span class="pre">next</span></code></a>()</p></td>
<td><p>Seeks to the next tree in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nodes</span></code></a>([root, order])</p></td>
<td><p>Returns an iterator over the node IDs in this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.num_children" title="tskit.Tree.num_children"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_children</span></code></a>(u)</p></td>
<td><p>Returns the number of children of the specified node (i.e.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_samples" title="tskit.Tree.num_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_samples</span></code></a>([u])</p></td>
<td><p>Returns the number of samples in this tree underneath the specified node (including the node itself).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_tracked_samples</span></code></a>([u])</p></td>
<td><p>Returns the number of samples in the set specified in the <code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameter of the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method underneath the specified node.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.parent" title="tskit.Tree.parent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parent</span></code></a>(u)</p></td>
<td><p>Returns the parent of the specified node.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.population" title="tskit.Tree.population"><code class="xref py py-obj docutils literal notranslate"><span class="pre">population</span></code></a>(u)</p></td>
<td><p>Returns the population associated with the specified node.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.prev" title="tskit.Tree.prev"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prev</span></code></a>()</p></td>
<td><p>Seeks to the previous tree in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rank</span></code></a>()</p></td>
<td><p>Produce the rank of this tree in the enumeration of all leaf-labelled trees of n leaves.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.right_child" title="tskit.Tree.right_child"><code class="xref py py-obj docutils literal notranslate"><span class="pre">right_child</span></code></a>(u)</p></td>
<td><p>Returns the rightmost child of the specified node.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.right_sib" title="tskit.Tree.right_sib"><code class="xref py py-obj docutils literal notranslate"><span class="pre">right_sib</span></code></a>(u)</p></td>
<td><p>Returns the sibling node to the right of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u does not have a right sibling.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">samples</span></code></a>([u])</p></td>
<td><p>Returns an iterator over all the samples in this tree that are underneath the specified node.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.seek" title="tskit.Tree.seek"><code class="xref py py-obj docutils literal notranslate"><span class="pre">seek</span></code></a>(position)</p></td>
<td><p>Sets the state to represent the tree that covers the specified position in the parent tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.seek_index" title="tskit.Tree.seek_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">seek_index</span></code></a>(index)</p></td>
<td><p>Sets the state to represent the tree at the specified index in the parent tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.sites" title="tskit.Tree.sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sites</span></code></a>()</p></td>
<td><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.split_polytomies" title="tskit.Tree.split_polytomies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">split_polytomies</span></code></a>(*[, epsilon, method, …])</p></td>
<td><p>Return a new <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> where extra nodes and edges have been inserted so that any any node <code class="docutils literal notranslate"><span class="pre">u</span></code> with greater than 2 children — a multifurcation or “polytomy” — is resolved into successive bifurcations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.time" title="tskit.Tree.time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">time</span></code></a>(u)</p></td>
<td><p>Returns the time of the specified node in generations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.tmrca" title="tskit.Tree.tmrca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tmrca</span></code></a>(u, v)</p></td>
<td><p>Returns the time of the most recent common ancestor of the specified nodes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.unrank" title="tskit.Tree.unrank"><code class="xref py py-obj docutils literal notranslate"><span class="pre">unrank</span></code></a>(num_leaves, rank, *[, span, …])</p></td>
<td><p>Reconstruct the tree of the given <code class="docutils literal notranslate"><span class="pre">rank</span></code> (see <a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.Tree.rank()</span></code></a>) with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p></td>
</tr>
</tbody>
</table>
<p><strong>Attributes</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.index" title="tskit.Tree.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a></p></td>
<td><p>Returns the index this tree occupies in the parent tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interval</span></code></a></p></td>
<td><p>Returns the coordinates of the genomic interval that this tree represents the history of.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.left_root" title="tskit.Tree.left_root"><code class="xref py py-obj docutils literal notranslate"><span class="pre">left_root</span></code></a></p></td>
<td><p>The leftmost root in this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_mutations" title="tskit.Tree.num_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_mutations</span></code></a></p></td>
<td><p>Returns the total number of mutations across all sites on this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.num_nodes" title="tskit.Tree.num_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_nodes</span></code></a></p></td>
<td><p>Returns the number of nodes in the <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> this tree is in.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_roots" title="tskit.Tree.num_roots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_roots</span></code></a></p></td>
<td><p>The number of roots in this tree, as defined in the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> attribute.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.num_sites" title="tskit.Tree.num_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_sites</span></code></a></p></td>
<td><p>Returns the number of sites on this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.root" title="tskit.Tree.root"><code class="xref py py-obj docutils literal notranslate"><span class="pre">root</span></code></a></p></td>
<td><p>The root of this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.root_threshold" title="tskit.Tree.root_threshold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">root_threshold</span></code></a></p></td>
<td><p>Returns the minimum number of samples that a node must be an ancestor of to be considered a potential root.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">roots</span></code></a></p></td>
<td><p>The list of roots in this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.sample_size" title="tskit.Tree.sample_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample_size</span></code></a></p></td>
<td><p>Returns the sample size for this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.span" title="tskit.Tree.span"><code class="xref py py-obj docutils literal notranslate"><span class="pre">span</span></code></a></p></td>
<td><p>Returns the genomic distance that this tree spans.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.total_branch_length" title="tskit.Tree.total_branch_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">total_branch_length</span></code></a></p></td>
<td><p>Returns the sum of all the branch lengths in this tree (in units of generations).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tree_sequence</span></code></a></p></td>
<td><p>Returns the tree sequence that this tree is from.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="tskit.Tree.as_dict_of_dicts">
<code class="sig-name descname">as_dict_of_dicts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.as_dict_of_dicts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.as_dict_of_dicts" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert tree to dict of dicts for conversion to a
<a class="reference external" href="https://networkx.github.io/documentation/stable/reference/classes/digraph.html">networkx graph</a>.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># undirected graphs work as well</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Dictionary of dictionaries of dictionaries where the first key
is the source, the second key is the target of an edge, and the
third key is an edge annotation. At this point the only annotation
is “branch_length”, the length of the branch (in generations).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.branch_length">
<code class="sig-name descname">branch_length</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.branch_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.branch_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the length of the branch (in generations) joining the
specified node to its parent. This is equivalent to</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">u</span><span class="p">))</span> <span class="o">-</span> <span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
</pre></div>
</div>
<p>The branch length for a node that has no parent (e.g., a root) is
defined as zero.</p>
<p>Note that this is not related to the property <cite>.length</cite> which
is a deprecated alias for the genomic <a class="reference internal" href="#tskit.Tree.span" title="tskit.Tree.span"><code class="xref py py-attr docutils literal notranslate"><span class="pre">span</span></code></a> covered by a tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The branch length from u to its parent.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.children">
<code class="sig-name descname">children</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.children"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.children" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the children of the specified node <code class="docutils literal notranslate"><span class="pre">u</span></code> as a tuple of integer node IDs.
If <code class="docutils literal notranslate"><span class="pre">u</span></code> is a leaf, return the empty tuple. The ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The children of <code class="docutils literal notranslate"><span class="pre">u</span></code> as a tuple of integers</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.clear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets this tree back to the initial null state. Calling this method
on a tree already in the null state has no effect.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.copy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this tree. The returned tree will have identical state
to this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A copy of this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.count_topologies">
<code class="sig-name descname">count_topologies</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_sets</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.count_topologies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.count_topologies" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the distribution of embedded topologies for every combination
of the sample sets in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>. <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> defaults to all
samples in the tree grouped by population.</p>
<p><code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> need not include all samples but must be pairwise disjoint.</p>
<p>The returned object is a <a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TopologyCounter</span></code></a> that contains
counts of topologies per combination of sample sets. For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">topology_counter</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">count_topologies</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rank</span><span class="p">,</span> <span class="n">count</span> <span class="o">=</span> <span class="n">topology_counter</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">most_common</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<p>produces the most common tree topology, with populations 0, 1
and 2 as its tips, according to the genealogies of those
populations’ samples in this tree.</p>
<p>The counts for each topology in the <a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TopologyCounter</span></code></a>
are absolute counts that we would get if we were to select all
combinations of samples from the relevant sample sets.
For sample sets <span class="math notranslate nohighlight">\([s_0, s_1, ..., s_n]\)</span>, the total number of
topologies for those sample sets is equal to
<span class="math notranslate nohighlight">\(|s_0| * |s_1| * ... * |s_n|\)</span>, so the counts in the counter
<code class="docutils literal notranslate"><span class="pre">topology_counter[0,</span> <span class="pre">1,</span> <span class="pre">...,</span> <span class="pre">n]</span></code> should sum to
<span class="math notranslate nohighlight">\(|s_0| * |s_1| * ... * |s_n|\)</span>.</p>
<p>To convert the topology counts to probabilities, divide by the total
possible number of sample combinations from the sample sets in question:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">set_sizes</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_set</span><span class="p">)</span> <span class="k">for</span> <span class="n">sample_set</span> <span class="ow">in</span> <span class="n">sample_sets</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">count</span> <span class="o">/</span> <span class="p">(</span><span class="n">set_sizes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">set_sizes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">set_sizes</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.
Defaults to all samples grouped by population.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter">tskit.TopologyCounter</a></p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If nodes in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> are invalid or are
internal samples.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.depth">
<code class="sig-name descname">depth</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.depth"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.depth" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of nodes on the path from <code class="docutils literal notranslate"><span class="pre">u</span></code> to a
root, not including <code class="docutils literal notranslate"><span class="pre">u</span></code>. Thus, the depth of a root is
zero.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The depth of u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.draw">
<code class="sig-name descname">draw</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">height</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node_colours</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mutation_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mutation_colours</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">format</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">edge_colours</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">tree_height_scale</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_tree_height</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.draw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.draw" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a drawing of this tree.</p>
<p>When working in a Jupyter notebook, use the <code class="docutils literal notranslate"><span class="pre">IPython.display.SVG</span></code>
function to display the SVG output from this function inline in the notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">SVG</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">())</span>
</pre></div>
</div>
<p>The unicode format uses unicode <a class="reference external" href="https://en.wikipedia.org/wiki/Box-drawing_character">box drawing characters</a> to render the tree.
This allows rendered trees to be printed out to the terminal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">))</span>
<span class="go">  6</span>
<span class="go">┏━┻━┓</span>
<span class="go">┃   5</span>
<span class="go">┃ ┏━┻┓</span>
<span class="go">┃ ┃  4</span>
<span class="go">┃ ┃ ┏┻┓</span>
<span class="go">3 0 1 2</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">node_labels</span></code> argument allows the user to specify custom labels
for nodes, or no labels at all:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">,</span> <span class="n">node_labels</span><span class="o">=</span><span class="p">{}))</span>
<span class="go">  ┃</span>
<span class="go">┏━┻━┓</span>
<span class="go">┃   ┃</span>
<span class="go">┃ ┏━┻┓</span>
<span class="go">┃ ┃  ┃</span>
<span class="go">┃ ┃ ┏┻┓</span>
<span class="go">┃ ┃ ┃ ┃</span>
</pre></div>
</div>
<p>Note: in some environments such as Jupyter notebooks with Windows or Mac,
users have observed that the Unicode box drawings can be misaligned. In
these cases, we recommend using the SVG or ASCII display formats instead.
If you have a strong preference for aligned Unicode, you can try out the
solution documented
<a class="reference external" href="https://github.com/tskit-dev/tskit/issues/189#issuecomment-499114811">here</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The path to the file to write the output. If None, do not
write to file.</p></li>
<li><p><strong>width</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The width of the image in pixels. If not specified, either
defaults to the minimum size required to depict the tree (text formats)
or 200 pixels.</p></li>
<li><p><strong>height</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The height of the image in pixels. If not specified, either
defaults to the minimum size required to depict the tree (text formats)
or 200 pixels.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – If specified, show custom labels for the nodes
that are present in the map. Any nodes not specified in the map will
not have a node label.</p></li>
<li><p><strong>node_colours</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – If specified, show custom colours for the nodes
given in the map. Any nodes not specified in the map will take the default
colour; a value of <code class="docutils literal notranslate"><span class="pre">None</span></code> is treated as transparent and hence the node
symbol is not plotted. (Only supported in the SVG format.)</p></li>
<li><p><strong>mutation_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – If specified, show custom labels for the mutations
(specified by ID) that are present in the map. Any mutations not in the map
will not have a label. (Showing mutations is currently only supported in the
SVG format)</p></li>
<li><p><strong>mutation_colours</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – If specified, show custom colours for the mutations
given in the map (specified by ID). As for <code class="docutils literal notranslate"><span class="pre">node_colours</span></code>, mutations not
present in the map take the default colour, and those mapping to <code class="docutils literal notranslate"><span class="pre">None</span></code>
are not drawn. (Only supported in the SVG format.)</p></li>
<li><p><strong>format</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The format of the returned image. Currently supported
are ‘svg’, ‘ascii’ and ‘unicode’. Note that the <a class="reference internal" href="#tskit.Tree.draw_svg" title="tskit.Tree.draw_svg"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.draw_svg()</span></code></a>
method provides more comprehensive functionality for creating SVGs.</p></li>
<li><p><strong>edge_colours</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – If specified, show custom colours for the edge
joining each node in the map to its parent. As for <code class="docutils literal notranslate"><span class="pre">node_colours</span></code>,
unspecified edges take the default colour, and <code class="docutils literal notranslate"><span class="pre">None</span></code> values result in the
edge being omitted. (Only supported in the SVG format.)</p></li>
<li><p><strong>tree_height_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Control how height values for nodes are computed.
If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;time&quot;</span></code>, node heights are proportional to their time
values. If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;log_time&quot;</span></code>, node heights are proportional to
their log(time) values. If it is equal to <code class="docutils literal notranslate"><span class="pre">&quot;rank&quot;</span></code>, node heights are spaced
equally according to their ranked times. For SVG output the default is
‘time’-scale whereas for text output the default is ‘rank’-scale.
Time scaling is not currently supported for text output.</p></li>
<li><p><strong>max_tree_height</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The maximum tree height value in the current
scaling system (see <code class="docutils literal notranslate"><span class="pre">tree_height_scale</span></code>). Can be either a string or a
numeric value. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code>, the maximum tree height is set to be
that of the oldest root in the tree. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;ts&quot;</span></code> the maximum
height is set to be the height of the oldest root in the tree sequence;
this is useful when drawing trees from the same tree sequence as it ensures
that node heights are consistent. If a numeric value, this is used as the
maximum tree height by which to scale other nodes. This parameter
is not currently supported for text output.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The left-to-right ordering of child nodes in the drawn tree.
This can be either: <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>, which minimises the differences
between adjacent trees (see also the <code class="docutils literal notranslate"><span class="pre">&quot;minlex_postorder&quot;</span></code> traversal
order for the <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">nodes()</span></code></a> method); or <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> which draws trees
in the left-to-right order defined by the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">quintuply linked tree structure</span></a>.
If not specified or None, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A representation of this tree in the requested format.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.draw_svg">
<code class="sig-name descname">draw_svg</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">tree_height_scale</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_tree_height</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mutation_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">root_svg_attributes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">style</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">force_root_branch</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.draw_svg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.draw_svg" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an SVG representation of a single tree.</p>
<p>When working in a Jupyter notebook, use the <code class="docutils literal notranslate"><span class="pre">IPython.display.SVG</span></code> function
to display the SVG output from this function inline in the notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">SVG</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
<p>The elements in the tree are grouped according to the structure of the tree,
using <a class="reference external" href="https://www.w3.org/TR/SVG2/struct.html#Groups">SVG groups</a>. This allows
easy styling and manipulation of elements and subtrees. Elements in the SVG file
are marked with SVG classes so that they can be targetted, allowing
different components of the drawing to be hidden, styled, or otherwise
manipulated. For example, when drawing (say) the first tree from a tree
sequence, all the SVG components will be placed in a group of class <code class="docutils literal notranslate"><span class="pre">tree</span></code>.
The group will have the additional class <code class="docutils literal notranslate"><span class="pre">t0</span></code>, indicating that this tree
has index 0 in the tree sequence. The general SVG structure is as follows:</p>
<p>The tree is contained in a group of class <code class="docutils literal notranslate"><span class="pre">tree</span></code>. Additionally, this group
has a class <code class="docutils literal notranslate"><span class="pre">tN</span></code> where <cite>N</cite> is the tree index.</p>
<p>Within the <code class="docutils literal notranslate"><span class="pre">tree</span></code> group there is a nested hierarchy of groups corresponding
to the tree structure. Any particular node in the tree will have a corresponding
group containing child groups (if any) followed by the edge above that node, a
node symbol, and (potentially) text containing the node label. For example, a
simple two tip tree, with tip node ids 0 and 1, and a root node id of 2 will have
a structure similar to the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">g</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;tree t0&quot;</span><span class="o">&gt;</span>
  <span class="o">&lt;</span><span class="n">g</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;node n2 root&quot;</span><span class="o">&gt;</span>
    <span class="o">&lt;</span><span class="n">g</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;node n1 a2 i1 p1 sample leaf&quot;</span><span class="o">&gt;</span>
      <span class="o">&lt;</span><span class="n">path</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;edge&quot;</span> <span class="o">...</span> <span class="o">/&gt;</span>
      <span class="o">&lt;</span><span class="n">circle</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;sym&quot;</span> <span class="o">...</span> <span class="o">/&gt;</span>
      <span class="o">&lt;</span><span class="n">text</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;lab&quot;</span> <span class="o">...&gt;</span><span class="n">Node</span> <span class="mi">1</span><span class="o">&lt;/</span><span class="n">text</span><span class="o">&gt;</span>
    <span class="o">&lt;/</span><span class="n">g</span><span class="o">&gt;</span>
    <span class="o">&lt;</span><span class="n">g</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;node n0 a2 i2 p1 sample leaf&quot;</span><span class="o">&gt;</span>
      <span class="o">&lt;</span><span class="n">path</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;edge&quot;</span> <span class="o">...</span> <span class="o">/&gt;</span>
      <span class="o">&lt;</span><span class="n">circle</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;sym&quot;</span> <span class="o">.../&gt;</span>
      <span class="o">&lt;</span><span class="n">text</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;lab&quot;</span> <span class="o">...&gt;</span><span class="n">Node</span> <span class="mi">0</span><span class="o">&lt;/</span><span class="n">text</span><span class="o">&gt;</span>
    <span class="o">&lt;/</span><span class="n">g</span><span class="o">&gt;</span>
    <span class="o">&lt;</span><span class="n">path</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;edge&quot;</span> <span class="o">...</span> <span class="o">/&gt;</span>
    <span class="o">&lt;</span><span class="n">circle</span> <span class="o">/&gt;</span>
    <span class="o">&lt;</span><span class="n">text</span> <span class="n">class</span><span class="o">=</span><span class="s2">&quot;lab&quot;</span><span class="o">&gt;</span><span class="n">Root</span> <span class="p">(</span><span class="n">Node</span> <span class="mi">2</span><span class="p">)</span><span class="o">&lt;/</span><span class="n">text</span><span class="o">&gt;</span>
  <span class="o">&lt;/</span><span class="n">g</span><span class="o">&gt;</span>
<span class="o">&lt;/</span><span class="n">g</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>The classes can be used to manipulate the element, e.g. by using
<a class="reference external" href="https://www.w3.org/TR/SVG2/styling.html">stylesheets</a>. Style strings can
be embedded in the svg by using the <code class="docutils literal notranslate"><span class="pre">style</span></code> parameter, or added to html
pages which contain the raw SVG (e.g. within a Jupyter notebook by using the
IPython <code class="docutils literal notranslate"><span class="pre">HTML()</span></code> function). As a simple example, passing the following
string as the <code class="docutils literal notranslate"><span class="pre">style</span></code> parameter will hide all labels:</p>
<div class="highlight-css notranslate"><div class="highlight"><pre><span></span><span class="p">.</span><span class="nc">tree</span> <span class="p">.</span><span class="nc">lab</span> <span class="p">{</span><span class="k">visibility</span><span class="p">:</span> <span class="kc">hidden</span><span class="p">}</span>
</pre></div>
</div>
<p>You can also change the format of various items: in SVG2-compatible viewers,
the following styles will rotate the leaf nodes labels by 90 degrees, colour
the leaf node symbols blue, and
hide the non-sample node labels. Note that SVG1.1 does not recognize the
<code class="docutils literal notranslate"><span class="pre">transform</span></code> style, so in some SVG viewers, the labels will not appear rotated:
a workaround is to convert the SVG to PDF first, using e.g. the programmable
chromium engine: <code class="docutils literal notranslate"><span class="pre">chromium</span> <span class="pre">--headless</span> <span class="pre">--print-to-pdf=out.pdf</span> <span class="pre">in.svg</span></code>)</p>
<div class="highlight-css notranslate"><div class="highlight"><pre><span></span><span class="p">.</span><span class="nc">tree</span> <span class="p">.</span><span class="nc">node</span><span class="p">.</span><span class="nc">leaf</span> <span class="o">&gt;</span> <span class="p">.</span><span class="nc">lab</span> <span class="p">{</span>
    <span class="k">transform</span><span class="p">:</span> <span class="nb">translateY</span><span class="p">(</span><span class="mf">0.5</span><span class="kt">em</span><span class="p">)</span> <span class="nb">rotate</span><span class="p">(</span><span class="mi">90</span><span class="kt">deg</span><span class="p">);</span> <span class="n">text-anchor</span><span class="p">:</span> <span class="kc">start</span><span class="p">}</span>
<span class="p">.</span><span class="nc">tree</span> <span class="p">.</span><span class="nc">node</span><span class="p">.</span><span class="nc">leaf</span> <span class="o">&gt;</span> <span class="p">.</span><span class="nc">sym</span> <span class="p">{</span><span class="n">fill</span><span class="p">:</span> <span class="kc">blue</span><span class="p">}</span>
<span class="p">.</span><span class="nc">tree</span> <span class="p">.</span><span class="nc">node</span><span class="p">:</span><span class="nd">not</span><span class="o">(</span><span class="p">.</span><span class="nc">sample</span><span class="o">)</span> <span class="o">&gt;</span> <span class="p">.</span><span class="nc">lab</span> <span class="p">{</span><span class="k">visibility</span><span class="p">:</span> <span class="kc">hidden</span><span class="p">}</span>
</pre></div>
</div>
<p>Nodes contain classes that allow them to be targetted by node id (<code class="docutils literal notranslate"><span class="pre">nX</span></code>),
ancestor (parent) id (<code class="docutils literal notranslate"><span class="pre">aX</span></code> or <code class="docutils literal notranslate"><span class="pre">root</span></code> if this node has no parent), and
(if defined) the id of the individual (<code class="docutils literal notranslate"><span class="pre">iX</span></code>) and population (<code class="docutils literal notranslate"><span class="pre">pX</span></code>) to
which this node belongs. Hence the following style will display
a large symbol for node 10, coloured red with a black border, and will also use
thick red lines for all the edges that have it as a direct or indirect parent
(note that, as with the <code class="docutils literal notranslate"><span class="pre">transform</span></code> style, changing the geometrical size of
symbols is only possible in SVG2 and above and therefore not all SVG viewers
will render such symbol size changes correctly).</p>
<div class="highlight-css notranslate"><div class="highlight"><pre><span></span><span class="p">.</span><span class="nc">tree</span> <span class="p">.</span><span class="nc">node</span><span class="p">.</span><span class="nc">n10</span> <span class="o">&gt;</span> <span class="p">.</span><span class="nc">sym</span> <span class="p">{</span><span class="n">fill</span><span class="p">:</span> <span class="kc">red</span><span class="p">;</span> <span class="n">stroke</span><span class="p">:</span> <span class="kc">black</span><span class="p">;</span> <span class="n">r</span><span class="p">:</span> <span class="mi">8</span><span class="kt">px</span><span class="p">}</span>
<span class="p">.</span><span class="nc">tree</span> <span class="p">.</span><span class="nc">node</span><span class="p">.</span><span class="nc">a10</span> <span class="p">.</span><span class="nc">edge</span> <span class="p">{</span><span class="n">stroke</span><span class="p">:</span> <span class="kc">red</span><span class="p">;</span> <span class="n">stroke-width</span><span class="p">:</span> <span class="mi">2</span><span class="kt">px</span><span class="p">}</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A feature of SVG style commands is that they apply not just to the contents
within the &lt;svg&gt; container, but to the entire file. Thus if an SVG file is
embedded in a larger document, such as an HTML file (e.g. when an SVG
is displayed inline in a Jupyter notebook), the style will apply to all SVG
drawings in the notebook. To avoid this, you can tag the SVG with a unique
SVG using <code class="docutils literal notranslate"><span class="pre">root_svg_attributes={'id':'MY_UID'}</span></code>, and prepend this to the
style string, as in <code class="docutils literal notranslate"><span class="pre">#MY_UID</span> <span class="pre">.tree</span> <span class="pre">.edges</span> <span class="pre">{stroke:</span> <span class="pre">gray}</span></code>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The path to the file to write the output. If None, do not
write to file.</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>)</em>) – A tuple of (width, height) giving the width and height of the
produced SVG drawing in abstract user units (usually interpreted as pixels on
initial display).</p></li>
<li><p><strong>tree_height_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Control how height values for nodes are computed.
If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;time&quot;</span></code> (the default), node heights are proportional
to their time values. If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;log_time&quot;</span></code>, node heights are
proportional to their log(time) values. If it is equal to <code class="docutils literal notranslate"><span class="pre">&quot;rank&quot;</span></code>, node
heights are spaced equally according to their ranked times.</p></li>
<li><p><strong>max_tree_height</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The maximum tree height value in the current
scaling system (see <code class="docutils literal notranslate"><span class="pre">tree_height_scale</span></code>). Can be either a string or a
numeric value. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> (the default), the maximum tree height
is set to be that of the oldest root in the tree. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;ts&quot;</span></code> the
maximum height is set to be the height of the oldest root in the tree
sequence; this is useful when drawing trees from the same tree sequence as it
ensures that node heights are consistent. If a numeric value, this is used as
the maximum tree height by which to scale other nodes.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – If specified, show custom labels for the nodes
(specified by ID) that are present in this map; any nodes not present will
not have a label.</p></li>
<li><p><strong>mutation_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – If specified, show custom labels for the
mutations (specified by ID) that are present in the map; any mutations
not present will not have a label.</p></li>
<li><p><strong>root_svg_attributes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – Additional attributes, such as an id, that will
be embedded in the root <code class="docutils literal notranslate"><span class="pre">&lt;svg&gt;</span></code> tag of the generated drawing.</p></li>
<li><p><strong>style</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A
<a class="reference external" href="https://www.w3.org/TR/CSS22/syndata.html">css style string</a> that will be
included in the <code class="docutils literal notranslate"><span class="pre">&lt;style&gt;</span></code> tag of the generated svg. Note that certain
styles, in particular transformations and changes in geometrical properties
of objects, will only be recognised by SVG2-compatible viewers.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string specifying the traversal type used to order the tips
in the tree, as detailed in <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a>. If None (default), use
the default order as described in that method.</p></li>
<li><p><strong>force_root_branch</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> always plot a branch (edge) above the
root(s) in the tree. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) then only plot such root branches
if there is a mutation above a root of the tree.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An SVG representation of a tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.first">
<code class="sig-name descname">first</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.first"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.first" title="Permalink to this definition">¶</a></dt>
<dd><p>Seeks to the first tree in the sequence. This can be called whether
the tree is in the null state or not.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.generate_balanced">
<em class="property">static </em><code class="sig-name descname">generate_balanced</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_leaves</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">arity</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_balanced"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.generate_balanced" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a :class:&lt;Tree&gt; with the specified number of leaves that is maximally
balanced. By default, the tree returned is binary, such that for each
node that subtends <span class="math notranslate nohighlight">\(n\)</span> leaves, the left child will subtend
<span class="math notranslate nohighlight">\(\floor{n / 2}\)</span> leaves and the right child the remainder. Balanced
trees with higher arity can also generated using the <code class="docutils literal notranslate"><span class="pre">arity</span></code> parameter,
where the leaves subtending a node are distributed among its children
analogously.</p>
<p>In the returned tree, the leaf nodes are all at time 0, marked as samples,
and labelled 0 to n from left-to-right. Internal node IDs are assigned
sequentially from n in a postorder traversal, and the time of an internal
node is the maximum time of its children plus the specified <code class="docutils literal notranslate"><span class="pre">branch_length</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of leaf nodes in the returned tree (must be
be 2 or greater).</p></li>
<li><p><strong>arity</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The maximum number of children a node can have in the returned
tree.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned :class:&lt;Tree&gt;.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The minimum length of a branch in the tree (see
above for details on how internal node times are assigned).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A balanced tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> is available
via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.generate_comb">
<em class="property">static </em><code class="sig-name descname">generate_comb</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_leaves</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_comb"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.generate_comb" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a :class:&lt;Tree&gt; in which all internal nodes have two children
and the left child is a leaf. This is a “comb”, “ladder” or “pectinate”
phylogeny, and also known as a <a class="reference external" href="https://en.wikipedia.org/wiki/Caterpillar_tree">caterpiller tree</a>.</p>
<p>The leaf nodes are all at time 0, marked as samples,
and labelled 0 to n from left-to-right. Internal node IDs are assigned
sequentially from n as we ascend the tree, and the time of an internal
node is the maximum time of its children plus the specified <code class="docutils literal notranslate"><span class="pre">branch_length</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of leaf nodes in the returned tree (must be
be 2 or greater).</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned :class:&lt;Tree&gt;.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The length of every branch in the tree (equivalent
to the time of the root node).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A star-shaped tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> is available
via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.generate_random">
<em class="property">static </em><code class="sig-name descname">generate_random</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_leaves</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">arity</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">random_seed</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_random"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.generate_random" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a random :class:&lt;Tree&gt; with <span class="math notranslate nohighlight">\(n\)</span> = <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code>
leaves with an equal probability of returning any valid topology.</p>
<p>The leaf nodes are marked as samples, labelled 0 to n, and placed at time 0.
The root node is placed at time <code class="docutils literal notranslate"><span class="pre">(num_leaves</span> <span class="pre">-</span> <span class="pre">1)</span> <span class="pre">*</span> <span class="pre">branch_length</span></code>, and
the other non-leaf nodes placed at a time of <code class="docutils literal notranslate"><span class="pre">branch_length</span></code> from each
other and from the root.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The returned tree has not been created under any explicit model of
evolution. In order to simulate such trees, additional software
such as <cite>msprime &lt;https://github.com/tskit-dev/msprime&gt;`</cite> is required.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of leaf nodes in the returned tree (must
be 2 or greater).</p></li>
<li><p><strong>arity</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of children of each internal node. If this is
2 (the default) then a strictly bifurcating (binary) tree is generated
chosen at random from the <span class="math notranslate nohighlight">\((2n - 3)! / (2^(n - 2) (n - 2!))\)</span>
possible topologies.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned :class:&lt;Tree&gt;.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The time separating successive non-leaf nodes
from each other.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A random tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> is available
via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.generate_star">
<em class="property">static </em><code class="sig-name descname">generate_star</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_leaves</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_star"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.generate_star" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a :class:&lt;Tree&gt; whose leaf nodes all have the same parent (i.e.
a “star” tree). The leaf nodes are all at time 0 and are marked as sample nodes.</p>
<p>The tree produced by this method is identical to
<code class="docutils literal notranslate"><span class="pre">tskit.Tree.unrank(n,</span> <span class="pre">(0,</span> <span class="pre">0))</span></code>, but generated more efficiently for large <code class="docutils literal notranslate"><span class="pre">n</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of leaf nodes in the returned tree (must be
be 2 or greater).</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned :class:&lt;Tree&gt;.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The length of every branch in the tree (equivalent
to the time of the root node).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A star-shaped tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> is available
via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.index">
<em class="property">property </em><code class="sig-name descname">index</code><a class="headerlink" href="#tskit.Tree.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index this tree occupies in the parent tree sequence.
This index is zero based, so the first tree in the sequence has index 0.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The index of this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.interval">
<em class="property">property </em><code class="sig-name descname">interval</code><a class="headerlink" href="#tskit.Tree.interval" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the coordinates of the genomic interval that this tree
represents the history of. The interval is returned as a tuple
<span class="math notranslate nohighlight">\((l, r)\)</span> and is a half-open interval such that the left
coordinate is inclusive and the right coordinate is exclusive. This
tree therefore applies to all genomic locations <span class="math notranslate nohighlight">\(x\)</span> such that
<span class="math notranslate nohighlight">\(l \leq x &lt; r\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A named tuple (l, r) representing the left-most (inclusive)
and right-most (exclusive) coordinates of the genomic region
covered by this tree. The coordinates can be accessed by index
(<code class="docutils literal notranslate"><span class="pre">0</span></code> or <code class="docutils literal notranslate"><span class="pre">1</span></code>) or equivalently by name (<code class="docutils literal notranslate"><span class="pre">.left</span></code> or <code class="docutils literal notranslate"><span class="pre">.right</span></code>)</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.is_descendant">
<code class="sig-name descname">is_descendant</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em>, <em class="sig-param"><span class="n">v</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_descendant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.is_descendant" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if the specified node u is a descendant of node v and False
otherwise. A node <span class="math notranslate nohighlight">\(u\)</span> is a descendant of another node <span class="math notranslate nohighlight">\(v\)</span> if
<span class="math notranslate nohighlight">\(v\)</span> is on the path from <span class="math notranslate nohighlight">\(u\)</span> to root. A node is considered
to be a descendant of itself, so <code class="docutils literal notranslate"><span class="pre">tree.is_descendant(u,</span> <span class="pre">u)</span></code> will be
True for any valid node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The descendant node.</p></li>
<li><p><strong>v</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ancestral node.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if u is a descendant of v.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If u or v are not valid node IDs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.is_internal">
<code class="sig-name descname">is_internal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_internal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.is_internal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if the specified node is not a leaf. A node is internal
if it has one or more children in the current tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if u is not a leaf node.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.is_isolated">
<code class="sig-name descname">is_isolated</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_isolated"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.is_isolated" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if the specified node is isolated in this tree: that is
it has no parents and no children. Sample nodes that are isolated
and have no mutations above them are used to represent
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if u is an isolated node.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.is_leaf">
<code class="sig-name descname">is_leaf</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_leaf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.is_leaf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if the specified node is a leaf. A node <span class="math notranslate nohighlight">\(u\)</span> is a
leaf if it has zero children.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if u is a leaf node.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.is_sample">
<code class="sig-name descname">is_sample</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.is_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if the specified node is a sample. A node <span class="math notranslate nohighlight">\(u\)</span> is a
sample if it has been marked as a sample in the parent tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if u is a sample.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.kc_distance">
<code class="sig-name descname">kc_distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">lambda_</span><span class="o">=</span><span class="default_value">0.0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.kc_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.kc_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the Kendall-Colijn distance between the specified pair of trees.
The <code class="docutils literal notranslate"><span class="pre">lambda_</span></code> parameter  determines the relative weight of topology
vs branch lengths in calculating the distance. If <code class="docutils literal notranslate"><span class="pre">lambda_</span></code> is 0
(the default) we only consider topology, and if it is 1 we only
consider branch lengths. See <a class="reference external" href="https://academic.oup.com/mbe/article/33/10/2735/2925548">Kendall &amp; Colijn (2016)</a> for details.</p>
<p>The trees we are comparing to must have identical lists of sample
nodes (i.e., the same IDs in the same order). The metric operates on
samples, not leaves, so internal samples are treated identically to
sample tips. Subtrees with no samples do not contribute to the metric.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><em>Tree</em></a>) – The other tree to compare to.</p></li>
<li><p><strong>lambda</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The KC metric lambda parameter determining the
relative weight of topology and branch length.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The computed KC distance between this tree and other.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.last">
<code class="sig-name descname">last</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.last"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.last" title="Permalink to this definition">¶</a></dt>
<dd><p>Seeks to the last tree in the sequence. This can be called whether
the tree is in the null state or not.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.leaves">
<code class="sig-name descname">leaves</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.leaves"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.leaves" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over all the leaves in this tree that are
underneath the specified node. If u is not specified, return all leaves
in the tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An iterator over all leaves in the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.9)">collections.abc.Iterable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.left_child">
<code class="sig-name descname">left_child</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.left_child"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.left_child" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the leftmost child of the specified node. Returns
<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u is a leaf or is not a node in
the current tree. The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<p>This is a low-level method giving access to the quintuply linked
tree structure in memory; the <a class="reference internal" href="#tskit.Tree.children" title="tskit.Tree.children"><code class="xref py py-meth docutils literal notranslate"><span class="pre">children()</span></code></a> method is a more
convenient way to obtain the children of a given node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The leftmost child of u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.left_root">
<em class="property">property </em><code class="sig-name descname">left_root</code><a class="headerlink" href="#tskit.Tree.left_root" title="Permalink to this definition">¶</a></dt>
<dd><p>The leftmost root in this tree. If there are multiple roots
in this tree, they are siblings of this node, and so we can
use <a class="reference internal" href="#tskit.Tree.right_sib" title="tskit.Tree.right_sib"><code class="xref py py-meth docutils literal notranslate"><span class="pre">right_sib()</span></code></a> to iterate over all roots:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">u</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">left_root</span>
<span class="k">while</span> <span class="n">u</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Root:&quot;</span><span class="p">,</span> <span class="n">u</span><span class="p">)</span>
    <span class="n">u</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">right_sib</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
</pre></div>
</div>
<p>The left-to-right ordering of roots is arbitrary and should
not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a>
section for details.</p>
<p>This is a low-level method giving access to the quintuply linked
tree structure in memory; the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> attribute is a more
convenient way to obtain the roots of a tree. If you are assuming
that there is a single root in the tree you should use the
<a class="reference internal" href="#tskit.Tree.root" title="tskit.Tree.root"><code class="xref py py-attr docutils literal notranslate"><span class="pre">root</span></code></a> property.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Do not use this property if you are assuming that there
is a single root in trees that are being processed. The
<a class="reference internal" href="#tskit.Tree.root" title="tskit.Tree.root"><code class="xref py py-attr docutils literal notranslate"><span class="pre">root</span></code></a> property should be used in this case, as it will
raise an error when multiple roots exists.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.left_sib">
<code class="sig-name descname">left_sib</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.left_sib"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.left_sib" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sibling node to the left of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>
if u does not have a left sibling.
The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sibling node to the left of u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.map_mutations">
<code class="sig-name descname">map_mutations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genotypes</span></em>, <em class="sig-param"><span class="n">alleles</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.map_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.map_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Given observations for the samples in this tree described by the specified
set of genotypes and alleles, return a parsimonious set of state transitions
explaining these observations. The genotypes array is interpreted as indexes
into the alleles list in the same manner as described in the
<a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a> method. Thus, if sample <code class="docutils literal notranslate"><span class="pre">j</span></code> carries the
allele at index <code class="docutils literal notranslate"><span class="pre">k</span></code>, then we have <code class="docutils literal notranslate"><span class="pre">genotypes[j]</span> <span class="pre">=</span> <span class="pre">k</span></code>.
Missing observations can be specified for a sample using the value
<code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code> (-1), in which case the state at this sample does not
influence the ancestral state or the position of mutations returned. At least
one non-missing observation must be provided. A maximum of 64 alleles are
supported.</p>
<p>The current implementation uses the Fitch parsimony algorithm to determine
the minimum number of state transitions required to explain the data. In this
model, transitions between any of the non-missing states is equally likely.</p>
<p>The returned values correspond directly to the data model for describing
variation at sites using mutations. See the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a>
and <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> definitions for details and background.</p>
<p>The state reconstruction is returned as two-tuple, <code class="docutils literal notranslate"><span class="pre">(ancestral_state,</span>
<span class="pre">mutations)</span></code>, where <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> is the allele assigned to the
tree root(s) and <code class="docutils literal notranslate"><span class="pre">mutations</span></code> is a list of <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> objects,
ordered as <a class="reference internal" href="data-model.html#sec-mutation-requirements"><span class="std std-ref">required in a mutation table</span></a>.
For each mutation, <code class="docutils literal notranslate"><span class="pre">node</span></code> is the tree node at the bottom of the branch
on which the transition occurred, and <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> is the new state
after this mutation. The <code class="docutils literal notranslate"><span class="pre">parent</span></code> property contains the index in the
returned list of the previous mutation on the path to root, or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code>
if there are no previous mutations (see the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a>
for more information on the concept of mutation parents). All other attributes
of the <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> object are undefined and should not be used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Sample states observed as missing in the input <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> need
not correspond to samples whose nodes are actually “missing” (i.e.
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">isolated</span></a>) in the tree. In this
case, mapping the mutations returned by this method onto the tree
will result in these missing observations being imputed to the
most parsimonious state.</p>
</div>
<p>See the <a class="reference internal" href="tutorial.html#sec-tutorial-parsimony"><span class="std std-ref">Parsimony</span></a> section in the tutorial for examples
of how to use this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genotypes</strong> (<em>array_like</em>) – The input observations for the samples in this tree.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – The alleles for the specified <code class="docutils literal notranslate"><span class="pre">genotypes</span></code>. Each
positive value in the <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> array is treated as an index into this
list of alleles.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The inferred ancestral state and list of mutations on this tree
that encode the specified observations.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation">tskit.Mutation</a>))</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.mrca">
<code class="sig-name descname">mrca</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em>, <em class="sig-param"><span class="n">v</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.mrca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.mrca" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the most recent common ancestor of the specified nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The first node.</p></li>
<li><p><strong>v</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The second node.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The most recent common ancestor of u and v.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.mutations">
<code class="sig-name descname">mutations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over all the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree.
Mutations are returned in their
<a class="reference internal" href="data-model.html#sec-mutation-requirements"><span class="std std-ref">order in the mutations table</span></a>,
that is, by nondecreasing site ID, and within a site, by decreasing
mutation time with parent mutations before their children.
See the <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> class for details on the available fields for
each mutation.</p>
<p>The returned iterator is equivalent to iterating over all sites
and all mutations in each site, i.e.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="k">yield</span> <span class="n">mutation</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterator over all <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> objects in this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>iter(<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.newick">
<code class="sig-name descname">newick</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">precision</span><span class="o">=</span><span class="default_value">14</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">root</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">include_branch_lengths</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.newick"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.newick" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">newick encoding</a>
of this tree. If the <code class="docutils literal notranslate"><span class="pre">root</span></code> argument is specified, return a representation
of the specified subtree, otherwise the full tree is returned. If the tree
has multiple roots then seperate newick strings for each rooted subtree
must be found (i.e., we do not attempt to concatenate the different trees).</p>
<p>By default, leaf nodes are labelled with their numerical ID + 1,
and internal nodes are not labelled. Arbitrary node labels can be specified
using the <code class="docutils literal notranslate"><span class="pre">node_labels</span></code> argument, which maps node IDs to the desired
labels.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Node labels are <strong>not</strong> Newick escaped, so care must be taken
to provide labels that will not break the encoding.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>precision</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The numerical precision with which branch lengths are
printed.</p></li>
<li><p><strong>root</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – If specified, return the tree rooted at this node.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – If specified, show custom labels for the nodes
that are present in the map. Any nodes not specified in the map will
not have a node label.</p></li>
<li><p><strong>include_branch_lengths</strong> – If True (default), output branch lengths in the
Newick string. If False, only output the topology, without branch lengths.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A newick representation of this tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.next">
<code class="sig-name descname">next</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.next"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.next" title="Permalink to this definition">¶</a></dt>
<dd><p>Seeks to the next tree in the sequence. If the tree is in the initial
null state we seek to the first tree (equivalent to calling <a class="reference internal" href="#tskit.Tree.first" title="tskit.Tree.first"><code class="xref py py-meth docutils literal notranslate"><span class="pre">first()</span></code></a>).
Calling <code class="docutils literal notranslate"><span class="pre">next</span></code> on the last tree in the sequence results in the tree
being cleared back into the null initial state (equivalent to calling
<a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clear()</span></code></a>). The return value of the function indicates whether the
tree is in a non-null state, and can be used to loop over the trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Iterate over the trees from left-to-right</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">tree_sequence</span><span class="p">)</span>
<span class="k">while</span> <span class="n">tree</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
    <span class="c1"># Do something with the tree.</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="c1"># tree is now back in the null state.</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the tree has been transformed into one of the trees
in the sequence; False if the tree has been transformed into the
null state.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.nodes">
<code class="sig-name descname">nodes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">root</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">'preorder'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.nodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over the node IDs in this tree. If the root parameter
is provided, iterate over the node IDs in the subtree rooted at this
node. If this is None, iterate over all node IDs. If the order parameter
is provided, iterate over the nodes in required tree traversal order.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Unlike the <a class="reference internal" href="#tskit.TreeSequence.nodes" title="tskit.TreeSequence.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.nodes()</span></code></a> method, this iterator produces
integer node IDs, not <a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a> objects.</p>
</div>
<p>The currently implemented traversal orders are:</p>
<blockquote>
<div><ul class="simple">
<li><p>‘preorder’: starting at root, yield the current node, then recurse
and do a preorder on each child of the current node. See also <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Pre-order_(NLR)">Wikipedia</a>.</p></li>
<li><p>‘inorder’: starting at root, assuming binary trees, recurse and do
an inorder on the first child, then yield the current node, then
recurse and do an inorder on the second child. In the case of <code class="docutils literal notranslate"><span class="pre">n</span></code>
child nodes (not necessarily 2), the first <code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">//</span> <span class="pre">2</span></code> children are
visited in the first stage, and the remaining <code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">-</span> <span class="pre">n</span> <span class="pre">//</span> <span class="pre">2</span></code> children
are visited in the second stage. See also <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#In-order_(LNR)">Wikipedia</a>.</p></li>
<li><p>‘postorder’: starting at root, recurse and do a postorder on each
child of the current node, then yield the current node. See also
<a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Post-order_(LRN)">Wikipedia</a>.</p></li>
<li><p>‘levelorder’ (‘breadthfirst’): visit the nodes under root (including
the root) in increasing order of their depth from root. See also
<a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Breadth-first_search_/_level_order">Wikipedia</a>.</p></li>
<li><p>‘timeasc’: visits the nodes in order of increasing time, falling back to
increasing ID if times are equal.</p></li>
<li><p>‘timedesc’: visits the nodes in order of decreasing time, falling back to
decreasing ID if times are equal.</p></li>
<li><p>‘minlex_postorder’: a usual postorder has ambiguity in the order in
which children of a node are visited. We constrain this by outputting
a postorder such that the leaves visited, when their IDs are
listed out, have minimum <a class="reference external" href="https://en.wikipedia.org/wiki/Lexicographical_order">lexicographic order</a> out of all valid
traversals. This traversal is useful for drawing multiple trees of
a <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code>, as it leads to more consistency between adjacent
trees. Note that internal non-leaf nodes are not counted in
assessing the lexicographic order.</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>root</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The root of the subtree we are traversing.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The traversal ordering. Currently ‘preorder’,
‘inorder’, ‘postorder’, ‘levelorder’ (‘breadthfirst’), ‘timeasc’ and
‘timedesc’ and ‘minlex_postorder’ are supported.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An iterator over the node IDs in the tree in some traversal order.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.9)">collections.abc.Iterable</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.num_children">
<code class="sig-name descname">num_children</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.num_children"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.num_children" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of children of the specified
node (i.e. <code class="docutils literal notranslate"><span class="pre">len(tree.children(u))</span></code>)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of immediate children of the node u in this tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.num_mutations">
<em class="property">property </em><code class="sig-name descname">num_mutations</code><a class="headerlink" href="#tskit.Tree.num_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of mutations across all sites on this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The total number of mutations over all sites on this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.num_nodes">
<em class="property">property </em><code class="sig-name descname">num_nodes</code><a class="headerlink" href="#tskit.Tree.num_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of nodes in the <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> this tree is in.
Equivalent to <code class="docutils literal notranslate"><span class="pre">tree.tree_sequence.num_nodes</span></code>. To find the number of
nodes that are reachable from all roots use <code class="docutils literal notranslate"><span class="pre">len(list(tree.nodes()))</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.num_roots">
<em class="property">property </em><code class="sig-name descname">num_roots</code><a class="headerlink" href="#tskit.Tree.num_roots" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of roots in this tree, as defined in the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a>
attribute.</p>
<p>Requires O(number of roots) time.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.num_samples">
<code class="sig-name descname">num_samples</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.num_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.num_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of samples in this tree underneath the specified
node (including the node itself). If u is not specified return
the total number of samples in the tree.</p>
<p>This is a constant time operation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of samples in the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.num_sites">
<em class="property">property </em><code class="sig-name descname">num_sites</code><a class="headerlink" href="#tskit.Tree.num_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of sites on this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of sites on this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.num_tracked_samples">
<code class="sig-name descname">num_tracked_samples</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.num_tracked_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.num_tracked_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of samples in the set specified in the
<code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameter of the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method
underneath the specified node. If the input node is not specified,
return the total number of tracked samples in the tree.</p>
<p>This is a constant time operation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of samples within the set of tracked samples in
the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.parent">
<code class="sig-name descname">parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.parent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the parent of the specified node. Returns
<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u is a root or is not a node in
the current tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The parent of u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.population">
<code class="sig-name descname">population</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.population"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.population" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the population associated with the specified node.
Equivalent to <code class="docutils literal notranslate"><span class="pre">tree.tree_sequence.node(u).population</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the population associated with node u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.prev">
<code class="sig-name descname">prev</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.prev"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.prev" title="Permalink to this definition">¶</a></dt>
<dd><p>Seeks to the previous tree in the sequence. If the tree is in the initial
null state we seek to the last tree (equivalent to calling <a class="reference internal" href="#tskit.Tree.last" title="tskit.Tree.last"><code class="xref py py-meth docutils literal notranslate"><span class="pre">last()</span></code></a>).
Calling <code class="docutils literal notranslate"><span class="pre">prev</span></code> on the first tree in the sequence results in the tree
being cleared back into the null initial state (equivalent to calling
<a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clear()</span></code></a>). The return value of the function indicates whether the
tree is in a non-null state, and can be used to loop over the trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Iterate over the trees from right-to-left</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">tree_sequence</span><span class="p">)</span>
<span class="k">while</span> <span class="n">tree</span><span class="o">.</span><span class="n">prev</span><span class="p">()</span>
    <span class="c1"># Do something with the tree.</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="c1"># tree is now back in the null state.</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the tree has been transformed into one of the trees
in the sequence; False if the tree has been transformed into the
null state.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.rank">
<code class="sig-name descname">rank</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.rank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.rank" title="Permalink to this definition">¶</a></dt>
<dd><p>Produce the rank of this tree in the enumeration of all leaf-labelled
trees of n leaves. See the <a class="reference internal" href="combinatorics.html#sec-tree-ranks"><span class="std std-ref">Interpreting Tree Ranks</span></a> section for
details on ranking and unranking trees.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If the tree has multiple roots.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.right_child">
<code class="sig-name descname">right_child</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.right_child"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.right_child" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the rightmost child of the specified node. Returns
<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u is a leaf or is not a node in
the current tree. The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<p>This is a low-level method giving access to the quintuply linked
tree structure in memory; the <a class="reference internal" href="#tskit.Tree.children" title="tskit.Tree.children"><code class="xref py py-meth docutils literal notranslate"><span class="pre">children()</span></code></a> method is a more
convenient way to obtain the children of a given node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The rightmost child of u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.right_sib">
<code class="sig-name descname">right_sib</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.right_sib"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.right_sib" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sibling node to the right of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>
if u does not have a right sibling.
The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sibling node to the right of u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.root">
<em class="property">property </em><code class="sig-name descname">root</code><a class="headerlink" href="#tskit.Tree.root" title="Permalink to this definition">¶</a></dt>
<dd><p>The root of this tree. If the tree contains multiple roots, a ValueError is
raised indicating that the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> attribute should be used instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The root node.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code></a> if this tree contains more than one root.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.root_threshold">
<em class="property">property </em><code class="sig-name descname">root_threshold</code><a class="headerlink" href="#tskit.Tree.root_threshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the minimum number of samples that a node must be an ancestor
of to be considered a potential root.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The root threshold.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.roots">
<em class="property">property </em><code class="sig-name descname">roots</code><a class="headerlink" href="#tskit.Tree.roots" title="Permalink to this definition">¶</a></dt>
<dd><p>The list of roots in this tree. A root is defined as a unique endpoint of
the paths starting at samples. We can define the set of roots as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">roots</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree_sequence</span><span class="o">.</span><span class="n">samples</span><span class="p">():</span>
    <span class="k">while</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
        <span class="n">u</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
    <span class="n">roots</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
<span class="c1"># roots is now the set of all roots in this tree.</span>
<span class="k">assert</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">roots</span><span class="p">)</span> <span class="o">==</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">roots</span><span class="p">)</span>
</pre></div>
</div>
<p>The roots of the tree are returned in a list, in no particular order.</p>
<p>Requires O(number of roots) time.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The list of roots in this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.sample_size">
<em class="property">property </em><code class="sig-name descname">sample_size</code><a class="headerlink" href="#tskit.Tree.sample_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sample size for this tree. This is the number of sample
nodes in the tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of sample nodes in the tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.samples">
<code class="sig-name descname">samples</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over all the samples in this tree that are
underneath the specified node. If u is a sample, it is included in the
returned iterator. If u is not specified, return all samples in the tree.</p>
<p>If the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method is called with
<code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>, this method uses an efficient algorithm to find
the samples. If not, a simple traversal based method is used.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An iterator over all samples in the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.9)">collections.abc.Iterable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.seek">
<code class="sig-name descname">seek</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.seek"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.seek" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the state to represent the tree that covers the specified
position in the parent tree sequence. After a successful return
of this method we have <code class="docutils literal notranslate"><span class="pre">tree.interval.left</span></code> &lt;= <code class="docutils literal notranslate"><span class="pre">position</span></code>
&lt; <code class="docutils literal notranslate"><span class="pre">tree.interval.right</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The position along the sequence length to
seek to.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.9)"><strong>ValueError</strong></a> – If 0 &lt; position or position &gt;=
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.seek_index">
<code class="sig-name descname">seek_index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.seek_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.seek_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the state to represent the tree at the specified
index in the parent tree sequence. Negative indexes following the
standard Python conventions are allowed, i.e., <code class="docutils literal notranslate"><span class="pre">index=-1</span></code> will
seek to the last tree in the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The tree index to seek to.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#IndexError" title="(in Python v3.9)"><strong>IndexError</strong></a> – If an index outside the acceptable range is provided.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.sites">
<code class="sig-name descname">sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a>
in this tree. Sites are returned in order of increasing ID
(and also position). See the <a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> class for details on
the available fields for each site.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An iterator over all sites in this tree.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.span">
<em class="property">property </em><code class="sig-name descname">span</code><a class="headerlink" href="#tskit.Tree.span" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the genomic distance that this tree spans.
This is defined as <span class="math notranslate nohighlight">\(r - l\)</span>, where <span class="math notranslate nohighlight">\((l, r)\)</span> is the genomic
interval returned by <a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The genomic distance covered by this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.split_polytomies">
<code class="sig-name descname">split_polytomies</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">epsilon</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">random_seed</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.split_polytomies"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.split_polytomies" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> where extra nodes and edges have been inserted
so that any any node <code class="docutils literal notranslate"><span class="pre">u</span></code> with greater than 2 children — a multifurcation
or “polytomy” — is resolved into successive bifurcations. New nodes are
inserted at times fractionally less than than the time of node <code class="docutils literal notranslate"><span class="pre">u</span></code>
(controlled by the <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> parameter).</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">method</span></code> is <code class="docutils literal notranslate"><span class="pre">&quot;random&quot;</span></code> (currently the only option, and the default
when no method is specified), then for a node with <span class="math notranslate nohighlight">\(n\)</span> children, the
<span class="math notranslate nohighlight">\((2n - 3)! / (2^(n - 2) (n - 2!))\)</span> possible binary trees with equal
probability.</p>
<p>The returned :class`.Tree` will have the same genomic span as this tree,
and node IDs will be conserved (that is, node <code class="docutils literal notranslate"><span class="pre">u</span></code> in this tree will
be the same node in the returned tree). The returned tree is derived from a
tree sequence that contains only one non-degenerate tree, that is, where
edges cover only the interval spanned by this tree.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A tree sequence <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">requires</span></a> that
parents be older than children and that mutations are younger than the
parent of the edge on which they lie. If <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> is not small enough,
compared to the distance between a polytomy and its oldest child (or oldest
child mutation) these requirements may not be met. In this case an error is
raised, recommending a smaller epsilon value be used.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>epsilon</strong> – A small time period used to separate each newly inserted node.
For a given polytomy of degree <span class="math notranslate nohighlight">\(n\)</span>, the <span class="math notranslate nohighlight">\(n-2\)</span> extra nodes are
inserted with the oldest at time <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> less than the original parent,
<code class="docutils literal notranslate"><span class="pre">u</span></code>, and successive nodes at time <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> from each other. Times
are allocated to different levels of the tree, such that any newly
inserted sibling nodes will have the same time. (Default <span class="math notranslate nohighlight">\(1e-10\)</span>).</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The method used to break polytomies. Currently only “random”
is supported, which can also be specified by <code class="docutils literal notranslate"><span class="pre">method=None</span></code>
(Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p></li>
<li><p><strong>random_seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The random seed. If this is None, a random seed will
be automatically generated. Valid random seeds must be between 1 and
<span class="math notranslate nohighlight">\(2^32 − 1\)</span>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new tree with polytomies split into random bifurcations.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.time">
<code class="sig-name descname">time</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.time"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.time" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the time of the specified node in generations.
Equivalent to <code class="docutils literal notranslate"><span class="pre">tree.tree_sequence.node(u).time</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node of interest.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The time of u.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.tmrca">
<code class="sig-name descname">tmrca</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">u</span></em>, <em class="sig-param"><span class="n">v</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.tmrca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.tmrca" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the time of the most recent common ancestor of the specified
nodes. This is equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">mrca</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">))</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The first node.</p></li>
<li><p><strong>v</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The second node.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The time of the most recent common ancestor of u and v.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.total_branch_length">
<em class="property">property </em><code class="sig-name descname">total_branch_length</code><a class="headerlink" href="#tskit.Tree.total_branch_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sum of all the branch lengths in this tree (in
units of generations). This is equivalent to</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">sum</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">branch_length</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span>
</pre></div>
</div>
<p>Note that the branch lengths for root nodes are defined as zero.</p>
<p>As this is defined by a traversal of the tree, technically we
return the sum of all branch lengths that are reachable from
roots. Thus, this is the sum of all branches that are ancestral
to at least one sample. This distinction is only important
in tree sequences that contain ‘dead branches’, i.e., those
that define topology not ancestral to any samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The sum of lengths of branches in this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.tree_sequence">
<em class="property">property </em><code class="sig-name descname">tree_sequence</code><a class="headerlink" href="#tskit.Tree.tree_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the tree sequence that this tree is from.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The parent tree sequence for this tree.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.Tree.unrank">
<em class="property">static </em><code class="sig-name descname">unrank</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_leaves</span></em>, <em class="sig-param"><span class="n">rank</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.unrank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Tree.unrank" title="Permalink to this definition">¶</a></dt>
<dd><p>Reconstruct the tree of the given <code class="docutils literal notranslate"><span class="pre">rank</span></code>
(see <a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.Tree.rank()</span></code></a>) with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.
The labels and times of internal nodes are assigned by a postorder
traversal of the nodes, such that the time of each internal node
is the maximum time of its children plus the specified <code class="docutils literal notranslate"><span class="pre">branch_length</span></code>.
The time of each leaf is 0.</p>
<p>See the <a class="reference internal" href="combinatorics.html#sec-tree-ranks"><span class="std std-ref">Interpreting Tree Ranks</span></a> section for details on ranking and
unranking trees and what constitutes valid ranks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of leaves of the tree to generate.</p></li>
<li><p><strong>rank</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>)</em>) – The rank of the tree to generate.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The genomic span of the returned tree. The tree will cover
the interval <span class="math notranslate nohighlight">\([0, \text{span})\)</span> and the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
from which the tree is taken will have its
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> equal to <code class="docutils literal notranslate"><span class="pre">span</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>float branch_length: The minimum length of a branch in this tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p>ValueError: If the given rank is out of bounds for trees
with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="constants">
<h3>Constants<a class="headerlink" href="#constants" title="Permalink to this headline">¶</a></h3>
<dl class="py data">
<dt id="tskit.NULL">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">NULL</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><em class="property"> = -1</em><a class="headerlink" href="#tskit.NULL" title="Permalink to this definition">¶</a></dt>
<dd><p>Special reserved value representing a null ID.</p>
</dd></dl>

<dl class="py data">
<dt id="tskit.NODE_IS_SAMPLE">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">NODE_IS_SAMPLE</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><em class="property"> = 1</em><a class="headerlink" href="#tskit.NODE_IS_SAMPLE" title="Permalink to this definition">¶</a></dt>
<dd><p>Node flag value indicating that it is a sample.</p>
</dd></dl>

<dl class="py data">
<dt id="tskit.MISSING_DATA">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">MISSING_DATA</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><em class="property"> = -1</em><a class="headerlink" href="#tskit.MISSING_DATA" title="Permalink to this definition">¶</a></dt>
<dd><p>Special value representing missing data in a genotype array</p>
</dd></dl>

<dl class="py data">
<dt id="tskit.FORWARD">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">FORWARD</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><em class="property"> = 1</em><a class="headerlink" href="#tskit.FORWARD" title="Permalink to this definition">¶</a></dt>
<dd><p>Constant representing the forward direction of travel (i.e.,
increasing genomic coordinate values).</p>
</dd></dl>

<dl class="py data">
<dt id="tskit.REVERSE">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">REVERSE</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><em class="property"> = -1</em><a class="headerlink" href="#tskit.REVERSE" title="Permalink to this definition">¶</a></dt>
<dd><p>Constant representing the reverse direction of travel (i.e.,
decreasing genomic coordinate values).</p>
</dd></dl>

<dl class="py data">
<dt id="tskit.ALLELES_ACGT">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">ALLELES_ACGT</code><em class="property"> = ('A', 'C', 'G', 'T')</em><a class="headerlink" href="#tskit.ALLELES_ACGT" title="Permalink to this definition">¶</a></dt>
<dd><p>Say what</p>
</dd></dl>

</div>
<div class="section" id="simple-container-classes">
<h3>Simple container classes<a class="headerlink" href="#simple-container-classes" title="Permalink to this headline">¶</a></h3>
<p>These classes are simple shallow containers representing the entities defined
in the <a class="reference internal" href="data-model.html#sec-data-model-definitions"><span class="std std-ref">Definitions</span></a>. These classes are not intended to be instantiated
directly, but are the return types for the various iterators provided by the
<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> and <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> classes.</p>
<dl class="py class">
<dt id="tskit.Individual">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Individual</code><a class="reference internal" href="_modules/tskit/trees.html#Individual"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Individual" title="Permalink to this definition">¶</a></dt>
<dd><p>An <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual</span></a> in a tree sequence.
Since nodes correspond to genomes, individuals are associated with a collection
of nodes (e.g., two nodes per diploid). See <a class="reference internal" href="data-model.html#sec-nodes-or-individuals"><span class="std std-ref">Nodes, Genomes, or Individuals?</span></a>
for more discussion of this distinction.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of this individual. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_individuals" title="tskit.TreeSequence.num_individuals"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_individuals</span></code></a> - 1.</p></li>
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The bitwise flags for this individual.</p></li>
<li><p><strong>location</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – The spatial location of this individual as a numpy array. The
location is an empty array if no spatial location is defined.</p></li>
<li><p><strong>nodes</strong> – The IDs of the nodes that are associated with this individual as
a numpy array (dtype=np.int32). If no nodes are associated with the
individual this array will be empty.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – The decoded <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a>
for this individual.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.Node">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Node</code><a class="reference internal" href="_modules/tskit/trees.html#Node"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Node" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node</span></a> in a tree sequence, corresponding
to a single genome. The <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">population</span></code> are attributes of the
<code class="docutils literal notranslate"><span class="pre">Node</span></code>, rather than the <code class="docutils literal notranslate"><span class="pre">Individual</span></code>, as discussed in
<a class="reference internal" href="data-model.html#sec-nodes-or-individuals"><span class="std std-ref">Nodes, Genomes, or Individuals?</span></a>.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of this node. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_nodes" title="tskit.TreeSequence.num_nodes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_nodes</span></code></a> - 1.</p></li>
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The bitwise flags for this node.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The birth time of this node.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of the population that this node was born in.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of the individual that this node was a part of.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – The decoded <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this node.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.Node.is_sample">
<code class="sig-name descname">is_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Node.is_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Node.is_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if this node is a sample. This value is derived from the
<code class="docutils literal notranslate"><span class="pre">flag</span></code> variable.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.Edge">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Edge</code><a class="reference internal" href="_modules/tskit/trees.html#Edge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Edge" title="Permalink to this definition">¶</a></dt>
<dd><p>An <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The left coordinate of this edge.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The right coordinate of this edge.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of the parent node for this edge.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of the child node for this edge.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p></li>
<li><p><strong>id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of this edge. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_edges" title="tskit.TreeSequence.num_edges"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_edges</span></code></a> - 1.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – The decoded <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this edge.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.Edge.span">
<em class="property">property </em><code class="sig-name descname">span</code><a class="headerlink" href="#tskit.Edge.span" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the span of this edge, i.e. the right position minus the left position</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The span of this edge.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.Interval">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Interval</code><a class="reference internal" href="_modules/tskit/trees.html#Interval"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Interval" title="Permalink to this definition">¶</a></dt>
<dd><p>A tuple of 2 numbers, <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code>, defining an interval over the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The left hand end of the interval. By convention this value is included
in the interval.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The right hand end of the iterval. By convention this value is <em>not</em>
included in the interval, i.e. the interval is half-open.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The span of the genome covered by this interval, simply <code class="docutils literal notranslate"><span class="pre">right-left</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.Site">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Site</code><a class="reference internal" href="_modules/tskit/trees.html#Site"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Site" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of this site. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_sites" title="tskit.TreeSequence.num_sites"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_sites</span></code></a> - 1.</p></li>
<li><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The floating point location of this site in genome coordinates.
Ranges from 0 (inclusive) to <a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a>
(exclusive).</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The ancestral state at this site (i.e., the state
inherited by nodes, unless mutations occur).</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – The decoded <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this site.</p></li>
<li><p><strong>mutations</strong> (list[<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a>]) – The list of mutations at this site. Mutations
within a site are returned in the order they are specified in the
underlying <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.Mutation">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Mutation</code><a class="reference internal" href="_modules/tskit/trees.html#Mutation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Mutation" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of this mutation. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_mutations" title="tskit.TreeSequence.num_mutations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_mutations</span></code></a> - 1.</p></li>
<li><p><strong>site</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of the site that this mutation occurs at. To obtain
further information about a site with a given ID use
<a class="reference internal" href="#tskit.TreeSequence.site" title="tskit.TreeSequence.site"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.site()</span></code></a>.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of the first node that inherits this mutation.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The occurrence time of this mutation.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The derived state for this mutation. This is the state
inherited by nodes in the subtree rooted at this mutation’s node, unless
another mutation occurs.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of this mutation’s parent mutation. When multiple
mutations occur at a site along a path in the tree, mutations must
record the mutation that is immediately above them. If the mutation does
not have a parent, this is equal to the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">NULL</span></code></a> (-1).
To obtain further information about a mutation with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.mutation" title="tskit.TreeSequence.mutation"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.mutation()</span></code></a>.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – The decoded <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this
mutation.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.Variant">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Variant</code><a class="reference internal" href="_modules/tskit/trees.html#Variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Variant" title="Permalink to this definition">¶</a></dt>
<dd><p>A variant represents the observed variation among samples
for a given site. A variant consists (a) of a reference to the
<a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> instance in question; (b) the <strong>alleles</strong> that may be
observed at the samples for this site; and (c) the <strong>genotypes</strong>
mapping sample IDs to the observed alleles.</p>
<p>Each element in the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> tuple is a string, representing the
actual observed state for a given sample. The <code class="docutils literal notranslate"><span class="pre">alleles</span></code> tuple is
generated in one of two ways. The first (and default) way is for
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> to generate the encoding on the fly as alleles are encountered
while generating genotypes. In this case, the first element of this
tuple is guaranteed to be the same as the site’s <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> value
and the list of alleles is also guaranteed not to contain any duplicates.
Note that allelic values may be listed that are not referred to by any
samples. For example, if we have a site that is fixed for the derived state
(i.e., we have a mutation over the tree root), all genotypes will be 1, but
the alleles list will be equal to <code class="docutils literal notranslate"><span class="pre">('0',</span> <span class="pre">'1')</span></code>. Other than the
ancestral state being the first allele, the alleles are listed in
no particular order, and the ordering should not be relied upon
(but see the notes on missing data below).</p>
<p>The second way is for the user to define the mapping between
genotype values and allelic state strings using the
<code class="docutils literal notranslate"><span class="pre">alleles</span></code> parameter to the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a> method.
In this case, there is no indication of which allele is the ancestral state,
as the ordering is determined by the user.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> represent the observed allelic states for each sample,
such that <code class="docutils literal notranslate"><span class="pre">var.alleles[var.genotypes[j]]</span></code> gives the string allele
for sample ID <code class="docutils literal notranslate"><span class="pre">j</span></code>. Thus, the elements of the genotypes array are
indexes into the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> list. The genotypes are provided in this
way via a numpy array to enable efficient calculations.</p>
<p>When <a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a> is present at a given
site boolean flag <code class="docutils literal notranslate"><span class="pre">has_missing_data</span></code> will be True, at least one element
of the <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> array will be equal to <code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code>, and the
last element of the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> array will be <code class="docutils literal notranslate"><span class="pre">None</span></code>. Note that in this
case <code class="docutils literal notranslate"><span class="pre">variant.num_alleles</span></code> will <strong>not</strong> be equal to
<code class="docutils literal notranslate"><span class="pre">len(variant.alleles)</span></code>. The rationale for adding <code class="docutils literal notranslate"><span class="pre">None</span></code> to the end of
the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> list is to help code that does not handle missing data
correctly fail early rather than introducing subtle and hard-to-find bugs.
As <code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code> is equal to -1, code that decodes genotypes into
allelic values without taking missing data into account would otherwise
output the last allele in the list rather missing data.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a>) – The site object for this variant.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – A tuple of the allelic values that may be observed at the
samples at the current site. The first element of this tuple is always
the site’s ancestral state.</p></li>
<li><p><strong>genotypes</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – An array of indexes into the list <code class="docutils literal notranslate"><span class="pre">alleles</span></code>, giving the
state of each sample at the current site.</p></li>
<li><p><strong>has_missing_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – True if there is missing data for any of the
samples at the current site.</p></li>
<li><p><strong>num_alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of distinct alleles at this site. Note that
this may be greater than the number of distinct values in the genotypes
array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.Migration">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Migration</code><a class="reference internal" href="_modules/tskit/trees.html#Migration"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Migration" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The left end of the genomic interval covered by this
migration (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The right end of the genomic interval covered by this migration
(exclusive).</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of the node involved in this migration event.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The source population ID.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The destination population ID.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The time at which this migration occured at.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – The decoded <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this
migration.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.Population">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Population</code><a class="reference internal" href="_modules/tskit/trees.html#Population"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Population" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The integer ID of this population. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_populations" title="tskit.TreeSequence.num_populations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_populations</span></code></a> - 1.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – The decoded <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a>
for this population.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.Provenance">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">Provenance</code><a class="reference internal" href="_modules/tskit/trees.html#Provenance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.Provenance" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="loading-data">
<h3>Loading data<a class="headerlink" href="#loading-data" title="Permalink to this headline">¶</a></h3>
<p>There are several methods for loading data into a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>
instance. The simplest and most convenient is the use the <a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load()</span></code></a>
function to load a <a class="reference internal" href="data-model.html#sec-tree-sequence-file-format"><span class="std std-ref">tree sequence file</span></a>. For small
scale data and debugging, it is often convenient to use the
<a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> to read data in the <a class="reference internal" href="data-model.html#sec-text-file-format"><span class="std std-ref">text file format</span></a>. The <a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a> function
efficiently creates a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> object from a set of tables
using the <a class="reference internal" href="#sec-tables-api"><span class="std std-ref">Tables API</span></a>.</p>
<dl class="py function">
<dt id="tskit.load">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">load</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads a tree sequence from the specified file object or path. The file must be in the
<a class="reference internal" href="data-model.html#sec-tree-sequence-file-format"><span class="std std-ref">tree sequence file format</span></a> produced by the
<a class="reference internal" href="#tskit.TreeSequence.dump" title="tskit.TreeSequence.dump"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.dump()</span></code></a> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The file object or path of the <code class="docutils literal notranslate"><span class="pre">.trees</span></code> file containing the
tree sequence we wish to load.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The tree sequence object containing the information
stored in the specified file path.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.load_text">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">load_text</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nodes</span></em>, <em class="sig-param"><span class="n">edges</span></em>, <em class="sig-param"><span class="n">sites</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mutations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">individuals</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">populations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sequence_length</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em>, <em class="sig-param"><span class="n">base64_metadata</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#load_text"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.load_text" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the tree sequence data from the specified file-like objects, and
returns the resulting <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> object. The format
for these files is documented in the <a class="reference internal" href="data-model.html#sec-text-file-format"><span class="std std-ref">Text file formats</span></a> section,
and is produced by the <a class="reference internal" href="#tskit.TreeSequence.dump_text" title="tskit.TreeSequence.dump_text"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.dump_text()</span></code></a> method. Further
properties required for an input tree sequence are described in the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a> section. This method is intended as a
convenient interface for importing external data into tskit; the binary
file format using by <a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.load()</span></code></a> is many times more efficient than
this text format.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">nodes</span></code> and <code class="docutils literal notranslate"><span class="pre">edges</span></code> parameters are mandatory and must be file-like
objects containing text with whitespace delimited columns,  parsable by
<a class="reference internal" href="#tskit.parse_nodes" title="tskit.parse_nodes"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_nodes()</span></code></a> and <a class="reference internal" href="#tskit.parse_edges" title="tskit.parse_edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_edges()</span></code></a>, respectively. <code class="docutils literal notranslate"><span class="pre">sites</span></code>,
<code class="docutils literal notranslate"><span class="pre">mutations</span></code>, <code class="docutils literal notranslate"><span class="pre">individuals</span></code> and <code class="docutils literal notranslate"><span class="pre">populations</span></code> are optional, and must
be parsable by <a class="reference internal" href="#tskit.parse_sites" title="tskit.parse_sites"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_sites()</span></code></a>, <a class="reference internal" href="#tskit.parse_individuals" title="tskit.parse_individuals"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_individuals()</span></code></a>,
<a class="reference internal" href="#tskit.parse_populations" title="tskit.parse_populations"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_populations()</span></code></a>, and <a class="reference internal" href="#tskit.parse_mutations" title="tskit.parse_mutations"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_mutations()</span></code></a>, respectively.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> parameter determines the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a> of the returned tree sequence. If it
is 0 or not specified, the value is taken to be the maximum right
coordinate of the input edges. This parameter is useful in degenerate
situations (such as when there are zero edges), but can usually be ignored.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">strict</span></code> parameter controls the field delimiting algorithm that
is used. If <code class="docutils literal notranslate"><span class="pre">strict</span></code> is True (the default), we require exactly one
tab character separating each field. If <code class="docutils literal notranslate"><span class="pre">strict</span></code> is False, a more relaxed
whitespace delimiting algorithm is used, such that any run of whitespace
is regarded as a field separator. In most situations, <code class="docutils literal notranslate"><span class="pre">strict=False</span></code>
is more convenient, but it can lead to error in certain situations. For
example, if a deletion is encoded in the mutation table this will not
be parseable when <code class="docutils literal notranslate"><span class="pre">strict=False</span></code>.</p>
<p>After parsing the tables, <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> is called to ensure that
the loaded tables satisfy the tree sequence <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">ordering requirements</span></a>. Note that this may result in the
IDs of various entities changing from their positions in the input file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing text describing a
<a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>.</p></li>
<li><p><strong>edges</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing text
describing an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>.</p></li>
<li><p><strong>sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing text describing a
<a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>.</p></li>
<li><p><strong>mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing text
describing a <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing text
describing a <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a>.</p></li>
<li><p><strong>populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing text
describing a <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The sequence length of the returned tree sequence. If
not supplied or zero this will be inferred from the set of edges.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The tree sequence object containing the information
stored in the specified file paths.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="tables-and-table-collections">
<span id="sec-tables-api"></span><h2>Tables and Table Collections<a class="headerlink" href="#tables-and-table-collections" title="Permalink to this headline">¶</a></h2>
<p>The information required to construct a tree sequence is stored in a collection
of <em>tables</em>, each defining a different aspect of the structure of a tree
sequence. These tables are described individually in <a class="reference internal" href="#sec-tables-api-tables"><span class="std std-ref">the next section</span></a>. However, these are interrelated, and so many operations work
on the entire collection of tables, known as a <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code>.
The <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> and <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> classes are
deeply related. A <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code> instance is based on the information
encoded in a <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code>. Tree sequences are <strong>immutable</strong>, and
provide methods for obtaining trees from the sequence. A <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code>
is <strong>mutable</strong>, and does not have any methods for obtaining trees.
The <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> class thus allows dynamic creation and modification of
tree sequences.</p>
<div class="section" id="the-tablecollection-class">
<h3>The <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> class<a class="headerlink" href="#the-tablecollection-class" title="Permalink to this headline">¶</a></h3>
<p>Many of the <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code> methods that return a modified tree sequence
are in fact wrappers around a corresponding <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> method
that modifies a copy of the origin tree sequence’s table collection.</p>
<dl class="py class">
<dt id="tskit.TableCollection">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">TableCollection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_length</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection" title="Permalink to this definition">¶</a></dt>
<dd><p>A collection of mutable tables defining a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-data-model"><span class="std std-ref">Data model</span></a> section for definition on the various tables
and how they together define a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. Arbitrary
data can be stored in a TableCollection, but there are certain
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">requirements</span></a> that must be
satisfied for these tables to be interpreted as a tree sequence.</p>
<p>To obtain an immutable <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance corresponding to the
current state of a <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code>, please use the <a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tree_sequence()</span></code></a>
method.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individuals</strong> (<a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><em>IndividualTable</em></a>) – The individual table.</p></li>
<li><p><strong>nodes</strong> (<a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><em>NodeTable</em></a>) – The node table.</p></li>
<li><p><strong>edges</strong> (<a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><em>EdgeTable</em></a>) – The edge table.</p></li>
<li><p><strong>migrations</strong> (<a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><em>MigrationTable</em></a>) – The migration table.</p></li>
<li><p><strong>sites</strong> (<a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><em>SiteTable</em></a>) – The site table.</p></li>
<li><p><strong>mutations</strong> (<a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><em>MutationTable</em></a>) – The mutation table.</p></li>
<li><p><strong>populations</strong> (<a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><em>PopulationTable</em></a>) – The population table.</p></li>
<li><p><strong>provenances</strong> (<a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><em>ProvenanceTable</em></a>) – The provenance table.</p></li>
<li><p><strong>index</strong> – The edge insertion and removal index.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The sequence length defining the coordinate
space.</p></li>
<li><p><strong>file_uuid</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The UUID for the file this TableCollection is derived
from, or None if not derived from a file.</p></li>
</ul>
</dd>
</dl>
<p><strong>Methods</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.asdict" title="tskit.TableCollection.asdict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">asdict</span></code></a>()</p></td>
<td><p>Returns a dictionary representation of this TableCollection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">build_index</span></code></a>()</p></td>
<td><p>Builds an index on this TableCollection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.clear" title="tskit.TableCollection.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code></a>([clear_provenance, …])</p></td>
<td><p>Remove all rows of the data tables, optionally remove provenance, metadata schemas and ts-level metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.compute_mutation_parents" title="tskit.TableCollection.compute_mutation_parents"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compute_mutation_parents</span></code></a>()</p></td>
<td><p>Modifies the tables in place, computing the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column of the mutation table.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.compute_mutation_times" title="tskit.TableCollection.compute_mutation_times"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compute_mutation_times</span></code></a>()</p></td>
<td><p>Modifies the tables in place, computing valid values for the <code class="docutils literal notranslate"><span class="pre">time</span></code> column of the mutation table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.copy" title="tskit.TableCollection.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Returns a deep copy of this TableCollection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.deduplicate_sites" title="tskit.TableCollection.deduplicate_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deduplicate_sites</span></code></a>()</p></td>
<td><p>Modifies the tables in place, removing entries in the site table with duplicate <code class="docutils literal notranslate"><span class="pre">position</span></code> (and keeping only the <em>first</em> entry for each site), and renumbering the <code class="docutils literal notranslate"><span class="pre">site</span></code> column of the mutation table appropriately.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.delete_intervals" title="tskit.TableCollection.delete_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_intervals</span></code></a>(intervals[, simplify, …])</p></td>
<td><p>Delete all information from this set of tables which lies <em>within</em> the specified list of genomic intervals.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.delete_sites" title="tskit.TableCollection.delete_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">delete_sites</span></code></a>(site_ids[, record_provenance])</p></td>
<td><p>Remove the specified sites entirely from the sites and mutations tables in this collection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.drop_index" title="tskit.TableCollection.drop_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">drop_index</span></code></a>()</p></td>
<td><p>Drops any indexes present on this table collection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.dump" title="tskit.TableCollection.dump"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dump</span></code></a>(file_or_path)</p></td>
<td><p>Writes the table collection to the specified path or file object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.equals" title="tskit.TableCollection.equals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">equals</span></code></a>(other, *[, ignore_metadata, …])</p></td>
<td><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.has_index" title="tskit.TableCollection.has_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_index</span></code></a>()</p></td>
<td><p>Returns True if this TableCollection is indexed.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.keep_intervals" title="tskit.TableCollection.keep_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">keep_intervals</span></code></a>(intervals[, simplify, …])</p></td>
<td><p>Delete all information from this set of tables which lies <em>outside</em> the specified list of genomic intervals.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.link_ancestors" title="tskit.TableCollection.link_ancestors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">link_ancestors</span></code></a>(samples, ancestors)</p></td>
<td><p>Returns an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance describing a subset of the genealogical relationships between the nodes in <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.ltrim" title="tskit.TableCollection.ltrim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ltrim</span></code></a>([record_provenance])</p></td>
<td><p>Reset the coordinate system used in these tables, changing the left and right genomic positions in the edge table such that the leftmost edge now starts at position 0.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.rtrim" title="tskit.TableCollection.rtrim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rtrim</span></code></a>([record_provenance])</p></td>
<td><p>Reset the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property so that the sequence ends at the end of the last edge.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.simplify" title="tskit.TableCollection.simplify"><code class="xref py py-obj docutils literal notranslate"><span class="pre">simplify</span></code></a>([samples, reduce_to_site_topology, …])</p></td>
<td><p>Simplifies the tables in place to retain only the information necessary to reconstruct the tree sequence describing the given <code class="docutils literal notranslate"><span class="pre">samples</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code></a>([edge_start])</p></td>
<td><p>Sorts the tables in place.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.subset" title="tskit.TableCollection.subset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subset</span></code></a>(nodes[, record_provenance])</p></td>
<td><p>Modifies the tables in place to contain only the entries referring to the provided list of nodes, with nodes reordered according to the order they appear in the list.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>()</p></td>
<td><p>Returns a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance with the structure defined by the tables in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.trim" title="tskit.TableCollection.trim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim</span></code></a>([record_provenance])</p></td>
<td><p>Trim away any empty regions on the right and left of the tree sequence encoded by these tables.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.union" title="tskit.TableCollection.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></a>(other, node_mapping[, …])</p></td>
<td><p>Modifies the table collection in place by adding the non-shared portions of <code class="docutils literal notranslate"><span class="pre">other</span></code> to itself.</p></td>
</tr>
</tbody>
</table>
<p><strong>Attributes</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.metadata" title="tskit.TableCollection.metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata</span></code></a></p></td>
<td><p>The decoded metadata for this TableCollection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.metadata_bytes" title="tskit.TableCollection.metadata_bytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata_bytes</span></code></a></p></td>
<td><p>The raw bytes of metadata for this TableCollection</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.metadata_schema" title="tskit.TableCollection.metadata_schema"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata_schema</span></code></a></p></td>
<td><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this TableCollection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.name_map" title="tskit.TableCollection.name_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">name_map</span></code></a></p></td>
<td><p>Returns a dictionary mapping table names to the corresponding table instances.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.nbytes" title="tskit.TableCollection.nbytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nbytes</span></code></a></p></td>
<td><p>Returns the total number of bytes required to store the data in this table collection.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="tskit.TableCollection.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.asdict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary representation of this TableCollection.</p>
<p>Note: the semantics of this method changed at tskit 0.1.0. Previously a
map of table names to the tables themselves was returned.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.build_index">
<code class="sig-name descname">build_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.build_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.build_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds an index on this TableCollection. Any existing indexes are automatically
dropped.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">clear_provenance</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">clear_metadata_schemas</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">clear_ts_metadata_and_schema</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.clear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove all rows of the data tables, optionally remove provenance, metadata
schemas and ts-level metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>clear_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, remove all rows of the provenance
table. (Default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
<li><p><strong>clear_metadata_schemas</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, clear the table metadata
schemas. (Default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
<li><p><strong>clear_ts_metadata_and_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, clear the tree-sequence
level metadata and schema (Default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.compute_mutation_parents">
<code class="sig-name descname">compute_mutation_parents</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.compute_mutation_parents"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.compute_mutation_parents" title="Permalink to this definition">¶</a></dt>
<dd><p>Modifies the tables in place, computing the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column of the
mutation table. For this to work, the node and edge tables must be
valid, and the site and mutation tables must be sorted (see
<a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a>).  This will produce an error if mutations
are not sorted (i.e., if a mutation appears before its mutation parent)
<em>unless</em> the two mutations occur on the same branch, in which case
there is no way to detect the error.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">parent</span></code> of a given mutation is the ID of the next mutation
encountered traversing the tree upwards from that mutation, or
<code class="docutils literal notranslate"><span class="pre">NULL</span></code> if there is no such mutation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>note: This method does not check that all mutations result
in a change of state, as required; see <a class="reference internal" href="data-model.html#sec-mutation-requirements"><span class="std std-ref">Mutation requirements</span></a>.</p>
</div>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.compute_mutation_times">
<code class="sig-name descname">compute_mutation_times</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.compute_mutation_times"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.compute_mutation_times" title="Permalink to this definition">¶</a></dt>
<dd><p>Modifies the tables in place, computing valid values for the <code class="docutils literal notranslate"><span class="pre">time</span></code> column of
the mutation table. For this to work, the node and edge tables must be
valid, and the site and mutation tables must be sorted and indexed(see
<a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> and <a class="reference internal" href="#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.build_index()</span></code></a>).</p>
<p>For a single mutation on an edge at a site, the <code class="docutils literal notranslate"><span class="pre">time</span></code> assigned to a mutation
by this method is the mid-point between the times of the nodes above and below
the mutation. In the case where there is more than one mutation on an edge for
a site, the times are evenly spread along the edge. For mutations that are
above a root node, the time of the root node is assigned.</p>
<p>The mutation table will be sorted if the new times mean that the original order
is no longer valid.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.copy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this TableCollection.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A deep copy of this TableCollection.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection">TableCollection</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.deduplicate_sites">
<code class="sig-name descname">deduplicate_sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.deduplicate_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.deduplicate_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Modifies the tables in place, removing entries in the site table with
duplicate <code class="docutils literal notranslate"><span class="pre">position</span></code> (and keeping only the <em>first</em> entry for each
site), and renumbering the <code class="docutils literal notranslate"><span class="pre">site</span></code> column of the mutation table
appropriately.  This requires the site table to be sorted by position.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.delete_intervals">
<code class="sig-name descname">delete_intervals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">intervals</span></em>, <em class="sig-param"><span class="n">simplify</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.delete_intervals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.delete_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete all information from this set of tables which lies <em>within</em> the
specified list of genomic intervals. This is identical to
<a class="reference internal" href="#tskit.TreeSequence.delete_intervals" title="tskit.TreeSequence.delete_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.delete_intervals()</span></code></a> but acts <em>in place</em> to alter
the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>) – A list (start, end) pairs describing the
genomic intervals to delete. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, run simplify on the tables so that nodes
no longer used are discarded. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.delete_sites">
<code class="sig-name descname">delete_sites</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site_ids</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.delete_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.delete_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the specified sites entirely from the sites and mutations tables in this
collection. This is identical to <a class="reference internal" href="#tskit.TreeSequence.delete_sites" title="tskit.TreeSequence.delete_sites"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.delete_sites()</span></code></a> but acts
<em>in place</em> to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>]</em>) – A list of site IDs specifying the sites to remove.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.drop_index">
<code class="sig-name descname">drop_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.drop_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.drop_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Drops any indexes present on this table collection. If the tables are not
currently indexed this method has no effect.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.dump">
<code class="sig-name descname">dump</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_or_path</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.dump"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.dump" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the table collection to the specified path or file object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_or_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The file object or path to write the TreeSequence to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_ts_metadata</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_provenance</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_timestamps</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.equals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two table
collections are considered equal if their</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> properties are identical;</p></li>
<li><p>top-level tree sequence metadata and metadata schemas are
byte-wise identical;</p></li>
<li><p>constituent tables are byte-wise identical.</p></li>
</ul>
<p>Some of the requirements in this definition can be relaxed using the
parameters, which can be used to remove certain parts of the data model
from the comparison.</p>
<p>Table indexes are not considered in the equality comparison.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><em>TableCollection</em></a>) – Another table collection.</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True <em>all</em> metadata and metadata schemas
will be excluded from the comparison. This includes the top-level
tree sequence and constituent table metadata (default=False).</p></li>
<li><p><strong>ignore_ts_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True the top-level tree sequence
metadata and metadata schemas will be excluded from the comparison.
If <code class="docutils literal notranslate"><span class="pre">ignore_metadata</span></code> is True, this parameter has no effect.</p></li>
<li><p><strong>ignore_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True the provenance tables are
not included in the comparison.</p></li>
<li><p><strong>ignore_timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True the provenance timestamp column
is ignored in the comparison. If <code class="docutils literal notranslate"><span class="pre">ignore_provenance</span></code> is True, this
parameter has no effect.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table collection; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.has_index">
<code class="sig-name descname">has_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.has_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.has_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if this TableCollection is indexed.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.keep_intervals">
<code class="sig-name descname">keep_intervals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">intervals</span></em>, <em class="sig-param"><span class="n">simplify</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.keep_intervals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.keep_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete all information from this set of tables which lies <em>outside</em> the
specified list of genomic intervals. This is identical to
<a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.keep_intervals()</span></code></a> but acts <em>in place</em> to alter
the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>) – A list (start, end) pairs describing the
genomic intervals to keep. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, run simplify on the tables so that nodes
no longer used are discarded. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.link_ancestors">
<code class="sig-name descname">link_ancestors</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span></em>, <em class="sig-param"><span class="n">ancestors</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.link_ancestors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.link_ancestors" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance describing a subset of the genealogical
relationships between the nodes in <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code>.</p>
<p>Each row <code class="docutils literal notranslate"><span class="pre">parent,</span> <span class="pre">child,</span> <span class="pre">left,</span> <span class="pre">right</span></code> in the output table indicates that
<code class="docutils literal notranslate"><span class="pre">child</span></code> has inherited the segment <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code> from <code class="docutils literal notranslate"><span class="pre">parent</span></code> more
recently than from any other node in these lists.</p>
<p>In particular, suppose <code class="docutils literal notranslate"><span class="pre">samples</span></code> is a list of nodes such that <code class="docutils literal notranslate"><span class="pre">time</span></code> is 0
for each node, and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> is a list of nodes such that <code class="docutils literal notranslate"><span class="pre">time</span></code> is
greater than 0.0 for each node. Then each row of the output table will show
an interval <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code> over which a node in <code class="docutils literal notranslate"><span class="pre">samples</span></code> has inherited
most recently from a node in <code class="docutils literal notranslate"><span class="pre">ancestors</span></code>, or an interval over which one of
these <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> has inherited most recently from another node in
<code class="docutils literal notranslate"><span class="pre">ancestors</span></code>.</p>
<p>The following table shows which <code class="docutils literal notranslate"><span class="pre">parent-&gt;child</span></code> pairs will be shown in the
output of <code class="docutils literal notranslate"><span class="pre">link_ancestors</span></code>.
A node is a relevant descendant on a given interval if it also appears somewhere
in the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column of the outputted table.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 34%" />
<col style="width: 66%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Type of relationship</p></td>
<td><p>Shown in output of <code class="docutils literal notranslate"><span class="pre">link_ancestors</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ancestor-&gt;sample</span></code></p></td>
<td><p>Always</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">ancestor1-&gt;ancestor2</span></code></p></td>
<td><p>Only if <code class="docutils literal notranslate"><span class="pre">ancestor2</span></code> has a relevant descendant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">sample1-&gt;sample2</span></code></p></td>
<td><p>Always</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sample-&gt;ancestor</span></code></p></td>
<td><p>Only if <code class="docutils literal notranslate"><span class="pre">ancestor</span></code> has a relevant descendant</p></td>
</tr>
</tbody>
</table>
<p>The difference between <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> is that information about
the ancestors of a node in <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> will only be retained if it also has a
relevant descendant, while information about the ancestors of a node in
<code class="docutils literal notranslate"><span class="pre">samples</span></code> will always be retained.
The node IDs in <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code> refer to the IDs in the node table
of the inputted tree sequence.</p>
<p>The supplied nodes must be non-empty lists of the node IDs in the tree sequence:
in particular, they do not have to be <em>samples</em> of the tree sequence. The lists
of <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> may overlap, although adding a node from
<code class="docutils literal notranslate"><span class="pre">samples</span></code> to <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> will not change the output. So, setting <code class="docutils literal notranslate"><span class="pre">samples</span></code>
and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> to the same list of nodes will find all genealogical
relationships within this list.</p>
<p>If none of the nodes in <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> or <code class="docutils literal notranslate"><span class="pre">samples</span></code> are ancestral to <code class="docutils literal notranslate"><span class="pre">samples</span></code>
anywhere in the tree sequence, an empty table will be returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>]</em>) – A list of node IDs to retain as samples.</p></li>
<li><p><strong>ancestors</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>]</em>) – A list of node IDs to use as ancestors.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance displaying relationships between
the <cite>samples</cite> and <cite>ancestors</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.ltrim">
<code class="sig-name descname">ltrim</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.ltrim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.ltrim" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset the coordinate system used in these tables, changing the left and right
genomic positions in the edge table such that the leftmost edge now starts at
position 0. This is identical to <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ltrim()</span></code></a> but acts <em>in place</em>
to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.metadata">
<em class="property">property </em><code class="sig-name descname">metadata</code><a class="headerlink" href="#tskit.TableCollection.metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>The decoded metadata for this TableCollection.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.metadata_bytes">
<em class="property">property </em><code class="sig-name descname">metadata_bytes</code><a class="headerlink" href="#tskit.TableCollection.metadata_bytes" title="Permalink to this definition">¶</a></dt>
<dd><p>The raw bytes of metadata for this TableCollection</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.TableCollection.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this TableCollection.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.name_map">
<em class="property">property </em><code class="sig-name descname">name_map</code><a class="headerlink" href="#tskit.TableCollection.name_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping table names to the corresponding
table instances. For example, the returned dictionary will contain the
key “edges” that maps to an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.TableCollection.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table collection. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.rtrim">
<code class="sig-name descname">rtrim</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.rtrim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.rtrim" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property so that the sequence ends at the end of
the last edge. This is identical to <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.rtrim()</span></code></a> but acts
<em>in place</em> to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.simplify">
<code class="sig-name descname">simplify</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">reduce_to_site_topology</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">filter_populations</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_individuals</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_sites</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">keep_unary</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">keep_input_roots</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">filter_zero_mutation_sites</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.simplify"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.simplify" title="Permalink to this definition">¶</a></dt>
<dd><p>Simplifies the tables in place to retain only the information necessary
to reconstruct the tree sequence describing the given <code class="docutils literal notranslate"><span class="pre">samples</span></code>.
This will change the ID of the nodes, so that the node
<code class="docutils literal notranslate"><span class="pre">samples[k]</span></code> will have ID <code class="docutils literal notranslate"><span class="pre">k</span></code> in the result. The resulting
NodeTable will have only the first <code class="docutils literal notranslate"><span class="pre">len(samples)</span></code> individuals marked
as samples. The mapping from node IDs in the current set of tables to
their equivalent values in the simplified tables is also returned as a
numpy array. If an array <code class="docutils literal notranslate"><span class="pre">a</span></code> is returned by this function and <code class="docutils literal notranslate"><span class="pre">u</span></code>
is the ID of a node in the input table, then <code class="docutils literal notranslate"><span class="pre">a[u]</span></code> is the ID of this
node in the output table. For any node <code class="docutils literal notranslate"><span class="pre">u</span></code> that is not mapped into
the output tables, this mapping will equal <code class="docutils literal notranslate"><span class="pre">-1</span></code>.</p>
<p>Tables operated on by this function must: be sorted (see
<a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a>), have children be born strictly after their
parents, and the intervals on which any individual is a child must be
disjoint. Other than this the tables need not satisfy remaining
requirements to specify a valid tree sequence (but the resulting tables
will).</p>
<p>This is identical to <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code></a> but acts <em>in place</em> to
alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>. Please see the
<a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code></a> method for a description of the remaining
parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>]</em>) – A list of node IDs to retain as samples. If
not specified or None, use all nodes marked with the IS_SAMPLE flag.</p></li>
<li><p><strong>reduce_to_site_topology</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to reduce the topology down
to the trees that are present at sites. (default: False).</p></li>
<li><p><strong>filter_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, remove any populations that are
not referenced by nodes after simplification; new population IDs are
allocated sequentially from zero. If False, the population table will
not be altered in any way. (Default: True)</p></li>
<li><p><strong>filter_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, remove any individuals that are
not referenced by nodes after simplification; new individual IDs are
allocated sequentially from zero. If False, the individual table will
not be altered in any way. (Default: True)</p></li>
<li><p><strong>filter_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, remove any sites that are
not referenced by mutations after simplification; new site IDs are
allocated sequentially from zero. If False, the site table will not
be altered in any way. (Default: True)</p></li>
<li><p><strong>keep_unary</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, any unary nodes (i.e. nodes with exactly
one child) that exist on the path from samples to root will be preserved
in the output. (Default: False)</p></li>
<li><p><strong>keep_input_roots</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, insert edges from the MRCAs of the
samples to the roots in the input trees. If False, no topology older
than the MRCAs of the samples will be included. (Default: False)</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, record details of this call to
simplify in the returned tree sequence’s provenance information
(Default: True).</p></li>
<li><p><strong>filter_zero_mutation_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Deprecated alias for <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A numpy array mapping node IDs in the input tables to their
corresponding node IDs in the output tables.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.sort">
<code class="sig-name descname">sort</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">edge_start</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.sort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.sort" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorts the tables in place. This ensures that all tree sequence ordering
requirements listed in the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a> section are met, as long
as each site has at most one mutation (see below).</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">edge_start</span></code> parameter is provided, this specifies the index
in the edge table where sorting should start. Only rows with index
greater than or equal to <code class="docutils literal notranslate"><span class="pre">edge_start</span></code> are sorted; rows before this index
are not affected. This parameter is provided to allow for efficient sorting
when the user knows that the edges up to a given index are already sorted.</p>
<p>The individual, node, population and provenance tables are not affected
by this method.</p>
<p>Edges are sorted as follows:</p>
<ul class="simple">
<li><p>time of parent, then</p></li>
<li><p>parent node ID, then</p></li>
<li><p>child node ID, then</p></li>
<li><p>left endpoint.</p></li>
</ul>
<p>Note that this sorting order exceeds the
<a class="reference internal" href="data-model.html#sec-edge-requirements"><span class="std std-ref">edge sorting requirements</span></a> for a valid
tree sequence. For a valid tree sequence, we require that all edges for a
given parent ID are adjacent, but we do not require that they be listed in
sorted order.</p>
<p>Sites are sorted by position, and sites with the same position retain
their relative ordering.</p>
<p>Mutations are sorted by site ID, and within the same site are sorted by time.
Those with equal or unknown time retain their relative ordering. This does not
currently rearrange tables so that mutations occur after their mutation parents,
which is a requirement for valid tree sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge_start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The index in the edge table where sorting starts
(default=0; must be &lt;= len(edges)).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.subset">
<code class="sig-name descname">subset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nodes</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.subset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Modifies the tables in place to contain only the entries referring to
the provided list of nodes, with nodes reordered according to the order
they appear in the list. See <a class="reference internal" href="#tskit.TreeSequence.subset" title="tskit.TreeSequence.subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.subset()</span></code></a> for a more
detailed description.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The list of nodes for which to retain information. This
may be a numpy array (or array-like) object (dtype=np.int32).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to record a provenance entry
in the provenance table for this operation.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.tree_sequence">
<code class="sig-name descname">tree_sequence</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.tree_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.tree_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance with the structure defined by the
tables in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>. If the table collection is not
in canonical form (i.e., does not meet sorting requirements) or cannot be
interpreted as a tree sequence an exception is raised. The
<a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort()</span></code></a> method may be used to ensure that input sorting requirements
are met. If the table collection does not have indexes they will be
built.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance reflecting the structures
defined in this set of tables.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.trim">
<code class="sig-name descname">trim</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.trim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.trim" title="Permalink to this definition">¶</a></dt>
<dd><p>Trim away any empty regions on the right and left of the tree sequence encoded by
these tables. This is identical to <a class="reference internal" href="#tskit.TreeSequence.trim" title="tskit.TreeSequence.trim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trim()</span></code></a> but acts <em>in place</em>
to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.TableCollection.union">
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">node_mapping</span></em>, <em class="sig-param"><span class="n">check_shared_equality</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">add_populations</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">record_provenance</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.union"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TableCollection.union" title="Permalink to this definition">¶</a></dt>
<dd><p>Modifies the table collection in place by adding the non-shared
portions of <code class="docutils literal notranslate"><span class="pre">other</span></code> to itself. To perform the node-wise union,
the method relies on a <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code> array, that maps nodes in
<code class="docutils literal notranslate"><span class="pre">other</span></code> to its equivalent node in <code class="docutils literal notranslate"><span class="pre">self</span></code> or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if
the node is exclusive to <code class="docutils literal notranslate"><span class="pre">other</span></code>. See <a class="reference internal" href="#tskit.TreeSequence.union" title="tskit.TreeSequence.union"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.union()</span></code></a> for a more
detailed description.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><em>TableCollection</em></a>) – Another table collection.</p></li>
<li><p><strong>node_mapping</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – An array of node IDs that relate nodes in
<code class="docutils literal notranslate"><span class="pre">other</span></code> to nodes in <code class="docutils literal notranslate"><span class="pre">self</span></code>: the k-th element of <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code>
should be the index of the equivalent node in <code class="docutils literal notranslate"><span class="pre">self</span></code>, or
<code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if the node is not present in <code class="docutils literal notranslate"><span class="pre">self</span></code> (in which case it
will be added to self).</p></li>
<li><p><strong>check_shared_equality</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, the shared portions of the
table collections will be checked for equality.</p></li>
<li><p><strong>add_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, nodes new to <code class="docutils literal notranslate"><span class="pre">self</span></code> will be
assigned new population IDs.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to record a provenance entry
in the provenance table for this operation.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="tables">
<span id="sec-tables-api-tables"></span><h3>Tables<a class="headerlink" href="#tables" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="data-model.html#sec-binary-interchange"><span class="std std-ref">tables API</span></a> provides an efficient way of working
with and interchanging <a class="reference internal" href="data-model.html#sec-data-model"><span class="std std-ref">tree sequence data</span></a>. Each table
class (e.g, <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>, <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>) has a specific set of
columns with fixed types, and a set of methods for setting and getting the data
in these columns. The number of rows in the table <code class="docutils literal notranslate"><span class="pre">t</span></code> is given by <code class="docutils literal notranslate"><span class="pre">len(t)</span></code>.
Each table supports accessing the data either by row or column. To access the
row <code class="docutils literal notranslate"><span class="pre">j</span></code> in table <code class="docutils literal notranslate"><span class="pre">t</span></code> simply use <code class="docutils literal notranslate"><span class="pre">t[j]</span></code>. The value returned by such an
access is an instance of <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.namedtuple" title="(in Python v3.9)"><code class="xref py py-func docutils literal notranslate"><span class="pre">collections.namedtuple()</span></code></a>, and therefore supports
either positional or named attribute access. To access the data in
a column, we can use standard attribute access which will return a numpy array
of the data. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tskit</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">EdgeTable</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">child</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">child</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="go">id      left            right           parent  child</span>
<span class="go">0       0.00000000      1.00000000      10      11</span>
<span class="go">1       1.00000000      2.00000000      9       11</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">EdgeTableRow(left=0.0, right=1.0, parent=10, child=11)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">EdgeTableRow(left=1.0, right=2.0, parent=9, child=11)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">left</span>
<span class="go">array([ 0.,  1.])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">parent</span>
<span class="go">array([10,  9], dtype=int32)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="go">2</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p>Tables also support the <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pickle</span></code></a> protocol, and so can be easily
serialised and deserialised (for example, when performing parallel
computations using the <a class="reference external" href="https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">multiprocessing</span></code></a> module).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">serialised</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t2</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">serialised</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t2</span><span class="p">)</span>
<span class="go">id      left            right           parent  child</span>
<span class="go">0       0.00000000      1.00000000      10      11</span>
<span class="go">1       1.00000000      2.00000000      9       11</span>
</pre></div>
</div>
<p>However, pickling will not be as efficient as storing tables
in the native <a class="reference internal" href="data-model.html#sec-tree-sequence-file-format"><span class="std std-ref">format</span></a>.</p>
<p>Tables support the equality operator <code class="docutils literal notranslate"><span class="pre">==</span></code> based on the data
held in the columns:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">==</span> <span class="n">t2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="ow">is</span> <span class="n">t2</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t2</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t2</span><span class="p">)</span>
<span class="go">id      left            right           parent  child</span>
<span class="go">0       0.00000000      1.00000000      10      11</span>
<span class="go">1       1.00000000      2.00000000      9       11</span>
<span class="go">2       0.00000000      1.00000000      2       3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">==</span> <span class="n">t2</span>
<span class="go">False</span>
</pre></div>
</div>
</div>
<div class="section" id="text-columns">
<span id="sec-tables-api-text-columns"></span><h3>Text columns<a class="headerlink" href="#text-columns" title="Permalink to this headline">¶</a></h3>
<p>As described in the <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>, working with
variable length columns is somewhat more involved. Columns
encoding text data store the <strong>encoded bytes</strong> of the flattened
strings, and the offsets into this column in two separate
arrays.</p>
<p>Consider the following example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">SiteTable</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;BB&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;CCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="go">id      position        ancestral_state metadata</span>
<span class="go">0       0.00000000      A</span>
<span class="go">1       1.00000000      BB</span>
<span class="go">2       2.00000000</span>
<span class="go">3       3.00000000      CCC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">SiteTableRow(position=0.0, ancestral_state=&#39;A&#39;, metadata=b&#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">SiteTableRow(position=1.0, ancestral_state=&#39;BB&#39;, metadata=b&#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="go">SiteTableRow(position=2.0, ancestral_state=&#39;&#39;, metadata=b&#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="go">SiteTableRow(position=3.0, ancestral_state=&#39;CCC&#39;, metadata=b&#39;&#39;)</span>
</pre></div>
</div>
<p>Here we create a <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> and add four rows, each with a different
<code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code>. We can then access this information from each
row in a straightforward manner. Working with the data in the columns
is a little trickier, however:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">ancestral_state</span>
<span class="go">array([65, 66, 66, 67, 67, 67], dtype=int8)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">ancestral_state_offset</span>
<span class="go">array([0, 1, 3, 3, 6], dtype=uint32)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tskit</span><span class="o">.</span><span class="n">unpack_strings</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">ancestral_state</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">ancestral_state_offset</span><span class="p">)</span>
<span class="go">[&#39;A&#39;, &#39;BB&#39;, &#39;&#39;, &#39;CCC&#39;]</span>
</pre></div>
</div>
<p>Here, the <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> array is the UTF8 encoded bytes of the flattened
strings, and the <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> is the offset into this array
for each row. The <a class="reference internal" href="#tskit.unpack_strings" title="tskit.unpack_strings"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.unpack_strings()</span></code></a> function, however, is a convient
way to recover the original strings from this encoding. We can also use the
<a class="reference internal" href="#tskit.pack_strings" title="tskit.pack_strings"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.pack_strings()</span></code></a> to insert data using this approach:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">,</span> <span class="n">off</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">pack_strings</span><span class="p">([</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="s2">&quot;12&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">ancestral_state_offset</span><span class="o">=</span><span class="n">off</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="go">id      position        ancestral_state metadata</span>
<span class="go">0       0.00000000      0</span>
<span class="go">1       1.00000000      12</span>
<span class="go">2       2.00000000</span>
</pre></div>
</div>
<p>When inserting many rows with standard infinite sites mutations (i.e.,
ancestral state is “0”), it is more efficient to construct the
numpy arrays directly than to create a list of strings and use
<a class="reference internal" href="#tskit.pack_strings" title="tskit.pack_strings"><code class="xref py py-func docutils literal notranslate"><span class="pre">pack_strings()</span></code></a>. When doing this, it is important to note that
it is the <strong>encoded</strong> byte values that are stored; by default, we
use UTF8 (which corresponds to ASCII for simple printable characters).:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t_s</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">SiteTable</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">off</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t_s</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="p">),</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">ancestral_state_offset</span><span class="o">=</span><span class="n">off</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t_s</span><span class="p">)</span>
<span class="go">id      position        ancestral_state metadata</span>
<span class="go">0       0.00000000      0</span>
<span class="go">1       1.00000000      0</span>
<span class="go">2       2.00000000      0</span>
<span class="go">3       3.00000000      0</span>
<span class="go">4       4.00000000      0</span>
<span class="go">5       5.00000000      0</span>
<span class="go">6       6.00000000      0</span>
<span class="go">7       7.00000000      0</span>
<span class="go">8       8.00000000      0</span>
<span class="go">9       9.00000000      0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t_s</span><span class="o">.</span><span class="n">ancestral_state</span>
<span class="go">array([48, 48, 48, 48, 48, 48, 48, 48, 48, 48], dtype=int8)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t_s</span><span class="o">.</span><span class="n">ancestral_state_offset</span>
<span class="go">array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10], dtype=uint32)</span>
</pre></div>
</div>
<p>Here we create 10 sites at regular positions, each with ancestral state equal to
“0”. Note that we use <code class="docutils literal notranslate"><span class="pre">ord(&quot;0&quot;)</span></code> to get the ASCII code for “0” (48), and create
10 copies of this by adding it to an array of zeros. We have done this for
illustration purposes: it is equivalent (though slower for large examples) to do
<code class="docutils literal notranslate"><span class="pre">a,</span> <span class="pre">off</span> <span class="pre">=</span> <span class="pre">tskit.pack_strings([&quot;0&quot;]</span> <span class="pre">*</span> <span class="pre">m)</span></code>.</p>
<p>Mutations can be handled similarly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t_m</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MutationTable</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">site</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">,</span> <span class="n">off</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">pack_strings</span><span class="p">([</span><span class="s2">&quot;1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">m</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">node</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t_m</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">site</span><span class="o">=</span><span class="n">site</span><span class="p">,</span> <span class="n">node</span><span class="o">=</span><span class="n">node</span><span class="p">,</span> <span class="n">derived_state</span><span class="o">=</span><span class="n">d</span><span class="p">,</span> <span class="n">derived_state_offset</span><span class="o">=</span><span class="n">off</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t_m</span><span class="p">)</span>
<span class="go">id      site    node    derived_state   parent  metadata</span>
<span class="go">0       0       0       1       -1</span>
<span class="go">1       1       0       1       -1</span>
<span class="go">2       2       0       1       -1</span>
<span class="go">3       3       0       1       -1</span>
<span class="go">4       4       0       1       -1</span>
<span class="go">5       5       0       1       -1</span>
<span class="go">6       6       0       1       -1</span>
<span class="go">7       7       0       1       -1</span>
<span class="go">8       8       0       1       -1</span>
<span class="go">9       9       0       1       -1</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</div>
<div class="section" id="binary-columns">
<span id="sec-tables-api-binary-columns"></span><h3>Binary columns<a class="headerlink" href="#binary-columns" title="Permalink to this headline">¶</a></h3>
<p>Columns storing binary data take the same approach as
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> to encoding
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">variable length data</span></a>.
The difference between the two is only raw <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bytes</span></code></a> values are accepted: no
character encoding or decoding is done on the data. Consider the following example
where a table has no <code class="docutils literal notranslate"><span class="pre">metadata_schema</span></code> such that arbitrary bytes can be stored and
no automatic encoding or decoding of objects is performed by the Python API and we can
store and retrive raw <code class="docutils literal notranslate"><span class="pre">bytes</span></code>. (See <a class="reference internal" href="metadata.html#sec-metadata"><span class="std std-ref">Metadata</span></a> for details):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NodeTable</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="sa">b</span><span class="s2">&quot;raw bytes&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">({</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="mf">1.1</span><span class="p">}))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">b&#39;raw bytes&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">b&#39;\x80\x03}q\x00X\x01\x00\x00\x00xq\x01G?\xf1\x99\x99\x99\x99\x99\x9as.&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>
<span class="go">{&#39;x&#39;: 1.1}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="go">id      flags   population      time    metadata</span>
<span class="go">0       0       -1      0.00000000000000        cmF3IGJ5dGVz</span>
<span class="go">1       0       -1      0.00000000000000        gAN9cQBYAQAAAHhxAUc/8ZmZmZmZmnMu</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">array([ 114,   97,  119,   32,   98,  121,  116,  101,  115, -128,    3,</span>
<span class="go">        125,  113,    0,   88,    1,    0,    0,    0,  120,  113,    1,</span>
<span class="go">         71,   63,  -15, -103, -103, -103, -103, -103, -102,  115,   46], dtype=int8)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">.</span><span class="n">metadata_offset</span>
<span class="go">array([ 0,  9, 33], dtype=uint32)</span>
</pre></div>
</div>
<p>Here we add two rows to a <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>, with different
<a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a>. The first row contains a simple
byte string, and the second contains a Python dictionary serialised using
<a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pickle</span></code></a>. We then show several different (and seemingly incompatible!)
different views on the same data.</p>
<p>When we access the data in a row (e.g., <code class="docutils literal notranslate"><span class="pre">t[0].metadata</span></code>) we are returned
a Python bytes object containing precisely the bytes that were inserted.
The pickled dictionary is encoded in 24 bytes containing unprintable
characters, and when we unpickle it using <a class="reference external" href="https://docs.python.org/3/library/pickle.html#pickle.loads" title="(in Python v3.9)"><code class="xref py py-func docutils literal notranslate"><span class="pre">pickle.loads()</span></code></a>, we obtain
the original dictionary.</p>
<p>When we print the table, however, we see some data which is seemingly
unrelated to the original contents. This is because the binary data is
<a class="reference external" href="https://en.wikipedia.org/wiki/Base64">base64 encoded</a> to ensure
that it is print-safe (and doesn’t break your terminal). (See the
<a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for more information on the
use of base64 encoding.).</p>
<p>Finally, when we print the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column, we see the raw byte values
encoded as signed integers. As for <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a>,
the <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> column encodes the offsets into this array. So, we
see that the first metadata value is 9 bytes long and the second is 24.</p>
<p>The <a class="reference internal" href="#tskit.pack_bytes" title="tskit.pack_bytes"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.pack_bytes()</span></code></a> and <a class="reference internal" href="#tskit.unpack_bytes" title="tskit.unpack_bytes"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.unpack_bytes()</span></code></a> functions are
also useful for encoding data in these columns.</p>
</div>
<div class="section" id="table-classes">
<h3>Table classes<a class="headerlink" href="#table-classes" title="Permalink to this headline">¶</a></h3>
<p>This section describes the methods and variables available for each
table class. For description and definition of each table’s meaning
and use, see <a class="reference internal" href="data-model.html#sec-table-definitions"><span class="std std-ref">the table definitions</span></a>.</p>
<dl class="py class">
<dt id="tskit.IndividualTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">IndividualTable</code><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.IndividualTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table defining the individuals in a tree sequence. Note that although
each Individual has associated nodes, reference to these is not stored in
the individual table, but rather reference to the individual is stored for
each node in the <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>.  This is similar to the way in which
the relationship between sites and mutations is modelled.</p>
<dl class="field-list">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>The numpy arrays returned by table attribute accesses are <strong>copies</strong>
of the underlying data. In particular, this means that you cannot edit
the values in the columns by updating the attribute arrays.</p>
<p><strong>NOTE:</strong> this behaviour may change in future.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of flags values.</p></li>
<li><p><strong>location</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The flattened array of floating point location values. See
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for more details.</p></li>
<li><p><strong>location_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the location column. See
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for more details.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>) – The metadata schema for this table’s metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.IndividualTable.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the specifed row of this table, decoding metadata if it is present.
Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the zero-index of the desired row</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flags</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">location</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.IndividualTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> and returns the ID of the
corresponding individual. Metadata, if specified, will be validated and encoded
according to the table’s
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The bitwise flags for the new node.</p></li>
<li><p><strong>location</strong> (<em>array-like</em>) – A list of numeric values or one-dimensional numpy
array describing the location of this individual. If not specified
or None, a zero-dimensional location is stored.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – Any object that is valid metadata for the table’s schema.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the newly added node.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flags</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">location</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">location_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.IndividualTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns in this
<a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code> array is mandatory and defines the number of
extra individuals to add to the table.
The <code class="docutils literal notranslate"><span class="pre">location</span></code> and <code class="docutils literal notranslate"><span class="pre">location_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The bitwise flags for each individual. Required.</p></li>
<li><p><strong>location</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The flattened location array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">location_offset</span></code>. If not specified or None, an empty location
value is stored for each individual.</p></li>
<li><p><strong>location_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">location</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each individual.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.IndividualTable.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.IndividualTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.packset_location">
<code class="sig-name descname">packset_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">locations</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.packset_location"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.IndividualTable.packset_location" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of location values and updates the <code class="docutils literal notranslate"><span class="pre">location</span></code>
and <code class="docutils literal notranslate"><span class="pre">location_offset</span></code> columns. The length of the locations array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>locations</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of locations interpreted as numpy float64
arrays.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.packset_metadata">
<code class="sig-name descname">packset_metadata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadatas</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.packset_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flags</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">location</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">location_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_schema</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.IndividualTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> using the
values in the specified arrays. Overwrites any data currently stored in
the table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code> array is mandatory and defines the number of individuals
the table will contain.
The <code class="docutils literal notranslate"><span class="pre">location</span></code> and <code class="docutils literal notranslate"><span class="pre">location_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The bitwise flags for each individual. Required.</p></li>
<li><p><strong>location</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The flattened location array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">location_offset</span></code>. If not specified or None, an empty location
value is stored for each individual.</p></li>
<li><p><strong>location_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">location</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each individual.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> – The encoded metadata schema.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.IndividualTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.NodeTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">NodeTable</code><a class="reference internal" href="_modules/tskit/tables.html#NodeTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.NodeTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table defining the nodes in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a node table to be a part of a valid tree sequence.</p>
<dl class="field-list">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>The numpy arrays returned by table attribute accesses are <strong>copies</strong>
of the underlying data. In particular, this means that you cannot edit
the values in the columns by updating the attribute arrays.</p>
<p><strong>NOTE:</strong> this behaviour may change in future.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of time values.</p></li>
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of flags values.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of population IDs.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of individual IDs that each node belongs to.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>) – The metadata schema for this table’s metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.NodeTable.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the specifed row of this table, decoding metadata if it is present.
Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the zero-index of the desired row</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flags</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">time</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">population</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">individual</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#NodeTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.NodeTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> and returns the ID of the
corresponding node. Metadata, if specified, will be validated and encoded
according to the table’s
<a class="reference internal" href="#tskit.NodeTable.metadata_schema" title="tskit.NodeTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The bitwise flags for the new node.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The birth time for the new node.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the population in which the new node was born.
Defaults to <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the individual in which the new node was born.
Defaults to <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – Any object that is valid metadata for the table’s schema.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the newly added node.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flags</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">individual</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#NodeTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.NodeTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns in this
<a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code>, <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">population</span></code> arrays must all be of the same length,
which is equal to the number of nodes that will be added to the table. The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The bitwise flags for each node. Required.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The time values for each node. Required.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The population values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The individual values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.NodeTable.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.NodeTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.packset_metadata">
<code class="sig-name descname">packset_metadata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadatas</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.packset_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flags</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">individual</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_schema</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#NodeTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.NodeTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> using the values in
the specified arrays. Overwrites any data currently stored in the table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code>, <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">population</span></code> arrays must all be of the same length,
which is equal to the number of nodes the table will contain. The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The bitwise flags for each node. Required.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The time values for each node. Required.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The population values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The individual values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> – The encoded metadata schema.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.NodeTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.EdgeTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">EdgeTable</code><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.EdgeTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table defining the edges in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for an edge table to be a part of a valid tree sequence.</p>
<dl class="field-list">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>The numpy arrays returned by table attribute accesses are <strong>copies</strong>
of the underlying data. In particular, this means that you cannot edit
the values in the columns by updating the attribute arrays.</p>
<p><strong>NOTE:</strong> this behaviour may change in future.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of left coordinates.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of right coordinates.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of parent node IDs.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of child node IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>) – The metadata schema for this table’s metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.EdgeTable.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the specifed row of this table, decoding metadata if it is present.
Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the zero-index of the desired row</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">left</span></em>, <em class="sig-param"><span class="n">right</span></em>, <em class="sig-param"><span class="n">parent</span></em>, <em class="sig-param"><span class="n">child</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.EdgeTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> and returns the ID of the
corresponding edge. Metadata, if specified, will be validated and encoded
according to the table’s
<a class="reference internal" href="#tskit.EdgeTable.metadata_schema" title="tskit.EdgeTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The left coordinate (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The right coordinate (exclusive).</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of parent node.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of child node.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – Any object that is valid metadata for the table’s schema.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the newly added edge.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">left</span></em>, <em class="sig-param"><span class="n">right</span></em>, <em class="sig-param"><span class="n">parent</span></em>, <em class="sig-param"><span class="n">child</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.EdgeTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">left</span></code>, <code class="docutils literal notranslate"><span class="pre">right</span></code>, <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code> parameters are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
additional edges to add to the table). The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and
<code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The right coordinates (exclusive).</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The parent node IDs.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The child node IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.EdgeTable.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.EdgeTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.packset_metadata">
<code class="sig-name descname">packset_metadata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadatas</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.packset_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">left</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">right</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">parent</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">child</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_schema</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.EdgeTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> using the values
in the specified arrays. Overwrites any data currently stored in the table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">left</span></code>, <code class="docutils literal notranslate"><span class="pre">right</span></code>, <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code> parameters are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
edges the table will contain).
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together,
and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The right coordinates (exclusive).</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The parent node IDs.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The child node IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> – The encoded metadata schema.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.squash">
<code class="sig-name descname">squash</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.squash"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.EdgeTable.squash" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorts, then condenses the table into the smallest possible number of rows by
combining any adjacent edges.
A pair of edges is said to be <cite>adjacent</cite> if they have the same parent and child
nodes, and if the left coordinate of one of the edges is equal to the right
coordinate of the other edge.
The <code class="docutils literal notranslate"><span class="pre">squash</span></code> method modifies an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> in place so that any set of
adjacent edges is replaced by a single edge.
The new edge will have the same parent and child node, a left coordinate
equal to the smallest left coordinate in the set, and a right coordinate
equal to the largest right coordinate in the set.
The new edge table will be sorted in the canonical order (P, C, L, R).</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.EdgeTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.MigrationTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">MigrationTable</code><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MigrationTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table defining the migrations in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a migration table to be a part of a valid tree
sequence.</p>
<dl class="field-list">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>The numpy arrays returned by table attribute accesses are <strong>copies</strong>
of the underlying data. In particular, this means that you cannot edit
the values in the columns by updating the attribute arrays.</p>
<p><strong>NOTE:</strong> this behaviour may change in future.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of left coordinates.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of right coordinates.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of node IDs.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of source population IDs.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of destination population IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of time values.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>) – The metadata schema for this table’s metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.MigrationTable.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the specifed row of this table, decoding metadata if it is present.
Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the zero-index of the desired row</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">left</span></em>, <em class="sig-param"><span class="n">right</span></em>, <em class="sig-param"><span class="n">node</span></em>, <em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">dest</span></em>, <em class="sig-param"><span class="n">time</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MigrationTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> and returns the ID of the
corresponding migration. Metadata, if specified, will be validated and encoded
according to the table’s
<a class="reference internal" href="#tskit.MigrationTable.metadata_schema" title="tskit.MigrationTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The left coordinate (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The right coordinate (exclusive).</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The node ID.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the source population.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the destination population.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The time of the migration event.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – Any object that is valid metadata for the table’s schema.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the newly added migration.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">left</span></em>, <em class="sig-param"><span class="n">right</span></em>, <em class="sig-param"><span class="n">node</span></em>, <em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">dest</span></em>, <em class="sig-param"><span class="n">time</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MigrationTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>All parameters except <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> and are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
additional migrations to add to the table). The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and
<code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The right coordinates (exclusive).</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The node IDs.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The source population IDs.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The destination population IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int64</em>) – The time of each migration.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each migration.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.MigrationTable.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.MigrationTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.packset_metadata">
<code class="sig-name descname">packset_metadata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadatas</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.packset_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">left</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">right</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">source</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dest</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_schema</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MigrationTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> using the values
in the specified arrays. Overwrites any data currently stored in the table.</p>
<p>All parameters except <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> and are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
migrations the table will contain).
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together,
and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The right coordinates (exclusive).</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The node IDs.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The source population IDs.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The destination population IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int64</em>) – The time of each migration.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each migration.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> – The encoded metadata schema.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MigrationTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.SiteTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">SiteTable</code><a class="reference internal" href="_modules/tskit/tables.html#SiteTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.SiteTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table defining the sites in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a site table to be a part of a valid tree
sequence.</p>
<dl class="field-list">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>The numpy arrays returned by table attribute accesses are <strong>copies</strong>
of the underlying data. In particular, this means that you cannot edit
the values in the columns by updating the attribute arrays.</p>
<p><strong>NOTE:</strong> this behaviour may change in future.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of site position coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of ancestral state strings.
See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>ancestral_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The offsets of rows in the ancestral_state
array. See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>) – The metadata schema for this table’s metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.SiteTable.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the specifed row of this table, decoding metadata if it is present.
Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the zero-index of the desired row</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em>, <em class="sig-param"><span class="n">ancestral_state</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.SiteTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> and returns the ID of the
corresponding site. Metadata, if specified, will be validated and encoded
according to the table’s
<a class="reference internal" href="#tskit.SiteTable.metadata_schema" title="tskit.SiteTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The position of this site in genome coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The state of this site at the root of the tree.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – Any object that is valid metadata for the table’s schema.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the newly added site.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span></em>, <em class="sig-param"><span class="n">ancestral_state</span></em>, <em class="sig-param"><span class="n">ancestral_state_offset</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.SiteTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">position</span></code>, <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The length
of the <code class="docutils literal notranslate"><span class="pre">position</span></code> array determines the number of additional rows
to add the table.
The <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The position of each site in genome coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened ancestral_state array. Required.</p></li>
<li><p><strong>ancestral_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.SiteTable.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.SiteTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.packset_ancestral_state">
<code class="sig-name descname">packset_ancestral_state</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ancestral_states</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.packset_ancestral_state"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.SiteTable.packset_ancestral_state" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of ancestral_state values and updates the
<code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> columns. The length
of the ancestral_states array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ancestral_states</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – A list of string ancestral state values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.packset_metadata">
<code class="sig-name descname">packset_metadata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadatas</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.packset_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">position</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ancestral_state</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ancestral_state_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_schema</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.SiteTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> using the values
in the specified arrays. Overwrites any data currently stored in the table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">position</span></code>, <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The length
of the <code class="docutils literal notranslate"><span class="pre">position</span></code> array determines the number of rows in table.
The <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The position of each site in genome coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened ancestral_state array. Required.</p></li>
<li><p><strong>ancestral_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> – The encoded metadata schema.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.SiteTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.MutationTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">MutationTable</code><a class="reference internal" href="_modules/tskit/tables.html#MutationTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MutationTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table defining the mutations in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a mutation table to be a part of a valid tree
sequence.</p>
<dl class="field-list">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>The numpy arrays returned by table attribute accesses are <strong>copies</strong>
of the underlying data. In particular, this means that you cannot edit
the values in the columns by updating the attribute arrays.</p>
<p><strong>NOTE:</strong> this behaviour may change in future.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of site IDs.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of node IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The array of time values.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of derived state strings.
See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>derived_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The offsets of rows in the derived_state
array. See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The array of parent mutation IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>) – The metadata schema for this table’s metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.MutationTable.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the specifed row of this table, decoding metadata if it is present.
Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the zero-index of the desired row</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site</span></em>, <em class="sig-param"><span class="n">node</span></em>, <em class="sig-param"><span class="n">derived_state</span></em>, <em class="sig-param"><span class="n">parent</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">time</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MutationTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> and returns the ID of the
corresponding mutation. Metadata, if specified, will be validated and encoded
according to the table’s
<a class="reference internal" href="#tskit.MutationTable.metadata_schema" title="tskit.MutationTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the site that this mutation occurs at.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the first node inheriting this mutation.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The state of the site at this mutation’s node.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the parent mutation. If not specified,
defaults to <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-attr docutils literal notranslate"><span class="pre">NULL</span></code></a>.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – Any object that is valid metadata for the table’s schema.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The occurrence time for the new mutation. If not specified,
defaults to <code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code>, indicating the time is unknown.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the newly added mutation.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site</span></em>, <em class="sig-param"><span class="n">node</span></em>, <em class="sig-param"><span class="n">derived_state</span></em>, <em class="sig-param"><span class="n">derived_state_offset</span></em>, <em class="sig-param"><span class="n">parent</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MutationTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">site</span></code>, <code class="docutils literal notranslate"><span class="pre">node</span></code>, <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The
<code class="docutils literal notranslate"><span class="pre">site</span></code> and <code class="docutils literal notranslate"><span class="pre">node</span></code> (also <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">parent</span></code>, if supplied) arrays
must be of equal length, and determine the number of additional
rows to add to the table.
The <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The ID of the site each mutation occurs at.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The ID of the node each mutation is associated with.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The time values for each mutation.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened derived_state array. Required.</p></li>
<li><p><strong>derived_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> array.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The ID of the parent mutation for each mutation.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.MutationTable.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.MutationTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.packset_derived_state">
<code class="sig-name descname">packset_derived_state</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">derived_states</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.packset_derived_state"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MutationTable.packset_derived_state" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of derived_state values and updates the
<code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code> columns. The length
of the derived_states array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>derived_states</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – A list of string derived state values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.packset_metadata">
<code class="sig-name descname">packset_metadata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadatas</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.packset_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">time</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">derived_state</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">derived_state_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">parent</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_schema</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MutationTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> using the values
in the specified arrays. Overwrites any data currently stored in the table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">site</span></code>, <code class="docutils literal notranslate"><span class="pre">node</span></code>, <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The
<code class="docutils literal notranslate"><span class="pre">site</span></code> and <code class="docutils literal notranslate"><span class="pre">node</span></code> (also <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">time</span></code>, if supplied) arrays
must be of equal length, and determine the number of rows in the table.
The <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The ID of the site each mutation occurs at.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The ID of the node each mutation is associated with.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>) – The time values for each mutation.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened derived_state array. Required.</p></li>
<li><p><strong>derived_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> array.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>) – The ID of the parent mutation for each mutation.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> – The encoded metadata schema.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.MutationTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.PopulationTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">PopulationTable</code><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.PopulationTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table defining the populations referred to in a tree sequence.
The PopulationTable stores metadata for populations that may be referred to
in the NodeTable and MigrationTable”.  Note that although nodes
may be associated with populations, this association is stored in
the <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>: only metadata on each population is stored
in the population table.</p>
<dl class="field-list">
<dt class="field-odd">Warning</dt>
<dd class="field-odd"><p>The numpy arrays returned by table attribute accesses are <strong>copies</strong>
of the underlying data. In particular, this means that you cannot edit
the values in the columns by updating the attribute arrays.</p>
<p><strong>NOTE:</strong> this behaviour may change in future.</p>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>) – The metadata schema for this table’s metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.PopulationTable.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the specifed row of this table, decoding metadata if it is present.
Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – the zero-index of the desired row</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.PopulationTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> and returns the ID of the
corresponding population. Metadata, if specified, will be validated and encoded
according to the table’s
<a class="reference internal" href="#tskit.PopulationTable.metadata_schema" title="tskit.PopulationTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.9)"><em>object</em></a>) – Any object that is valid metadata for the table’s schema.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The ID of the newly added population.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.PopulationTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_metadata</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.metadata_schema">
<em class="property">property </em><code class="sig-name descname">metadata_schema</code><a class="headerlink" href="#tskit.PopulationTable.metadata_schema" title="Permalink to this definition">¶</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.PopulationTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.packset_metadata">
<code class="sig-name descname">packset_metadata</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadatas</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.packset_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">metadata_schema</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.PopulationTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> using the
values in the specified arrays. Overwrites any data currently stored in the
table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> – The encoded metadata schema.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.PopulationTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="tskit.ProvenanceTable">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">ProvenanceTable</code><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.ProvenanceTable" title="Permalink to this definition">¶</a></dt>
<dd><p>A table recording the provenance (i.e., history) of this table, so that the
origin of the underlying data and sequence of subsequent operations can be
traced. Each row contains a “record” string (recommended format: JSON) and
a timestamp.</p>
<div class="admonition-todo admonition" id="id5">
<p class="admonition-title">Todo</p>
<p>The format of the <cite>record</cite> field will be more precisely specified in
the future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>record</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array containing the record strings.
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>record_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the record column. See
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened array containing the timestamp strings.
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>timestamp_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>) – The array of offsets into the timestamp column. See
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.ProvenanceTable.add_row">
<code class="sig-name descname">add_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">record</span></em>, <em class="sig-param"><span class="n">timestamp</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.add_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.ProvenanceTable.add_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a new row to this ProvenanceTable consisting of the specified record and
timestamp. If timestamp is not specified, it is automatically generated from
the current time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>record</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A provenance record, describing the parameters and
environment used to generate the current set of tables.</p></li>
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – A string timestamp. This should be in ISO8601 form.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.append_columns">
<code class="sig-name descname">append_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">timestamp</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">timestamp_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.append_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.ProvenanceTable.append_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> and <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information). Likewise
for the <code class="docutils literal notranslate"><span class="pre">record</span></code> and <code class="docutils literal notranslate"><span class="pre">record_offset</span></code> columns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened timestamp array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code>. If not specified or None, an empty timestamp
value is stored for each node.</p></li>
<li><p><strong>timestamp_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> array.</p></li>
<li><p><strong>record</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened record array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">record_offset</span></code>. If not specified or None, an empty record
value is stored for each node.</p></li>
<li><p><strong>record_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">record</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.asdict">
<code class="sig-name descname">asdict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.asdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.clear">
<code class="sig-name descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.equals">
<code class="sig-name descname">equals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">ignore_timestamps</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.equals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.ProvenanceTable.equals" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two provenance
tables are considered equal if their columns are byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> – Another provenance table instance</p></li>
<li><p><strong>ignore_timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True exclude the timestamp column
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if other is equal to this provenance table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.nbytes">
<em class="property">property </em><code class="sig-name descname">nbytes</code><a class="headerlink" href="#tskit.ProvenanceTable.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.packset_record">
<code class="sig-name descname">packset_record</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">records</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.packset_record"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.ProvenanceTable.packset_record" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of record values and updates the
<code class="docutils literal notranslate"><span class="pre">record</span></code> and <code class="docutils literal notranslate"><span class="pre">record_offset</span></code> columns. The length
of the records array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>records</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – A list of string record values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.packset_timestamp">
<code class="sig-name descname">packset_timestamp</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">timestamps</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.packset_timestamp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.ProvenanceTable.packset_timestamp" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of timestamp values and updates the
<code class="docutils literal notranslate"><span class="pre">timestamp</span></code> and <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code> columns. The length
of the timestamps array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – A list of string timestamp values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.set_columns">
<code class="sig-name descname">set_columns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">timestamp</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">timestamp_offset</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">record_offset</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.set_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.ProvenanceTable.set_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a> using the
values in the specified arrays. Overwrites any data currently stored in the
table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> and <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information). Likewise
for the <code class="docutils literal notranslate"><span class="pre">record</span></code> and <code class="docutils literal notranslate"><span class="pre">record_offset</span></code> columns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened timestamp array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code>. If not specified or None, an empty timestamp
value is stored for each node.</p></li>
<li><p><strong>timestamp_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> array.</p></li>
<li><p><strong>record</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>) – The flattened record array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">record_offset</span></code>. If not specified or None, an empty record
value is stored for each node.</p></li>
<li><p><strong>record_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>) – The offsets into the <code class="docutils literal notranslate"><span class="pre">record</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.ProvenanceTable.truncate">
<code class="sig-name descname">truncate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_rows</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="table-functions">
<h3>Table functions<a class="headerlink" href="#table-functions" title="Permalink to this headline">¶</a></h3>
<dl class="py function">
<dt id="tskit.parse_nodes">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">parse_nodes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em>, <em class="sig-param"><span class="n">base64_metadata</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_nodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.parse_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a node table and returns the corresponding <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>
instance. See the <a class="reference internal" href="data-model.html#sec-node-text-format"><span class="std std-ref">node text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node table definition</span></a> section for the
required properties of the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><em>NodeTable</em></a>) – If specified write into this table. If not,
create a new <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.parse_edges">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">parse_edges</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_edges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.parse_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a edge table and returns the corresponding <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>
instance. See the <a class="reference internal" href="data-model.html#sec-edge-text-format"><span class="std std-ref">edge text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge table definition</span></a> section for the
required properties of the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code> parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><em>EdgeTable</em></a>) – If specified, write the edges into this table. If
not, create a new <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance and return.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.parse_sites">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">parse_sites</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em>, <em class="sig-param"><span class="n">base64_metadata</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.parse_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a site table and returns the corresponding <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>
instance. See the <a class="reference internal" href="data-model.html#sec-site-text-format"><span class="std std-ref">site text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site table definition</span></a> section for the
required properties of the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><em>SiteTable</em></a>) – If specified write site into this table. If not,
create a new <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.parse_mutations">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">parse_mutations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em>, <em class="sig-param"><span class="n">base64_metadata</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.parse_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a mutation table and returns the corresponding <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>
instance. See the <a class="reference internal" href="data-model.html#sec-mutation-text-format"><span class="std std-ref">mutation text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation table definition</span></a> section for the
required properties of the contents. Note that if the <code class="docutils literal notranslate"><span class="pre">time</span></code> column is missing its
entries are filled with <code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code>.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><em>MutationTable</em></a>) – If specified, write mutations into this table.
If not, create a new <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.parse_individuals">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">parse_individuals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em>, <em class="sig-param"><span class="n">base64_metadata</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_individuals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.parse_individuals" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of an individual table and returns the corresponding
<a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> instance. See the <a class="reference internal" href="data-model.html#sec-individual-text-format"><span class="std std-ref">individual text format</span></a> section for the details of the required
format and the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual table definition</span></a> section for the required properties of
the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><em>IndividualTable</em></a>) – If specified write into this table. If not,
create a new <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.parse_populations">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">parse_populations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em>, <em class="sig-param"><span class="n">base64_metadata</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_populations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.parse_populations" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a population table and returns the corresponding
<a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> instance. See the <a class="reference internal" href="data-model.html#sec-population-text-format"><span class="std std-ref">population text format</span></a> section for the details of the required
format and the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population table definition</span></a> section for the required properties of
the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.9)"><em>io.TextIOBase</em></a>) – The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><em>PopulationTable</em></a>) – If specified write into this table. If not,
create a new <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.pack_strings">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">pack_strings</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strings</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#pack_strings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.pack_strings" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of strings into a flattened numpy array of 8 bit integers
and corresponding offsets using the specified text encoding.
See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details of this encoding of
columns of variable length data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>]</em>) – The list of strings to encode.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The text encoding to use when converting string data
to bytes. See the <a class="reference external" href="https://docs.python.org/3/library/codecs.html#module-codecs" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">codecs</span></code></a> module for information on available
string encodings.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The tuple (packed, offset) of numpy arrays representing the flattened
input data and offsets.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a> (dtype=np.int8), <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a> (dtype=np.uint32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.unpack_strings">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">unpack_strings</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">packed</span></em>, <em class="sig-param"><span class="n">offset</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf8'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#unpack_strings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.unpack_strings" title="Permalink to this definition">¶</a></dt>
<dd><p>Unpacks a list of strings from the specified numpy arrays of packed byte
data and corresponding offsets using the specified text encoding.
See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details of this encoding of
columns of variable length data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>packed</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – The flattened array of byte values.</p></li>
<li><p><strong>offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – The array of offsets into the <code class="docutils literal notranslate"><span class="pre">packed</span></code> array.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The text encoding to use when converting string data
to bytes. See the <a class="reference external" href="https://docs.python.org/3/library/codecs.html#module-codecs" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">codecs</span></code></a> module for information on available
string encodings.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The list of strings unpacked from the parameter arrays.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.pack_bytes">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">pack_bytes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#pack_bytes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.pack_bytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Packs the specified list of bytes into a flattened numpy array of 8 bit integers
and corresponding offsets. See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details
of this encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.9)"><em>bytes</em></a><em>]</em>) – The list of bytes values to encode.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The tuple (packed, offset) of numpy arrays representing the flattened
input data and offsets.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a> (dtype=np.int8), <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a> (dtype=np.uint32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.unpack_bytes">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">unpack_bytes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">packed</span></em>, <em class="sig-param"><span class="n">offset</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#unpack_bytes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.unpack_bytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Unpacks a list of bytes from the specified numpy arrays of packed byte
data and corresponding offsets. See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details
of this encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>packed</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – The flattened array of byte values.</p></li>
<li><p><strong>offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)"><em>numpy.ndarray</em></a>) – The array of offsets into the <code class="docutils literal notranslate"><span class="pre">packed</span></code> array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The list of bytes values unpacked from the parameter arrays.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.9)">bytes</a>]</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="metadata-api">
<span id="sec-metadata-api"></span><h2>Metadata API<a class="headerlink" href="#metadata-api" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> module provides validation, encoding and decoding of metadata
using a schema. See <a class="reference internal" href="metadata.html#sec-metadata"><span class="std std-ref">Metadata</span></a>, <a class="reference internal" href="metadata.html#sec-metadata-api-overview"><span class="std std-ref">Python Metadata API Overview</span></a> and
<a class="reference internal" href="tutorial.html#sec-tutorial-metadata"><span class="std std-ref">Working with Metadata</span></a>.</p>
<dl class="py class">
<dt id="tskit.MetadataSchema">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">MetadataSchema</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">schema</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>Mapping<span class="p">[</span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a><span class="p">, </span>Any<span class="p">]</span><span class="p">]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MetadataSchema" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for validating, encoding and decoding metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – A dict containing a valid JSONSchema object.</p>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.MetadataSchema.decode_row">
<code class="sig-name descname">decode_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">row</span><span class="p">:</span> <span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.9)">bytes</a></span></em><span class="sig-paren">)</span> &#x2192; Any<a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.decode_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MetadataSchema.decode_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Decode an encoded row (bytes) of metadata, using the codec specifed in the schema
and return a python dict. Note that no validation of the metadata against the
schema is performed.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MetadataSchema.encode_row">
<code class="sig-name descname">encode_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">row</span><span class="p">:</span> <span class="n">Any</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.9)">bytes</a><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.encode_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MetadataSchema.encode_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode a row (dict) of metadata to its binary representation (bytes)
using the codec specified in the schema. Note that unlike
<a class="reference internal" href="#tskit.MetadataSchema.validate_and_encode_row" title="tskit.MetadataSchema.validate_and_encode_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">validate_and_encode_row()</span></code></a> no validation against the schema is performed.
This should only be used for performance if a validation check is not needed.</p>
</dd></dl>

<dl class="py method">
<dt id="tskit.MetadataSchema.validate_and_encode_row">
<code class="sig-name descname">validate_and_encode_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">row</span><span class="p">:</span> <span class="n">Any</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.9)">bytes</a><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.validate_and_encode_row"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.MetadataSchema.validate_and_encode_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Validate a row (dict) of metadata against this schema and return the encoded
representation (bytes) using the codec specified in the schema.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="tskit.register_metadata_codec">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">register_metadata_codec</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">codec_cls</span><span class="p">:</span> <span class="n">Type<span class="p">[</span>tskit.metadata.AbstractMetadataCodec<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">codec_id</span><span class="p">:</span> <span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></span></em><span class="sig-paren">)</span> &#x2192; <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a><a class="reference internal" href="_modules/tskit/metadata.html#register_metadata_codec"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.register_metadata_codec" title="Permalink to this definition">¶</a></dt>
<dd><p>Register a metadata codec class.
This function maintains a mapping from metadata codec identifiers used in schemas
to codec classes. When a codec class is registered, it will replace any class
previously registered under the same codec identifier, if present.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>codec_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – String to use to refer to the codec in the schema.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="combinatorics-api">
<span id="sec-combinatorics-api"></span><h2>Combinatorics API<a class="headerlink" href="#combinatorics-api" title="Permalink to this headline">¶</a></h2>
<p>The combinatorics API deals with tree topologies, allowing them to be counted,
listed and generated: see <a class="reference internal" href="combinatorics.html#sec-combinatorics"><span class="std std-ref">Combinatorics</span></a> for a detailed description. Briefly,
the position of a tree in the enumeration <code class="docutils literal notranslate"><span class="pre">all_trees</span></code> can be obtained using the tree’s
<a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rank()</span></code></a> method. Inversely, a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> can be constructed from a position
in the enumeration with <a class="reference internal" href="#tskit.Tree.unrank" title="tskit.Tree.unrank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.unrank()</span></code></a>. Generated trees are associated with a new
tree sequence containing only that tree for the entire genome (i.e. with
<a class="reference internal" href="#tskit.TreeSequence.num_trees" title="tskit.TreeSequence.num_trees"><code class="xref py py-attr docutils literal notranslate"><span class="pre">num_trees</span></code></a> = 1 and a <a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> equal to
the <a class="reference internal" href="#tskit.Tree.span" title="tskit.Tree.span"><code class="xref py py-attr docutils literal notranslate"><span class="pre">span</span></code></a> of the tree).</p>
<dl class="py function">
<dt id="tskit.all_trees">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">all_trees</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_leaves</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/combinatorics.html#all_trees"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.all_trees" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates all unique leaf-labelled trees with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code>
leaves. See <a class="reference internal" href="combinatorics.html#sec-combinatorics"><span class="std std-ref">Combinatorics</span></a> on the details of this
enumeration. The leaf labels are selected from the set
<code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">num_leaves)</span></code>. The times and labels on internal nodes are
chosen arbitrarily.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of leaves of the tree to generate.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The genomic span of each returned tree.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.all_tree_shapes">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">all_tree_shapes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_leaves</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/combinatorics.html#all_tree_shapes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.all_tree_shapes" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates all unique shapes of trees with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of leaves of the tree to generate.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The genomic span of each returned tree.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="tskit.all_tree_labellings">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">all_tree_labellings</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tree</span></em>, <em class="sig-param"><span class="n">span</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/combinatorics.html#all_tree_labellings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.all_tree_labellings" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates all unique labellings of the leaves of a
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.Tree</span></code></a>. Leaves are labelled from the set
<code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">n)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of leaves of <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><em>tskit.Tree</em></a>) – The tree used to generate
labelled trees of the same shape.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The genomic span of each returned tree.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="tskit.TopologyCounter">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">TopologyCounter</code><a class="reference internal" href="_modules/tskit/combinatorics.html#TopologyCounter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.TopologyCounter" title="Permalink to this definition">¶</a></dt>
<dd><p>Contains the distributions of embedded topologies for every combination
of the sample sets used to generate the <code class="docutils literal notranslate"><span class="pre">TopologyCounter</span></code>. It is
indexable by a combination of sample set indexes and returns a
<code class="docutils literal notranslate"><span class="pre">collections.Counter</span></code> whose keys are topology ranks
(see <a class="reference internal" href="combinatorics.html#sec-tree-ranks"><span class="std std-ref">Interpreting Tree Ranks</span></a>). See <a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a> for more
detail on how this structure is used.</p>
</dd></dl>

</div>
<div class="section" id="linkage-disequilibrium">
<h2>Linkage disequilibrium<a class="headerlink" href="#linkage-disequilibrium" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This API will soon be deprecated in favour of multi-site extensions
to the <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a> API.</p>
</div>
<dl class="py class">
<dt id="tskit.LdCalculator">
<em class="property">class </em><code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">LdCalculator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tree_sequence</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.LdCalculator" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for calculating <a class="reference external" href="https://en.wikipedia.org/wiki/Linkage_disequilibrium">linkage disequilibrium</a> coefficients
between pairs of mutations in a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. This class requires
the <a class="reference external" href="http://www.numpy.org/">numpy</a> library.</p>
<p>This class supports multithreaded access using the Python <a class="reference external" href="https://docs.python.org/3/library/threading.html#module-threading" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">threading</span></code></a>
module. Separate instances of <a class="reference internal" href="#tskit.LdCalculator" title="tskit.LdCalculator"><code class="xref py py-class docutils literal notranslate"><span class="pre">LdCalculator</span></code></a> referencing the
same tree sequence can operate in parallel in multiple threads.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class does not currently support sites that have more than one
mutation. Using it on such a tree sequence will raise a LibraryError with
an “Unsupported operation” message.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tree_sequence</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>) – The tree sequence containing the
mutations we are interested in.</p>
</dd>
</dl>
<dl class="py method">
<dt id="tskit.LdCalculator.r2">
<code class="sig-name descname">r2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">a</span></em>, <em class="sig-param"><span class="n">b</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator.r2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.LdCalculator.r2" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the value of the <span class="math notranslate nohighlight">\(r^2\)</span> statistic between the pair of
mutations at the specified indexes. This method is <em>not</em> an efficient
method for computing large numbers of pairwise; please use either
<a class="reference internal" href="#tskit.LdCalculator.r2_array" title="tskit.LdCalculator.r2_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">r2_array()</span></code></a> or <a class="reference internal" href="#tskit.LdCalculator.r2_matrix" title="tskit.LdCalculator.r2_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">r2_matrix()</span></code></a> for this purpose.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The index of the first mutation.</p></li>
<li><p><strong>b</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The index of the second mutation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The value of <span class="math notranslate nohighlight">\(r^2\)</span> between the mutations at indexes
<code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">b</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.LdCalculator.r2_array">
<code class="sig-name descname">r2_array</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">a</span></em>, <em class="sig-param"><span class="n">direction</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">max_mutations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_distance</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator.r2_array"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.LdCalculator.r2_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the value of the <span class="math notranslate nohighlight">\(r^2\)</span> statistic between the focal
mutation at index <span class="math notranslate nohighlight">\(a\)</span> and a set of other mutations. The method
operates by starting at the focal mutation and iterating over adjacent
mutations (in either the forward or backwards direction) until either a
maximum number of other mutations have been considered (using the
<code class="docutils literal notranslate"><span class="pre">max_mutations</span></code> parameter), a maximum distance in sequence
coordinates has been reached (using the <code class="docutils literal notranslate"><span class="pre">max_distance</span></code> parameter) or
the start/end of the sequence has been reached. For every mutation
<span class="math notranslate nohighlight">\(b\)</span> considered, we then insert the value of <span class="math notranslate nohighlight">\(r^2\)</span> between
<span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(b\)</span> at the corresponding index in an array, and
return the entire array. If the returned array is <span class="math notranslate nohighlight">\(x\)</span> and
<code class="docutils literal notranslate"><span class="pre">direction</span></code> is <a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.FORWARD</span></code></a> then <span class="math notranslate nohighlight">\(x[0]\)</span> is the
value of the statistic for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a + 1\)</span>, <span class="math notranslate nohighlight">\(x[1]\)</span>
the value for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a + 2\)</span>, etc. Similarly, if
<code class="docutils literal notranslate"><span class="pre">direction</span></code> is <a class="reference internal" href="#tskit.REVERSE" title="tskit.REVERSE"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.REVERSE</span></code></a> then <span class="math notranslate nohighlight">\(x[0]\)</span> is the
value of the statistic for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a - 1\)</span>, <span class="math notranslate nohighlight">\(x[1]\)</span>
the value for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a - 2\)</span>, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The index of the focal mutation.</p></li>
<li><p><strong>direction</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The direction in which to travel when
examining other mutations. Must be either
<a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.FORWARD</span></code></a> or <a class="reference internal" href="#tskit.REVERSE" title="tskit.REVERSE"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.REVERSE</span></code></a>. Defaults
to <a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.FORWARD</span></code></a>.</p></li>
<li><p><strong>max_mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The maximum number of mutations to return
<span class="math notranslate nohighlight">\(r^2\)</span> values for. Defaults to as many mutations as
possible.</p></li>
<li><p><strong>max_distance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The maximum absolute distance between
the focal mutation and those for which <span class="math notranslate nohighlight">\(r^2\)</span> values
are returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An array of double precision floating point values
representing the <span class="math notranslate nohighlight">\(r^2\)</span> values for mutations in the
specified direction.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a></p>
</dd>
<dt class="field-even">Warning</dt>
<dd class="field-even"><p>For efficiency reasons, the underlying memory used to
store the returned array is shared between calls. Therefore,
if you wish to store the results of a single call to
<code class="docutils literal notranslate"><span class="pre">get_r2_array()</span></code> for later processing you <strong>must</strong> take a
copy of the array!</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="tskit.LdCalculator.r2_matrix">
<code class="sig-name descname">r2_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator.r2_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.LdCalculator.r2_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the complete <span class="math notranslate nohighlight">\(m \times m\)</span> matrix of pairwise
<span class="math notranslate nohighlight">\(r^2\)</span> values in a tree sequence with <span class="math notranslate nohighlight">\(m\)</span> mutations.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An 2 dimensional square array of double precision
floating point values representing the <span class="math notranslate nohighlight">\(r^2\)</span> values for
all pairs of mutations.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.19)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="provenance">
<span id="sec-provenance-api"></span><h2>Provenance<a class="headerlink" href="#provenance" title="Permalink to this headline">¶</a></h2>
<p>We provide some preliminary support for validating JSON documents against the
<a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">provenance schema</span></a>. Programmatic access to provenance
information is planned for future versions.</p>
<dl class="py function">
<dt id="tskit.validate_provenance">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">validate_provenance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">provenance</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/provenance.html#validate_provenance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tskit.validate_provenance" title="Permalink to this definition">¶</a></dt>
<dd><p>Validates the specified dict-like object against the tskit
<a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">provenance schema</span></a>. If the input does
not represent a valid instance of the schema an exception is
raised.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – The dictionary representing a JSON document
to be validated against the schema.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.ProvenanceValidationError" title="tskit.ProvenanceValidationError"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.ProvenanceValidationError</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="tskit.ProvenanceValidationError">
<code class="sig-prename descclassname">tskit.</code><code class="sig-name descname">ProvenanceValidationError</code><a class="headerlink" href="#tskit.ProvenanceValidationError" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="stats.html" class="btn btn-neutral float-right" title="Statistics" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2018-2020, Tskit developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>