

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Statistics &mdash; tskit 0.3.4.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="C API" href="c-api.html" />
    <link rel="prev" title="Python API" href="python-api.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> tskit
          

          
            
            <img src="_static/tskit_logo_pale.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Statistics</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#available-statistics">Available statistics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#single-site-statistics">Single site statistics</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#patterson-s-f-statistics">Patterson’s f statistics</a></li>
<li class="toctree-l4"><a class="reference internal" href="#y-statistics">Y statistics</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trait-correlations">Trait correlations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#derived-statistics">Derived statistics</a></li>
<li class="toctree-l4"><a class="reference internal" href="#general-methods">General methods</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#allele-frequency-spectrum">Allele frequency spectrum</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#interface">Interface</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mode">Mode</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sample-sets-and-indexes">Sample sets and indexes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#one-way-methods">One-way methods</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multi-way-methods">Multi-way methods</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#windows">Windows</a></li>
<li class="toctree-l3"><a class="reference internal" href="#span-normalise">Span normalise</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-format">Output format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-dimensions">Output dimensions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#general-api">General API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#polarisation">Polarisation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#summary-functions">Summary functions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="c-api.html">C API</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="data-model.html">Data model</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="combinatorics.html">Combinatorics</a></li>
<li class="toctree-l1"><a class="reference internal" href="provenance.html">Provenance</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">tskit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Statistics</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/tskit-dev/tskit/blob/main/docs/stats.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="statistics">
<span id="sec-stats"></span><h1>Statistics<a class="headerlink" href="#statistics" title="Permalink to this headline">¶</a></h1>
<p>There is a unified interface for computing many types of summary statistics from tree sequences.
These are implemented in a flexible way that
– like the tree sequence itself –
exploits the duality between mutations and branches in the trees
to compute statistics of both genome sequence
(whose definition does not depend on the trees)
and of the underlying trees (whose definition does not depend on the genome sequence).
Furthermore, these statistics have a common interface to easily compute
(a) averaged statistics in windows along the genome,
and (b) summary statistics of many combinations of sets of samples simultaneously.
All methods return numpy arrays whose dimensions are
determined by the parameters (see <a class="reference internal" href="#sec-stats-output-dimensions"><span class="std std-ref">Output dimensions</span></a>).</p>
<p>Please see the <a class="reference internal" href="tutorial.html#sec-tutorial-stats"><span class="std std-ref">tutorial</span></a> for examples of the
statistics API in use.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">Missing data</span></a> is not currently
handled correctly by site statistics defined here, as we always
assign missing data to be equal to the ancestral state. Later
versions will add this behaviour as an option and will account
for the presence of missing data by default.</p>
</div>
<div class="section" id="available-statistics">
<h2>Available statistics<a class="headerlink" href="#available-statistics" title="Permalink to this headline">¶</a></h2>
<p>Here are the statistics that can be computed using <code class="docutils literal notranslate"><span class="pre">tskit</span></code>,
grouped by basic classification and type.</p>
<div class="section" id="single-site-statistics">
<h3>Single site statistics<a class="headerlink" href="#single-site-statistics" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.diversity()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.divergence()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.segregating_sites" title="tskit.TreeSequence.segregating_sites"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.segregating_sites()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.allele_frequency_spectrum" title="tskit.TreeSequence.allele_frequency_spectrum"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.allele_frequency_spectrum()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness()</span></code></a></p></li>
</ul>
<div class="section" id="patterson-s-f-statistics">
<h4>Patterson’s f statistics<a class="headerlink" href="#patterson-s-f-statistics" title="Permalink to this headline">¶</a></h4>
<p>These are the <cite>f</cite> statistics (also called <cite>F</cite> statistics) introduced in
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2842210/">Reich et al (2009)</a>.
See the documentation (link below) for the definition,
and <a class="reference external" href="https://www.genetics.org/content/202/4/1485">Peter (2016)</a> for readable
discussion of their use.</p>
<ul class="simple">
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.f4()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.f3" title="tskit.TreeSequence.f3"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.f3()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.f2" title="tskit.TreeSequence.f2"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.f2()</span></code></a></p></li>
</ul>
</div>
<div class="section" id="y-statistics">
<h4>Y statistics<a class="headerlink" href="#y-statistics" title="Permalink to this headline">¶</a></h4>
<p>These are the <cite>Y</cite> statistics introduced by
<a class="reference external" href="https://www.biorxiv.org/content/10.1101/354530v1">Ashander et al (2018)</a>
as a three-sample intermediate between diversity/divergence (which are
pairwise) and Patterson’s f statistics (which are four-way).</p>
<ul class="simple">
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.Y3" title="tskit.TreeSequence.Y3"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.Y3()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.Y2" title="tskit.TreeSequence.Y2"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.Y2()</span></code></a></p></li>
</ul>
</div>
<div class="section" id="trait-correlations">
<h4>Trait correlations<a class="headerlink" href="#trait-correlations" title="Permalink to this headline">¶</a></h4>
<p>These methods compute correlations and covariances of traits (i.e., an
arbitrary vector) with allelic state, possibly in the context of a multivariate
linear model with other covariates (as in GWAS).</p>
<ul class="simple">
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trait_covariance()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.trait_correlation" title="tskit.TreeSequence.trait_correlation"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trait_correlation()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.trait_linear_model" title="tskit.TreeSequence.trait_linear_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trait_linear_model()</span></code></a></p></li>
</ul>
</div>
<div class="section" id="derived-statistics">
<h4>Derived statistics<a class="headerlink" href="#derived-statistics" title="Permalink to this headline">¶</a></h4>
<p>The other statistics above all have the property that <cite>mode=”branch”</cite> and
<cite>mode=”site”</cite> are “dual” in the sense that they are equal, on average, under
a high neutral mutation rate. The following statistics do not have this
property (since both are ratios of statistics that do have this property).</p>
<ul class="simple">
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.Fst" title="tskit.TreeSequence.Fst"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.Fst()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.Tajimas_D" title="tskit.TreeSequence.Tajimas_D"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.Tajimas_D()</span></code></a></p></li>
</ul>
</div>
<div class="section" id="general-methods">
<h4>General methods<a class="headerlink" href="#general-methods" title="Permalink to this headline">¶</a></h4>
<p>These methods allow access to the general method of computing statistics,
using weights or sample counts, and summary functions. See the documentation
for more details. The pre-implemented statistics above will be faster than
using these methods directly, so they should be preferred.</p>
<ul class="simple">
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.general_stat()</span></code></a></p></li>
<li><p><a class="reference internal" href="python-api.html#tskit.TreeSequence.sample_count_stat" title="tskit.TreeSequence.sample_count_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.sample_count_stat()</span></code></a></p></li>
</ul>
</div>
</div>
<div class="section" id="allele-frequency-spectrum">
<span id="sec-stats-afs"></span><h3>Allele frequency spectrum<a class="headerlink" href="#allele-frequency-spectrum" title="Permalink to this headline">¶</a></h3>
<p>We compute allele frequency spectra (AFS), including windowed and joint spectra,
using the same general pattern as other statistics,
but some of the details about how it is defined,
especially in the presence of multiple alleles per site, need to be explained.
If all sites are biallelic, then the result is just as you’d expect:
see the method documentation at <a class="reference internal" href="python-api.html#tskit.TreeSequence.allele_frequency_spectrum" title="tskit.TreeSequence.allele_frequency_spectrum"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.allele_frequency_spectrum()</span></code></a>
for the description.
Note that with <code class="docutils literal notranslate"><span class="pre">mode=&quot;site&quot;</span></code>, we really do tabulate <em>allele</em> counts:
if more than one mutation on different parts of the tree produce the same allele,
it is the total number with this allele (i.e., inheriting <em>either</em> mutation)
that goes into the AFS.
The AFS with <code class="docutils literal notranslate"><span class="pre">mode=&quot;branch&quot;</span></code> is the expected value for the Site AFS
with infinite-sites, biallelic mutation, so there is nothing surprising there,
either.</p>
<p>But, how do we deal with sites at which there are more than two alleles?
At each site, we iterate over the distinct alleles at that site,
and for each allele, count how many samples in each sample set
have inherited that allele.
For a concrete example, suppose that we are computing the AFS of a single
sample set with 10 samples, and are considering a site with three alleles:
<em>a</em>, <em>b</em>, and <em>c</em>,
which have been inherited by 6, 3, and 1 samples, respectively,
and that allele <em>a</em> is ancestral.
What we do at this site depends on if the AFS is polarised or not.</p>
<p>If we are computing the <em>polarised</em> AFS,
we add 1 to each entry of the output corresponding to each allele count
<em>except</em> the ancestral allele.
In our example, we’d add 1 to both <code class="docutils literal notranslate"><span class="pre">AFS[3]</span></code> and <code class="docutils literal notranslate"><span class="pre">AFS[1]</span></code>.
This means that the sum of all entries of a polarised, site AFS
should equal the total number of non-ancestral alleles in the tree sequence
that are ancestral to at least one of the samples in the tree sequence.
The reason for this last caveat is that like with most statistics,
mutations that are not ancestral to <em>any</em> samples (not just those in the sample sets)
are not counted (and so don’t even enter into <code class="docutils literal notranslate"><span class="pre">AFS[0]</span></code>.</p>
<p>Now, if we are computing the <em>unpolarised</em> AFS,
we add <em>one half</em> to each entry of the <em>folded</em> output
corresponding to each allele count <em>including</em> the ancestral allele.
What does this mean?
Well, <code class="docutils literal notranslate"><span class="pre">polarised=False</span></code> means that we cannot distinguish between an
allele count of 6 and an allele count of 4.
So, <em>folding</em> means that we would add our allele that is seen in 6 samples
to <code class="docutils literal notranslate"><span class="pre">AFS[4]</span></code> instead of <code class="docutils literal notranslate"><span class="pre">AFS[6]</span></code>.
So, in total, we will add 0.5 to each of <code class="docutils literal notranslate"><span class="pre">AFS[4]</span></code>, <code class="docutils literal notranslate"><span class="pre">AFS[3]</span></code>, and <code class="docutils literal notranslate"><span class="pre">AFS[1]</span></code>.
This means that the sum of an unpolarised AFS
will be equal to the total number of alleles that are inherited
by any of the samples in the tree sequence, divided by two.
Why one-half? Well, notice that if in fact the mutation that produced the <em>b</em>
allele had instead produced an <em>a</em> allele,
so that the site had only two alleles, with frequencies 7 and 3.
Then, we would have added 0.5 to <code class="docutils literal notranslate"><span class="pre">AFS[3]</span></code> <em>twice</em>.</p>
</div>
</div>
<div class="section" id="interface">
<span id="sec-stats-interface"></span><h2>Interface<a class="headerlink" href="#interface" title="Permalink to this headline">¶</a></h2>
<p>Tskit offers a powerful and flexible interface for computing population genetic
statistics. Consequently, the interface is a little complicated and there are a
lot of options. However, we provide sensible defaults for these options and
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> should do what you want in most cases. There are several major options
shared by many statistics, which we describe in detail in the following subsections:</p>
<dl class="simple">
<dt><a class="reference internal" href="#sec-stats-mode"><span class="std std-ref">Mode</span></a></dt><dd><p>What are we summarising information about?</p>
</dd>
<dt><a class="reference internal" href="#sec-stats-windows"><span class="std std-ref">Windows</span></a></dt><dd><p>What section(s) of the genome are we interested in?</p>
</dd>
<dt><a class="reference internal" href="#sec-stats-span-normalise"><span class="std std-ref">Span normalise</span></a></dt><dd><p>Should the statistic calculated for each window be normalised by the span
(i.e. the sequence length) of that window?</p>
</dd>
</dl>
<p>The statistics functions are highly efficient and are based where possible
on numpy arrays. Each of these statistics will return the results as a numpy
array, and the format of this array will depend on the statistic being
computed (see the <a class="reference internal" href="#sec-stats-output-format"><span class="std std-ref">Output format</span></a> section for details).
A convenient feature of the statistics API is that the dimensions of the
output array is defined in a simple and intuitive manner by the
parameters provided. The <a class="reference internal" href="#sec-stats-output-dimensions"><span class="std std-ref">Output dimensions</span></a> section
defines the rules that are used.</p>
<p>Please see the <a class="reference internal" href="tutorial.html#sec-tutorial-stats"><span class="std std-ref">tutorial</span></a> for examples of the
statistics API in use.</p>
<div class="section" id="mode">
<span id="sec-stats-mode"></span><h3>Mode<a class="headerlink" href="#mode" title="Permalink to this headline">¶</a></h3>
<p>There are three <strong>modes</strong> of statistic: <code class="docutils literal notranslate"><span class="pre">site</span></code>, <code class="docutils literal notranslate"><span class="pre">branch</span></code>, and <code class="docutils literal notranslate"><span class="pre">node</span></code>,
that each summarize aspects of the tree sequence in different but related ways.
Roughly speaking, these answer the following sorts of question:</p>
<dl class="simple">
<dt>site</dt><dd><p>How many mutations differentiate these two genomes?</p>
</dd>
<dt>branch</dt><dd><p>How long since these genomes’ common ancestor?</p>
</dd>
<dt>node</dt><dd><p>On how much of the genome is each node an ancestor of only one of these genomes, but not both?</p>
</dd>
</dl>
<p>These three examples can all be answered in the same way with the tree sequence:
first, draw all the paths from one genome to the other through the tree sequence
(back up to their common ancestor and back down in each marginal tree).
Then,
(<code class="docutils literal notranslate"><span class="pre">site</span></code>) count the number of mutations falling on the paths,
(<code class="docutils literal notranslate"><span class="pre">branch</span></code>) measure the length of the paths, or
(<code class="docutils literal notranslate"><span class="pre">node</span></code>) count how often the path goes through each node.
There is more discussion of this correspondence in the paper describing these statistics,
and precise definitions are given in each statistic.</p>
<p>One important thing to know is that <code class="docutils literal notranslate"><span class="pre">node</span></code> statistics have somewhat different output.
While <code class="docutils literal notranslate"><span class="pre">site</span></code> and <code class="docutils literal notranslate"><span class="pre">branch</span></code> statistics naturally return one number
for each portion of the genome (and thus incorporates information about many nodes: see below),
the <code class="docutils literal notranslate"><span class="pre">node</span></code> statistics return one number <strong>for each node</strong> in the tree sequence (and for each window).
There can be a lot of nodes in the tree sequence, so beware.</p>
<p>Also remember that in a tree sequence the “sites” are usually just the <strong>variant</strong> sites,
e.g., the sites of the SNPs.
(Although tree sequence may in principle have monomorphic sites, those produced by simulation usually don’t.)</p>
</div>
<div class="section" id="sample-sets-and-indexes">
<span id="sec-stats-sample-sets"></span><h3>Sample sets and indexes<a class="headerlink" href="#sample-sets-and-indexes" title="Permalink to this headline">¶</a></h3>
<p>Many standard population genetics statistics
are defined with respect to some number of groups of genomes,
usually called “populations”.
(However, it’s clear from the correspondence to descriptors of tree shape
that the definitions can usefully describe <em>something</em>
even if the groups of samples don’t come from “separate populations” in some sense.)
Basically, statistics defined in terms of sample sets can use the frequency of any allele
in each of the sample sets when computing the statistic.
For instance, nucleotide divergence is defined for a <em>pair</em> of groups of samples,
so if you wanted to compute pairwise divergences between some groups of samples,
you’d specify these as your <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Then, if <code class="docutils literal notranslate"><span class="pre">p[i]</span></code> is the derived allele frequency in sample set <code class="docutils literal notranslate"><span class="pre">i</span></code>,
under the hood we (essentially) compute the divergence between sample sets <code class="docutils literal notranslate"><span class="pre">i</span></code> and <code class="docutils literal notranslate"><span class="pre">j</span></code>
by averaging <code class="docutils literal notranslate"><span class="pre">p[i]</span> <span class="pre">*</span> <span class="pre">(1</span> <span class="pre">-</span> <span class="pre">p[j])</span> <span class="pre">+</span> <span class="pre">(1</span> <span class="pre">-</span> <span class="pre">p[i])</span> <span class="pre">*</span> <span class="pre">p[j]</span></code> across the genome.</p>
<p>So, what if you
have samples from each of 10 populations,
and want to compute <strong>all</strong> fourty-five pairwise divergences?
You could call <code class="docutils literal notranslate"><span class="pre">divergence</span></code> fourty-five times, but this would be tedious
and also inefficient, because the allele frequencies for one population
gets used in computing many of those values.
So, statistics that take a <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument also take an <code class="docutils literal notranslate"><span class="pre">indexes</span></code> argument,
which for a statistic that operates on <code class="docutils literal notranslate"><span class="pre">k</span></code> sample sets will be a list of <code class="docutils literal notranslate"><span class="pre">k</span></code>-tuples.
If <code class="docutils literal notranslate"><span class="pre">indexes</span></code> is a length <code class="docutils literal notranslate"><span class="pre">n</span></code> list of <code class="docutils literal notranslate"><span class="pre">k</span></code>-tuples,
then the output will have <code class="docutils literal notranslate"><span class="pre">n</span></code> columns,
and if <code class="docutils literal notranslate"><span class="pre">indexes[j]</span></code> is a tuple <code class="docutils literal notranslate"><span class="pre">(i0,</span> <span class="pre">...,</span> <span class="pre">ik)</span></code>,
then the <code class="docutils literal notranslate"><span class="pre">j</span></code>-th column will contain values of the statistic computed on
<code class="docutils literal notranslate"><span class="pre">(sample_sets[i0],</span> <span class="pre">sample_sets[i1],</span> <span class="pre">...,</span> <span class="pre">sample_sets[ik])</span></code>.</p>
<p>How multiple statistics are handled differs slightly between statistics
that operate on single sample sets and multiple sample sets.</p>
<div class="section" id="one-way-methods">
<span id="sec-stats-sample-sets-one-way"></span><h4>One-way methods<a class="headerlink" href="#one-way-methods" title="Permalink to this headline">¶</a></h4>
<p>One-way statistics such as <a class="reference internal" href="python-api.html#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.diversity()</span></code></a> are defined over a single
sample set. For these methods, <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> is interpreted in the following way:</p>
<ul class="simple">
<li><p>If it is a single list of node IDs (e.g., <code class="docutils literal notranslate"><span class="pre">sample_sets=[0,</span> <span class="pre">1</span> <span class="pre">,2]</span></code>), this is
interpreted as running the calculation over one sample set and we remove
the last dimension in the result array as described in the
<a class="reference internal" href="#sec-stats-output-dimensions"><span class="std std-ref">Output dimensions</span></a> section.</p></li>
<li><p>If it is <code class="docutils literal notranslate"><span class="pre">None</span></code> (the default), this is equivalent to <code class="docutils literal notranslate"><span class="pre">sample_sets=ts.samples()</span></code>,
and we therefore compute the statistic over all samples in the tree sequence. <strong>Note
that we also drop the outer dimension of the result array in this case</strong>.</p></li>
<li><p>If it is a list of lists of samples we return an array for each window in the output,
which contains the value of the statistic separately for each of <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>
in the order they are given.</p></li>
</ul>
</div>
<div class="section" id="multi-way-methods">
<span id="sec-stats-sample-sets-multi-way"></span><h4>Multi-way methods<a class="headerlink" href="#multi-way-methods" title="Permalink to this headline">¶</a></h4>
<p>Multi-way statistics such as <a class="reference internal" href="python-api.html#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.divergence()</span></code></a> are defined over a
<code class="docutils literal notranslate"><span class="pre">k</span></code> sample sets.</p>
<p>In this case, <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> must be a list of lists of sample IDs, and there
is no default.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">indexes</span></code> parameter is interpreted in the following way:</p>
<ul class="simple">
<li><p>If it is a single <code class="docutils literal notranslate"><span class="pre">k</span></code>-tuple, this is interpreted as computing a single
statistic selecting the specified sample sets and we remove the last dimension
in the result array as described in the <a class="reference internal" href="#sec-stats-output-dimensions"><span class="std std-ref">Output dimensions</span></a> section.</p></li>
<li><p>If if is <code class="docutils literal notranslate"><span class="pre">None</span></code> and <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> contains exactly <code class="docutils literal notranslate"><span class="pre">k</span></code> sample sets,
this is equivalent to <code class="docutils literal notranslate"><span class="pre">indexes=range(k)</span></code>. <strong>Note
that we also drop the outer dimension of the result array in this case</strong>.</p></li>
<li><p>If is is a list of <code class="docutils literal notranslate"><span class="pre">k</span></code>-tuples (each consisting of integers
of integers between <code class="docutils literal notranslate"><span class="pre">0</span></code> and <code class="docutils literal notranslate"><span class="pre">len(sample_sets)</span> <span class="pre">-</span> <span class="pre">1</span></code>) of length <code class="docutils literal notranslate"><span class="pre">n</span></code> we
compute <code class="docutils literal notranslate"><span class="pre">n</span></code> statistics based on these selections of sample sets.</p></li>
</ul>
</div>
</div>
<div class="section" id="windows">
<span id="sec-stats-windows"></span><h3>Windows<a class="headerlink" href="#windows" title="Permalink to this headline">¶</a></h3>
<p>Each statistic has an argument, <code class="docutils literal notranslate"><span class="pre">windows</span></code>,
which defines a collection of contiguous windows spanning the genome.
<code class="docutils literal notranslate"><span class="pre">windows</span></code> should be a list of <code class="docutils literal notranslate"><span class="pre">n+1</span></code> increasing numbers beginning with 0
and ending with the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>.
The statistic will be computed separately in each of the <code class="docutils literal notranslate"><span class="pre">n</span></code> windows,
and the <code class="docutils literal notranslate"><span class="pre">k</span></code>-th row of the output will report the values of the statistic
in the <code class="docutils literal notranslate"><span class="pre">k</span></code>-th window, i.e., from (and including) <code class="docutils literal notranslate"><span class="pre">windows[k]</span></code> to (but not including) <code class="docutils literal notranslate"><span class="pre">windows[k+1]</span></code>.</p>
<p>Most windowed statistics by default return <strong>averages</strong> within each of the windows,
so the values are comparable between windows, even of different spans.
(However, shorter windows may be noisier.)
Suppose for instance  that you compute some statistic with <code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">[0,</span> <span class="pre">a,</span> <span class="pre">b]</span></code>
for some valid positions <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;</span> <span class="pre">a</span> <span class="pre">&lt;</span> <span class="pre">b</span></code>,
and get an output array <code class="docutils literal notranslate"><span class="pre">S</span></code> with two rows.
Then, computing the same statistic with <code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">[0,</span> <span class="pre">b]</span></code>
would be equivalent to averaging the rows of <code class="docutils literal notranslate"><span class="pre">S</span></code>,
obtaining <code class="docutils literal notranslate"><span class="pre">((a</span> <span class="pre">-</span> <span class="pre">0)</span> <span class="pre">*</span> <span class="pre">S[0]</span> <span class="pre">+</span> <span class="pre">(b</span> <span class="pre">-</span> <span class="pre">a)</span> <span class="pre">*</span> <span class="pre">S[1])</span> <span class="pre">/</span> <span class="pre">(b</span> <span class="pre">-</span> <span class="pre">0)</span></code>.</p>
<p>There are some shortcuts to other useful options:</p>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">None</span></code></dt><dd><p>This is the default and computes statistics in single window over the whole
sequence. As the first returned array contains only a single
value, we drop this dimension as described in the <a class="reference internal" href="#sec-stats-output-dimensions"><span class="std std-ref">output dimensions</span></a> section. <strong>Note:</strong> if you really do
want to have an array with a single value as the result, please use
<code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">[0.0,</span> <span class="pre">ts.sequence_length]</span></code>.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">&quot;trees&quot;</span></code></dt><dd><p>This says that you want statistics computed separately on the portion of the genome
spanned by each tree, so is equivalent to passing <code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">ts.breakpoints()</span></code>.
(Beware: there can be a lot of trees!)</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">&quot;sites&quot;</span></code></dt><dd><p>This says to output one set of values for <strong>each site</strong>,
and is equivalent to passing <code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">[s.position</span> <span class="pre">for</span> <span class="pre">s</span> <span class="pre">in</span> <span class="pre">ts.sites()]</span> <span class="pre">+</span> <span class="pre">[ts.sequence_length]</span></code>.
This will return one statistic for each site (beware!);
since the windows are all different sizes you probably want to also pass
<code class="docutils literal notranslate"><span class="pre">span_normalise=False</span></code> (see below).</p>
</dd>
</dl>
</div>
<div class="section" id="span-normalise">
<span id="sec-stats-span-normalise"></span><h3>Span normalise<a class="headerlink" href="#span-normalise" title="Permalink to this headline">¶</a></h3>
<p>In addition to windowing there is an option, <code class="docutils literal notranslate"><span class="pre">span_normalise</span></code> (which defaults to <code class="docutils literal notranslate"><span class="pre">True</span></code>),
All the primary statistics defined here are <em>sums</em> across locations in the genome:
something is computed for each position, and these values are added up across all positions in each window.
Whether the total span of the window is then taken into account is determined by the option <code class="docutils literal notranslate"><span class="pre">span_normalise</span></code>:
if it is <code class="docutils literal notranslate"><span class="pre">True</span></code> (the default), the sum for each window is converted into an <em>average</em>,
by dividing by the window’s <em>span</em> (i.e. the length of genome that it covers).
Otherwise, the sum itself is returned.
The default is <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>,
because this makes the values comparable across windows of different sizes.
To make this more concrete: <a class="reference internal" href="python-api.html#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pairwise</span> <span class="pre">sequence</span> <span class="pre">divergence</span></code></a>
between two samples with <code class="docutils literal notranslate"><span class="pre">mode=&quot;site&quot;</span></code> is the density of sites that differ between the samples;
this is computed for each window by counting up the number of sites
at which the two differ, and dividing by the total span of the window.
If we wanted the number of sites at which the two differed in each window,
we’d calculate divergence with <code class="docutils literal notranslate"><span class="pre">span_normalise=False</span></code>.</p>
<p>Following on from above, suppose we computed the statistic <code class="docutils literal notranslate"><span class="pre">S</span></code> with
<code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">[0,</span> <span class="pre">a,</span> <span class="pre">b]</span></code> and <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>,
and then computed <code class="docutils literal notranslate"><span class="pre">T</span></code> in just the same way except with <code class="docutils literal notranslate"><span class="pre">span_normalize=False</span></code>.
Then <code class="docutils literal notranslate"><span class="pre">S[0]</span></code> would be equal to <code class="docutils literal notranslate"><span class="pre">T[0]</span> <span class="pre">/</span> <span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">S[1]</span> <span class="pre">=</span> <span class="pre">T[1]</span> <span class="pre">/</span> <span class="pre">(b</span> <span class="pre">-</span> <span class="pre">a)</span></code>.
Furthermore, the value obtained with <code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">[0,</span> <span class="pre">b]</span></code> would be equal to <code class="docutils literal notranslate"><span class="pre">T[0]</span> <span class="pre">+</span> <span class="pre">T[1]</span></code>.
However, you probably usually want the (default) normalized version:
don’t get unnormalised values unless you’re sure that’s what you want.
The exception is when computing a site statistic with <code class="docutils literal notranslate"><span class="pre">windows</span> <span class="pre">=</span> <span class="pre">&quot;sites&quot;</span></code>:
this case, computes a statistic with the pattern of genotypes at each site,
and normalising would divide these statistics by the distance to the previous variant site
(probably not what you want to do).</p>
<p>And, a final note about “span”: in tree sequences produced by <code class="docutils literal notranslate"><span class="pre">msprime</span></code>
coordinates along the sequence are <strong>continuous</strong>,
so the “spans” used here may not correspond to distance along the genome in (say) base pairs.
For instance, pairwise sequence divergence is usually a number between 0 and 1
because it is the proportion of bases that differ;
this will only be true if the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>, and hence the “spans” are measured in base pairs
(which you ensure in <code class="docutils literal notranslate"><span class="pre">msprime</span></code> by setting recombination and mutation rates equal to the values
in units of crossovers and mutations per base pair, respectively).</p>
</div>
<div class="section" id="output-format">
<span id="sec-stats-output-format"></span><h3>Output format<a class="headerlink" href="#output-format" title="Permalink to this headline">¶</a></h3>
<p>Each of the statistics methods returns a <code class="docutils literal notranslate"><span class="pre">numpy</span></code> ndarray.
Suppose that the output is named <code class="docutils literal notranslate"><span class="pre">out</span></code>.
If <code class="docutils literal notranslate"><span class="pre">windows</span></code> has been specified, the number of rows of the output is equal to the
number of windows, so that <code class="docutils literal notranslate"><span class="pre">out.shape[0]</span></code> is equal to <code class="docutils literal notranslate"><span class="pre">len(windows)</span> <span class="pre">-</span> <span class="pre">1</span></code>
and <code class="docutils literal notranslate"><span class="pre">out[i]</span></code> is an array of statistics describing the portion of the tree sequence
from <code class="docutils literal notranslate"><span class="pre">windows[i]</span></code> to <code class="docutils literal notranslate"><span class="pre">windows[i</span> <span class="pre">+</span> <span class="pre">1]</span></code> (including the left but not the right endpoint).
What is returned within each window depends on the <a class="reference internal" href="#sec-stats-mode"><span class="std std-ref">mode</span></a>:</p>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">mode=&quot;site&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">mode=&quot;branch&quot;</span></code></dt><dd><p>The output is a two-dimensional array,
with columns corresponding to the different statistics computed: <code class="docutils literal notranslate"><span class="pre">out[i,</span> <span class="pre">j]</span></code> is the <code class="docutils literal notranslate"><span class="pre">j</span></code>-th statistic
in the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th window.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">mode=&quot;node&quot;</span></code></dt><dd><p>The output is a three-dimensional array,
with the second dimension corresponding to node id.
In other words, <code class="docutils literal notranslate"><span class="pre">out.shape[1]</span></code> is equal to <code class="docutils literal notranslate"><span class="pre">ts.num_nodes</span></code>,
and <code class="docutils literal notranslate"><span class="pre">out[i,j]</span></code> is an array of statistics computed for node <code class="docutils literal notranslate"><span class="pre">j</span></code> on the <code class="docutils literal notranslate"><span class="pre">i</span></code>-th window.</p>
</dd>
</dl>
<p>The final dimension of the arrays in other cases is specified by the method.</p>
<p>Note, however, that empty dimensions can optionally be dropped,
as described in the <a class="reference internal" href="#sec-stats-output-dimensions"><span class="std std-ref">Output dimensions</span></a> section.</p>
<p>A note about <strong>default values</strong> and <strong>division by zero</strong>:
Under the hood, statistics computation fills in zeros everywhere, then updates these
(since statistics are all <strong>additive</strong>, this makes sense).
But now suppose that you’ve got a statistic that returns <code class="docutils literal notranslate"><span class="pre">nan</span></code>
(“not a number”) sometimes, like if you’re taking the diversity of a sample set with only <code class="docutils literal notranslate"><span class="pre">n=1</span></code> sample,
which involves dividing by <code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">*</span> <span class="pre">(n</span> <span class="pre">-</span> <span class="pre">1)</span></code>.
Usually, you’ll just get <code class="docutils literal notranslate"><span class="pre">nan</span></code> everywhere that the division by zero happens.
But there’s a couple of caveats.
For <code class="docutils literal notranslate"><span class="pre">site</span></code> statistics, any windows without any sites in them never get touched,
so they will have a value of 0.
For <code class="docutils literal notranslate"><span class="pre">branch</span></code> statistics, any windows with <strong>no branches</strong> will similarly remain 0.
That said, you should <strong>not</strong> rely on the specific behavior of whether <code class="docutils literal notranslate"><span class="pre">0</span></code> or <code class="docutils literal notranslate"><span class="pre">nan</span></code> is returned
for “empty” cases like these: it is subject to change.</p>
</div>
<div class="section" id="output-dimensions">
<span id="sec-stats-output-dimensions"></span><h3>Output dimensions<a class="headerlink" href="#output-dimensions" title="Permalink to this headline">¶</a></h3>
<p>In the general case, tskit outputs two dimensional (or three dimensional, in the case of node
stats) numpy arrays, as described in the <a class="reference internal" href="#sec-stats-output-format"><span class="std std-ref">Output format</span></a> section.
The first dimension corresponds to the window along the genome
such that for some result array <code class="docutils literal notranslate"><span class="pre">x</span></code>, <code class="docutils literal notranslate"><span class="pre">x[j]</span></code> contains information about the jth window.
The last dimension corresponds to the statistics being computed, so that <code class="docutils literal notranslate"><span class="pre">x[j,</span> <span class="pre">k]</span></code> is the
value of the kth statistic in the jth window (in the two dimensional case). This is
a powerful and general interface, but in many cases we will not use this full generality
and the extra dimensions in the numpy arrays are inconvenient.</p>
<p>Tskit optionally removes empty dimensions from the output arrays following a few
simple rules.</p>
<ol class="arabic simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">windows</span></code> is None we are computing over the single window covering the
full sequence. We therefore drop the first dimension of the array.</p></li>
<li><p>In one-way stats, if the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is a 1D array we interpret
this as specifying a single sample set (and therefore a single statistic), and
drop the last dimension of the output array. If <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> is None
(the default), we use the sample set <code class="docutils literal notranslate"><span class="pre">ts.samples()</span></code>, invoking
this rule (we therefore drop the last dimension by default).</p></li>
<li><p>In k-way stats, if the <code class="docutils literal notranslate"><span class="pre">indexes</span></code> argument is a 1D array of length k
we intepret this as specifying a single statistic and drop the last
dimension of the array. If <code class="docutils literal notranslate"><span class="pre">indexes</span></code> is None (the default) and
there are k sample sets, we compute the statistic from these sample sets
and drop the last dimension.</p></li>
</ol>
<p>Rules 2 and 3 can be summarised by “the dimensions of the input determines
the dimensions of the output”. Note that dropping these dimensions is
<strong>optional</strong>: it is always possible to keep the full dimensions of the
output arrays.</p>
<p>Please see the <a class="reference internal" href="tutorial.html#sec-tutorial-stats"><span class="std std-ref">tutorial</span></a> for examples of the
various output dimension options.</p>
</div>
</div>
<div class="section" id="general-api">
<span id="sec-stats-general-api"></span><h2>General API<a class="headerlink" href="#general-api" title="Permalink to this headline">¶</a></h2>
<p>The methods <a class="reference internal" href="python-api.html#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.general_stat()</span></code></a> and <a class="reference internal" href="python-api.html#tskit.TreeSequence.sample_count_stat" title="tskit.TreeSequence.sample_count_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.sample_count_stat()</span></code></a>
provide access to the general-purpose algorithm for computing statistics.
Here is a bit more discussion of how to use these.</p>
<div class="section" id="polarisation">
<span id="sec-stats-polarisation"></span><h3>Polarisation<a class="headerlink" href="#polarisation" title="Permalink to this headline">¶</a></h3>
<p>Many statistics calculated from genome sequence treat all alleles on equal footing,
as one must without knowledge of the ancestral state and sequence of mutations that produced the data.
Separating out the <em>ancestral</em> allele (e.g., as inferred using an outgroup)
is known as <em>polarisiation</em>.
For instance, in the allele frequency spectrum, a site with alleles at 20% and 80% frequency
is no different than another whose alleles are at 80% and 20%,
unless we know in each case which allele is ancestral,
and so while the unpolarised allele frequency spectrum gives the distribution of frequencies of <em>all</em> alleles,
the <em>polarised</em> allele frequency spectrum gives the distribution of frequencies of only <em>derived</em> alleles.</p>
<p>This concept is extended to more general statistics as follows.
For site statistics, summary functions are applied to the total weight or number of samples
associated with each allele; but if polarised, then the ancestral allele is left out of this sum.
For branch or node statistics, summary functions are applied to the total weight or number of samples
below, and above each branch or node; if polarised, then only the weight below is used.</p>
</div>
<div class="section" id="summary-functions">
<span id="sec-stats-summary-functions"></span><h3>Summary functions<a class="headerlink" href="#summary-functions" title="Permalink to this headline">¶</a></h3>
<p>For convenience, here are the summary functions used for many of the statistics.
Below, <span class="math notranslate nohighlight">\(x\)</span> denotes the number of samples in a sample set below a node,
<cite>n</cite> denotes the total size of a sample set, <span class="math notranslate nohighlight">\(p = x / n\)</span>,
and boolean expressions (e.g., <span class="math notranslate nohighlight">\((x &gt; 0)\)</span>) are interpreted as 0/1.</p>
<dl>
<dt><code class="docutils literal notranslate"><span class="pre">diversity</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x) = \frac{x (n - x)}{n (n-1)}\)</span></p>
<p>For an unpolarized statistic with biallelic loci, this calculates
<span class="math notranslate nohighlight">\(2 p (1-p)\)</span>.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">segregating_sites</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x) =  (x &gt; 0) (1 - x / n)\)</span></p>
<p>(Note: this works because if <span class="math notranslate nohighlight">\(\sum_i p_1 = 1\)</span> then <span class="math notranslate nohighlight">\(\sum_{i=1}^k (1-p_i) = k-1\)</span>.)</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">Y1</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x) = \frac{x (n - x) (n - x - 1)}{n (n-1) (n-2)}\)</span></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">divergence</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x_1, x_2) = \frac{x_1 (n_2 - x_2)}{n_1 n_2}\)</span>,</p>
<p>unless the two indices are the same, when the diversity function is used.</p>
<p>For an unpolarized statistic with biallelic loci, this calculates
<span class="math notranslate nohighlight">\(p_1 (1-p_2) + (1 - p_1) p_2\)</span>.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">genetic_relatedness</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x_i, x_j) = \frac{1}{2}(x_i - m)(x_j - m)\)</span>,</p>
<p>where <span class="math notranslate nohighlight">\(m = \frac{1}{n}\sum_{k=1}^n x_k\)</span> with <span class="math notranslate nohighlight">\(n\)</span> the total number
of samples.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">Y2</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x_1, x_2) = \frac{x_1 (n_2 - x_2) (n_2 - x_2 - 1)}{n_1 n_2 (n_2 - 1)}\)</span></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">f2</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x_1, x_2) = \frac{x_1 (x_1 - 1) (n_2 - x_2) (n_2 - x_2 - 1)}{n_1 (n_1 - 1) n_2 (n_2 - 1)} - \frac{x_1 (n_1 - x_1) (n_2 - x_2) x_2}{n_1 (n_1 - 1) n_2 (n_2 - 1)}\)</span></p>
<p>For an unpolarized statistic with biallelic loci, this calculates
<span class="math notranslate nohighlight">\(((p_1 - p_2)^2 - (p_1 (1-p_2)^2 + (1-p_1) p_2^2)/n_1 - (p_1^2 (1-p_2) + (1-p_1)^2 p_2)/n_2\)</span>
<span class="math notranslate nohighlight">\(+ (p_1 p_2 + (1-p_1)(1-p_2))/ n_1 n_2)(1 + \frac{1}{n_1 - 1})(1 + \frac{1}{n_2 - 1})\)</span>,
which is the unbiased estimator for <span class="math notranslate nohighlight">\((p_1 - p_2)^2\)</span> from a finite sample.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">Y3</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x_1, x_2, x_3) = \frac{x_1 (n_2 - x_2) (n_3 - x_3)}{n_1 n_2 n_3}\)</span></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">f3</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x_1, x_2, x_3) = \frac{x_1 (x_1 - 1) (n_2 - x_2) (n_3 - x_3)}{n_1 (n_1 - 1) n_2 n_3} - \frac{x_1 (n_1 - x_1) (n_2 - x_2) x_3}{n_1 (n_1 - 1) n_2 n_3}\)</span></p>
<p>For an unpolarized statistic with biallelic loci, this calculates
<span class="math notranslate nohighlight">\(((p_1 - p_2)(p_1 - p_3) - p_1 (1-p_2)(1-p_3)/n_1 - (1-p_1) p_2 p_3/n_1)(1 + \frac{1}{n_1 - 1})\)</span>,
which is the unbiased estimator for <span class="math notranslate nohighlight">\((p_1 - p_2)(p_1 - p_3)\)</span> from a finite sample.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">f4</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(x_1, x_2, x_3, x_4) = \frac{x_1 x_3 (n_2 - x_2) (n_4 - x_4)}{n_1 n_2 n_3 n_4} - \frac{x_1 x_4 (n_2 - x_2) (n_3 - x_3)}{n_1 n_2 n_3 n_4}\)</span></p>
<p>For an unpolarized statistic with biallelic loci, this calculates
<span class="math notranslate nohighlight">\((p_1 - p_2)(p_3 - p_4)\)</span>.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">trait_covariance</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(w) = \frac{w^2}{2 (n-1)^2}\)</span>,</p>
<p>where <span class="math notranslate nohighlight">\(w\)</span> is the sum of all trait values of the samples below the node.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">trait_correlation</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(w, x) = \frac{w^2}{2 x (1 - x/n) (n - 1)}\)</span>,</p>
<p>where as before <span class="math notranslate nohighlight">\(x\)</span> is the total number of samples below the node,
and <span class="math notranslate nohighlight">\(n\)</span> is the total number of samples.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">trait_linear_model</span></code></dt><dd><p><span class="math notranslate nohighlight">\(f(w, z, x) = \frac{1}{2}\left( \frac{w - \sum_{j=1}^k z_j v_j}{x - \sum_{j=1}^k z_j^2} \right)^2\)</span>,</p>
<p>where <span class="math notranslate nohighlight">\(w\)</span> and <span class="math notranslate nohighlight">\(x\)</span> are as before,
<span class="math notranslate nohighlight">\(z_j\)</span> is the sum of the j-th normalised covariate values below the node,
and <span class="math notranslate nohighlight">\(v_j\)</span> is the covariance of the trait with the j-th covariate.</p>
</dd>
</dl>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="c-api.html" class="btn btn-neutral float-right" title="C API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="python-api.html" class="btn btn-neutral float-left" title="Python API" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2018-2020, Tskit developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>