

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tutorial &mdash; tskit 0.3.3 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Changelogs" href="changelogs.html" />
    <link rel="prev" title="Development" href="development.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> tskit
          

          
            
            <img src="_static/tskit_logo_pale.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="python-api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="stats.html">Statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="c-api.html">C API</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="data-model.html">Data model</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="combinatorics.html">Combinatorics</a></li>
<li class="toctree-l1"><a class="reference internal" href="provenance.html">Provenance</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#trees">Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#traversals">Traversals</a></li>
<li class="toctree-l3"><a class="reference internal" href="#traversing-upwards">Traversing upwards</a></li>
<li class="toctree-l3"><a class="reference internal" href="#traversals-with-information">Traversals with information</a></li>
<li class="toctree-l3"><a class="reference internal" href="#traversals-with-networkx">Traversals with networkx</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Traversing upwards</a></li>
<li class="toctree-l3"><a class="reference internal" href="#calculating-distances-to-the-root">Calculating distances to the root</a></li>
<li class="toctree-l3"><a class="reference internal" href="#finding-nearest-neighbors">Finding nearest neighbors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#moving-along-a-tree-sequence">Moving along a tree sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#editing-tree-sequences">Editing tree sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#working-with-tables">Working with Tables</a></li>
<li class="toctree-l2"><a class="reference internal" href="#overview-of-the-tables-format">Overview of the Tables Format</a></li>
<li class="toctree-l2"><a class="reference internal" href="#working-with-metadata">Working with Metadata</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#reading-metadata-and-schemas">Reading metadata and schemas</a></li>
<li class="toctree-l3"><a class="reference internal" href="#modifying-metadata-and-schemas">Modifying metadata and schemas</a></li>
<li class="toctree-l3"><a class="reference internal" href="#viewing-raw-metadata">Viewing raw metadata</a></li>
<li class="toctree-l3"><a class="reference internal" href="#metadata-for-bulk-table-methods">Metadata for bulk table methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="#binary-metadata">Binary metadata</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#calculating-ld">Calculating LD</a></li>
<li class="toctree-l2"><a class="reference internal" href="#working-with-threads">Working with threads</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parsimony">Parsimony</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#building-tables">Building tables</a></li>
<li class="toctree-l3"><a class="reference internal" href="#missing-data">Missing data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#computing-statistics">Computing statistics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#one-way-statistics">One-way statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multi-way-statistics">Multi-way statistics</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#allele-frequency-spectra">Allele frequency spectra</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#joint-spectra">Joint spectra</a></li>
<li class="toctree-l3"><a class="reference internal" href="#branch-length-spectra">Branch length spectra</a></li>
<li class="toctree-l3"><a class="reference internal" href="#zeroth-and-final-entries-in-the-afs">Zeroth and final entries in the AFS</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">tskit</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/tskit-dev/tskit/blob/main/docs/tutorial.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tutorial">
<span id="sec-tutorial"></span><h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>The content here has been ported from the msprime tutorial and
needs to be reorganised to make a coherent narrative.</p>
</div>
<div class="section" id="trees">
<span id="sec-tutorial-trees"></span><h2>Trees<a class="headerlink" href="#trees" title="Permalink to this headline">¶</a></h2>
<p>A <a class="reference internal" href="python-api.html#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> represents a single tree in a <a class="reference internal" href="python-api.html#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.
The <code class="docutils literal notranslate"><span class="pre">tskit</span></code> Tree implementation differs from most tree libraries by
using <strong>integer IDs</strong> to refer to nodes rather than objects. Thus, when we wish to
find the parent of the node with ID ‘0’, we use <code class="docutils literal notranslate"><span class="pre">tree.parent(0)</span></code>, which
returns another integer. If ‘0’ does not have a parent in the current tree
(e.g., if it is a root), then the special value <a class="reference internal" href="python-api.html#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>
(<span class="math notranslate nohighlight">\(-1\)</span>) is returned. The children of a node are found using the
<a class="reference internal" href="python-api.html#tskit.Tree.children" title="tskit.Tree.children"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.children()</span></code></a> method. To obtain information about a particular node,
one may either use <code class="docutils literal notranslate"><span class="pre">tree.tree_sequence.node(u)</span></code> to which returns the
corresponding <a class="reference internal" href="python-api.html#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a> instance, or use the <a class="reference internal" href="python-api.html#tskit.Tree.time" title="tskit.Tree.time"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.time()</span></code></a> or
<a class="reference internal" href="python-api.html#tskit.Tree.population" title="tskit.Tree.population"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.population()</span></code></a> shorthands.</p>
<div class="section" id="traversals">
<span id="sec-tutorial-trees-traversals"></span><h3>Traversals<a class="headerlink" href="#traversals" title="Permalink to this headline">¶</a></h3>
<p>Tree traversals in various orders are possible using the <a class="reference internal" href="python-api.html#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a> iterator.
For example, in the following tree we can visit the nodes in different orders:</p>
<a class="reference internal image-reference" href="_images/tree_structure1.svg"><img alt="An example tree" src="_images/tree_structure1.svg" width="200px" /></a>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">order</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;preorder&quot;</span><span class="p">,</span> <span class="s2">&quot;inorder&quot;</span><span class="p">,</span> <span class="s2">&quot;postorder&quot;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">order</span><span class="si">}</span><span class="s2">:</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)))</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">preorder</span><span class="p">:</span>        <span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
<span class="n">inorder</span><span class="p">:</span>         <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
<span class="n">postorder</span><span class="p">:</span>       <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
</pre></div>
</div>
<p>Much of the time, the specific ordering of the nodes is not important
and we can leave it out (defaulting to preorder traversal). For example,
here we compute the total branch length of a tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">total_branch_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">branch_length</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span>
</pre></div>
</div>
<p>This is also available as the <a class="reference internal" href="python-api.html#tskit.Tree.total_branch_length" title="tskit.Tree.total_branch_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tskit.Tree.total_branch_length</span></code></a> attribute.</p>
</div>
<div class="section" id="traversing-upwards">
<h3>Traversing upwards<a class="headerlink" href="#traversing-upwards" title="Permalink to this headline">¶</a></h3>
<p>For many applications it is useful to be able to traverse upwards from the
leaves. We can do this using the <a class="reference internal" href="python-api.html#tskit.Tree.parent" title="tskit.Tree.parent"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.parent()</span></code></a> method, which
returns the parent of a node. For example, we can traverse upwards from
each of the samples in the tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">samples</span><span class="p">():</span>
    <span class="n">path</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">v</span> <span class="o">=</span> <span class="n">u</span>
    <span class="k">while</span> <span class="n">v</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
        <span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
        <span class="n">v</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="s2">&quot;-&gt;&quot;</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">1</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">2</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">3</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">4</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="traversals-with-information">
<h3>Traversals with information<a class="headerlink" href="#traversals-with-information" title="Permalink to this headline">¶</a></h3>
<p>Sometimes we will need to traverse down the tree while maintaining
some information about the nodes that are above it. While this
can be done using recursive algorithms, it is often more convenient
and efficient to use an iterative approach. Here, for example,
we define an iterator that yields all nodes in preorder along with
their path length to root:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">preorder_dist</span><span class="p">(</span><span class="n">tree</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">root</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">roots</span><span class="p">:</span>
        <span class="n">stack</span> <span class="o">=</span> <span class="p">[(</span><span class="n">root</span><span class="p">,</span> <span class="mi">0</span><span class="p">)]</span>
        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">stack</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">u</span><span class="p">,</span> <span class="n">distance</span> <span class="o">=</span> <span class="n">stack</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="k">yield</span> <span class="n">u</span><span class="p">,</span> <span class="n">distance</span>
            <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">u</span><span class="p">):</span>
                <span class="n">stack</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">v</span><span class="p">,</span> <span class="n">distance</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">preorder_dist</span><span class="p">(</span><span class="n">tree</span><span class="p">)))</span>
</pre></div>
</div>
<p>Running this on the example above gives us:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="section" id="traversals-with-networkx">
<span id="sec-tutorial-networkx"></span><h3>Traversals with networkx<a class="headerlink" href="#traversals-with-networkx" title="Permalink to this headline">¶</a></h3>
<p>Traversals and other network analysis can also be performed using the sizeable
<a class="reference external" href="https://networkx.github.io/documentation/stable/index.html">networkx</a>
library. This can be achieved by calling <a class="reference internal" href="python-api.html#tskit.Tree.as_dict_of_dicts" title="tskit.Tree.as_dict_of_dicts"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.as_dict_of_dicts()</span></code></a> to
convert a <a class="reference internal" href="python-api.html#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance to a format that can be imported by networkx to
create a graph:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>

<span class="n">g</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">edges</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">6</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="section" id="id2">
<h3>Traversing upwards<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>We can revisit the above examples and traverse upwards with
networkx using a depth-first search algorithm:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>

<span class="n">g</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
<span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">samples</span><span class="p">():</span>
    <span class="n">path</span> <span class="o">=</span> <span class="p">[</span><span class="n">u</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">parent</span> <span class="k">for</span> <span class="n">parent</span><span class="p">,</span> <span class="n">child</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span>
                  <span class="n">nx</span><span class="o">.</span><span class="n">edge_dfs</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">u</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s2">&quot;reverse&quot;</span><span class="p">)]</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="s2">&quot;-&gt;&quot;</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">1</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">2</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">3</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="mi">4</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="calculating-distances-to-the-root">
<h3>Calculating distances to the root<a class="headerlink" href="#calculating-distances-to-the-root" title="Permalink to this headline">¶</a></h3>
<p>Similarly, we can yield the nodes of a tree along with their distance to the
root in pre-order in networkx as well:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>

<span class="n">g</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
<span class="k">for</span> <span class="n">root</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">roots</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">shortest_path_length</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">root</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
</pre></div>
</div>
<p>Running this on the example above gives us the same result as before:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="section" id="finding-nearest-neighbors">
<h3>Finding nearest neighbors<a class="headerlink" href="#finding-nearest-neighbors" title="Permalink to this headline">¶</a></h3>
<p>If some samples in a tree are not at time 0, then finding the nearest neighbor
of a sample is a bit more involved. Instead of writing our own traversal code
we can again draw on a networkx algorithm.
Let us start with an example tree with three samples that were sampled at
different time points:</p>
<a class="reference internal image-reference" href="_images/different_time_samples.svg"><img alt="An example tree with samples taken at different times" src="_images/different_time_samples.svg" width="200px" /></a>
<p>The generation times for these nodes are:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">u</span><span class="p">))</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">4</span> <span class="mf">20.005398778263334</span>
<span class="mi">2</span> <span class="mf">20.0</span>
<span class="mi">3</span> <span class="mf">17.833492457579652</span>
<span class="mi">0</span> <span class="mf">0.0</span>
<span class="mi">1</span> <span class="mf">1.0</span>
</pre></div>
</div>
<p>Note that samples 0 and 1 are about 35 generations apart from each other even though
they were sampled at almost the same time. This is why samples 0 and 1 are
closer to sample 2 than to each other.</p>
<p>For this nearest neighbor search we will be traversing up and down the tree,
so it is easier to treat the tree as an undirected graph:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">g</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
</pre></div>
</div>
<p>When converting the tree to a networkx graph the edges are annotated with their
branch length:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;branch_length&#39;</span><span class="p">:</span> <span class="mf">0.005398778263334236</span><span class="p">}),</span>
 <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;branch_length&#39;</span><span class="p">:</span> <span class="mf">2.171906320683682</span><span class="p">}),</span>
 <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;branch_length&#39;</span><span class="p">:</span> <span class="mf">17.833492457579652</span><span class="p">}),</span>
 <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;branch_length&#39;</span><span class="p">:</span> <span class="mf">16.833492457579652</span><span class="p">})]</span>
</pre></div>
</div>
<p>We can now use the “branch_length” field as a weight for a weighted shortest path
search:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># a dictionary of dictionaries to represent our distance matrix</span>
<span class="n">dist_dod</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
<span class="k">for</span> <span class="n">source</span><span class="p">,</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">samples</span><span class="p">(),</span> <span class="mi">2</span><span class="p">):</span>
    <span class="n">dist_dod</span><span class="p">[</span><span class="n">source</span><span class="p">][</span><span class="n">target</span><span class="p">]</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">shortest_path_length</span><span class="p">(</span>
        <span class="n">g</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">,</span> <span class="n">target</span><span class="o">=</span><span class="n">target</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;branch_length&quot;</span>
    <span class="p">)</span>
    <span class="n">dist_dod</span><span class="p">[</span><span class="n">target</span><span class="p">][</span><span class="n">source</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist_dod</span><span class="p">[</span><span class="n">source</span><span class="p">][</span><span class="n">target</span><span class="p">]</span>

<span class="c1"># extract the nearest neighbor of nodes 0, 1, and 2</span>
<span class="n">nearest_neighbor_of</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">dist_dod</span><span class="p">[</span><span class="n">u</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="n">dist_dod</span><span class="p">[</span><span class="n">u</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">)]</span>

<span class="nb">print</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span> <span class="n">nearest_neighbor_of</span><span class="p">)))</span>
</pre></div>
</div>
<p>gives:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="moving-along-a-tree-sequence">
<span id="sec-tutorial-moving-along-a-tree-sequence"></span><h2>Moving along a tree sequence<a class="headerlink" href="#moving-along-a-tree-sequence" title="Permalink to this headline">¶</a></h2>
<p>Most of the time we will want to iterate over all the trees in a tree sequence
sequentially as efficiently as possible. The simplest way to do this is to
use the <a class="reference internal" href="python-api.html#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Tree sequence has </span><span class="si">{}</span><span class="s2"> trees&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_trees</span><span class="p">))</span>
<span class="nb">print</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="s2">&quot;Tree </span><span class="si">{}</span><span class="s2"> covers [</span><span class="si">{:.2f}</span><span class="s2">, </span><span class="si">{:.2f}</span><span class="s2">); TMRCA = </span><span class="si">{:.4f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="o">*</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">root</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
</pre></div>
</div>
<p>Running the code, we get:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span> <span class="n">sequence</span> <span class="n">has</span> <span class="mi">7</span> <span class="n">trees</span>

<span class="n">Tree</span> <span class="mi">0</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.08</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">4.2542</span>
<span class="n">Tree</span> <span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.08</span><span class="p">,</span> <span class="mf">0.27</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
<span class="n">Tree</span> <span class="mi">2</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.27</span><span class="p">,</span> <span class="mf">0.37</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">4.2542</span>
<span class="n">Tree</span> <span class="mi">3</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.37</span><span class="p">,</span> <span class="mf">0.66</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
<span class="n">Tree</span> <span class="mi">4</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.66</span><span class="p">,</span> <span class="mf">0.71</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">4.2542</span>
<span class="n">Tree</span> <span class="mi">5</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.71</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
<span class="n">Tree</span> <span class="mi">6</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">1.00</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
</pre></div>
</div>
<p>Here we run a small simulation using <a class="reference external" href="https://msprime.readthedocs.io">msprime</a>
which results in 7 distinct trees along a genome of length 1. We then iterate
over these trees sequentially using the <a class="reference internal" href="python-api.html#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method,
and print out each tree’s index, the interval over which the tree applies
and the time of the most recent common ancestor of all the samples. This
method is very efficient, and allows us to quickly iterate over very large
tree sequences.</p>
<p>We can also efficiently iterate over the trees backwards, using Python’s
<a class="reference external" href="https://docs.python.org/3/library/functions.html#reversed" title="(in Python v3.9)"><code class="xref py py-func docutils literal notranslate"><span class="pre">reversed()</span></code></a> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="nb">reversed</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="s2">&quot;Tree </span><span class="si">{}</span><span class="s2"> covers [</span><span class="si">{:.2f}</span><span class="s2">, </span><span class="si">{:.2f}</span><span class="s2">); TMRCA = </span><span class="si">{:.4f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="o">*</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">root</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span> <span class="mi">6</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">1.00</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
<span class="n">Tree</span> <span class="mi">5</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.71</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
<span class="n">Tree</span> <span class="mi">4</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.66</span><span class="p">,</span> <span class="mf">0.71</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">4.2542</span>
<span class="n">Tree</span> <span class="mi">3</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.37</span><span class="p">,</span> <span class="mf">0.66</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
<span class="n">Tree</span> <span class="mi">2</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.27</span><span class="p">,</span> <span class="mf">0.37</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">4.2542</span>
<span class="n">Tree</span> <span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.08</span><span class="p">,</span> <span class="mf">0.27</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">2.5973</span>
<span class="n">Tree</span> <span class="mi">0</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.08</span><span class="p">);</span> <span class="n">TMRCA</span> <span class="o">=</span> <span class="mf">4.2542</span>
</pre></div>
</div>
<p>One of the reasons that the <code class="docutils literal notranslate"><span class="pre">trees</span></code> iterator allows us to access
the trees in a tree sequence so efficiently is because we use the
same underlying instance of the <code class="docutils literal notranslate"><span class="pre">.Tree</span></code> class each time. That is,
each time the iterator returns a value, it is actually the same tree
instance each time which has been updated internally to reflect the
(usually small) changes in the tree along the sequence. As a
result of this, if we store the results of the tree iterator in a
list, we will get unexpected results:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="s2">&quot;Tree </span><span class="si">{}</span><span class="s2"> covers [</span><span class="si">{:.2f}</span><span class="s2">, </span><span class="si">{:.2f}</span><span class="s2">): id=</span><span class="si">{:x}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="o">*</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="p">,</span> <span class="nb">id</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span> <span class="o">-</span><span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f290becb3c8</span>
<span class="n">Tree</span> <span class="o">-</span><span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f290becb3c8</span>
<span class="n">Tree</span> <span class="o">-</span><span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f290becb3c8</span>
<span class="n">Tree</span> <span class="o">-</span><span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f290becb3c8</span>
<span class="n">Tree</span> <span class="o">-</span><span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f290becb3c8</span>
<span class="n">Tree</span> <span class="o">-</span><span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f290becb3c8</span>
<span class="n">Tree</span> <span class="o">-</span><span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f290becb3c8</span>
</pre></div>
</div>
<p>We have stored seven copies of the same <a class="reference internal" href="python-api.html#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance in the
list. Because iteration has ended, this tree is in the “null” state (see
below for more details) which means that it doesn’t represent any of the
trees in the tree sequence.</p>
<p>If we do wish to obtain a list of the trees, we can do so by using the
<a class="reference internal" href="python-api.html#tskit.TreeSequence.aslist" title="tskit.TreeSequence.aslist"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.aslist()</span></code></a> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">aslist</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="s2">&quot;Tree </span><span class="si">{}</span><span class="s2"> covers [</span><span class="si">{:.2f}</span><span class="s2">, </span><span class="si">{:.2f}</span><span class="s2">): id=</span><span class="si">{:x}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="o">*</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="p">,</span> <span class="nb">id</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span> <span class="mi">0</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.08</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">fd2c50a40f0</span>
<span class="n">Tree</span> <span class="mi">1</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.08</span><span class="p">,</span> <span class="mf">0.27</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">fd2b2aca6d8</span>
<span class="n">Tree</span> <span class="mi">2</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.27</span><span class="p">,</span> <span class="mf">0.37</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">fd2b2adde10</span>
<span class="n">Tree</span> <span class="mi">3</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.37</span><span class="p">,</span> <span class="mf">0.66</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">fd2b2adddd8</span>
<span class="n">Tree</span> <span class="mi">4</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.66</span><span class="p">,</span> <span class="mf">0.71</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">fd2b2addd68</span>
<span class="n">Tree</span> <span class="mi">5</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.71</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">fd2b2addcf8</span>
<span class="n">Tree</span> <span class="mi">6</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">1.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">fd2b2addeb8</span>
</pre></div>
</div>
<p>Note that we now have a different object for each tree in the list. Please
note that this is <strong>much</strong> less efficient than iterating over the trees
using the <a class="reference internal" href="python-api.html#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method (and uses far more memory!),
and should only be used as a convenience when working with small trees.</p>
<p>We can also obtain specific trees along the sequence, using the
<a class="reference internal" href="python-api.html#tskit.TreeSequence.first" title="tskit.TreeSequence.first"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.first()</span></code></a>,
<a class="reference internal" href="python-api.html#tskit.TreeSequence.last" title="tskit.TreeSequence.last"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.last()</span></code></a>
<a class="reference internal" href="python-api.html#tskit.TreeSequence.at" title="tskit.TreeSequence.at"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.at()</span></code></a> and
<a class="reference internal" href="python-api.html#tskit.TreeSequence.at_index" title="tskit.TreeSequence.at_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.at_index()</span></code></a> methods. The <code class="docutils literal notranslate"><span class="pre">first()</span></code> and <code class="docutils literal notranslate"><span class="pre">last()</span></code>
methods return the first and last trees in the sequence, as might be
imagined. The <code class="docutils literal notranslate"><span class="pre">at()</span></code> method returns the tree that covers a
given genomic location, and the <code class="docutils literal notranslate"><span class="pre">at_index()</span></code> method returns the
tree at a given index along the sequence:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">at</span><span class="p">(</span><span class="mf">0.5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Tree </span><span class="si">{}</span><span class="s2"> covers [</span><span class="si">{:.2f}</span><span class="s2">, </span><span class="si">{:.2f}</span><span class="s2">): id=</span><span class="si">{:x}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="o">*</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="p">,</span> <span class="nb">id</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">at_index</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Tree </span><span class="si">{}</span><span class="s2"> covers [</span><span class="si">{:.2f}</span><span class="s2">, </span><span class="si">{:.2f}</span><span class="s2">): id=</span><span class="si">{:x}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="o">*</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="p">,</span> <span class="nb">id</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">at_index</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Tree </span><span class="si">{}</span><span class="s2"> covers [</span><span class="si">{:.2f}</span><span class="s2">, </span><span class="si">{:.2f}</span><span class="s2">): id=</span><span class="si">{:x}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="o">*</span><span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="p">,</span> <span class="nb">id</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span> <span class="mi">3</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.37</span><span class="p">,</span> <span class="mf">0.66</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f9fdb469630</span>
<span class="n">Tree</span> <span class="mi">0</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.00</span><span class="p">,</span> <span class="mf">0.08</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f9fdb46d160</span>
<span class="n">Tree</span> <span class="mi">6</span> <span class="n">covers</span> <span class="p">[</span><span class="mf">0.75</span><span class="p">,</span> <span class="mf">1.00</span><span class="p">):</span> <span class="nb">id</span><span class="o">=</span><span class="mi">7</span><span class="n">f9fdb469630</span>
</pre></div>
</div>
<p>Note that each call to these methods returns a different <a class="reference internal" href="python-api.html#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance
and so it is much, much less efficient to sequentially access trees
by their index values than it is to use the <a class="reference internal" href="python-api.html#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a>
iterator.</p>
</div>
<div class="section" id="editing-tree-sequences">
<h2>Editing tree sequences<a class="headerlink" href="#editing-tree-sequences" title="Permalink to this headline">¶</a></h2>
<p>Sometimes we wish to make some minor modifications to a tree sequence that has
been generated by a simulation. However, tree sequence objects are <strong>immutable</strong>
and so we cannot edit them in place. To modify a tree sequence, we need to
extract the underlying <a class="reference internal" href="data-model.html#sec-table-definitions"><span class="std std-ref">tables</span></a> of information, edit these tables,
and then create a new tree sequence from them.
These tables succinctly store everything we need to know
about a tree sequence, and can be manipulated using the <a class="reference internal" href="python-api.html#sec-tables-api"><span class="std std-ref">Tables and Table Collections</span></a>.
In the following example, we use this approach
to remove all singleton sites from a given tree sequence.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">strip_singletons</span><span class="p">(</span><span class="n">ts</span><span class="p">):</span>
    <span class="n">tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>  <span class="c1"># Only supports infinite sites muts.</span>
            <span class="n">mut</span> <span class="o">=</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">num_samples</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">node</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">site_id</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span>
                    <span class="n">position</span><span class="o">=</span><span class="n">site</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="n">site</span><span class="o">.</span><span class="n">ancestral_state</span>
                <span class="p">)</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span>
                    <span class="n">site</span><span class="o">=</span><span class="n">site_id</span><span class="p">,</span> <span class="n">node</span><span class="o">=</span><span class="n">mut</span><span class="o">.</span><span class="n">node</span><span class="p">,</span> <span class="n">derived_state</span><span class="o">=</span><span class="n">mut</span><span class="o">.</span><span class="n">derived_state</span>
                <span class="p">)</span>
    <span class="k">return</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
</pre></div>
</div>
<p>This function takes a tree sequence containing some infinite sites mutations as
input, and returns a copy in which all singleton sites have been removed.
The approach is very simple: we get a copy of the underlying
table data in a <a class="reference internal" href="python-api.html#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> object, and first clear the
site and mutation tables. We then consider each site in turn,
and if the number of samples with
the mutation is greater than one, we add the site and mutation to our
output tables using <a class="reference internal" href="python-api.html#tskit.SiteTable.add_row" title="tskit.SiteTable.add_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SiteTable.add_row()</span></code></a> and <a class="reference internal" href="python-api.html#tskit.MutationTable.add_row" title="tskit.MutationTable.add_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MutationTable.add_row()</span></code></a>.
(In this case we consider only simple infinite sites mutations,
where we cannot have back or recurrent mutations. These would require a slightly
more involved approach where we keep a map of mutation IDs so that
mutation <code class="docutils literal notranslate"><span class="pre">parent</span></code> values could be computed. We have also omitted the
site and mutation metadata in the interest of simplicity.)</p>
<p>After considering each site, we then create a new tree sequence using
the <a class="reference internal" href="python-api.html#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a> method on our updated tables.
Using this function then, we get:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span>
<span class="go">50</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ts_new</span> <span class="o">=</span> <span class="n">strip_singletons</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ts_new</span><span class="o">.</span><span class="n">num_sites</span>
<span class="go">44</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p>Thus, we have removed 6 singleton sites from the tree sequence.</p>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<p>Add another example here where we use the array oriented API to edit
the nodes and edges of a tree sequence. Perhaps decapitating would be a
good example?</p>
</div>
</div>
<div class="section" id="working-with-tables">
<h2>Working with Tables<a class="headerlink" href="#working-with-tables" title="Permalink to this headline">¶</a></h2>
<p>Tables provide a convenient method for viewing, importing and exporting tree
sequences, and are closely tied to the underlying data structures.
There are eight tables that together define a tree sequence,
although some may be empty,
and together they form a <a class="reference internal" href="python-api.html#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.
The tables are defined in <a class="reference internal" href="data-model.html#sec-table-definitions"><span class="std std-ref">Table Definitions</span></a>,
and the <a class="reference internal" href="python-api.html#sec-tables-api"><span class="std std-ref">Tables API</span></a> section describes how to work with them.
Here we make some general remarks about what you can and cannot do with them.</p>
<p><code class="docutils literal notranslate"><span class="pre">tskit</span></code> provides direct access to the columns of each table as
<code class="docutils literal notranslate"><span class="pre">numpy</span></code> arrays: for instance, if <code class="docutils literal notranslate"><span class="pre">n</span></code> is a <code class="docutils literal notranslate"><span class="pre">NodeTable</span></code>, then <code class="docutils literal notranslate"><span class="pre">n.time</span></code>
will return an array containing the birth times of the individuals whose genomes
are represented by the nodes in the table.
<em>However</em>, it is important to note that this is <em>not</em> a shallow copy:
modifying <code class="docutils literal notranslate"><span class="pre">n.time</span></code> will <em>not</em> change the node table <code class="docutils literal notranslate"><span class="pre">n</span></code>.  This may change in
the future, but currently there are three ways to modify tables: <code class="docutils literal notranslate"><span class="pre">.add_row()</span></code>,
<code class="docutils literal notranslate"><span class="pre">.set_columns()</span></code>, and <code class="docutils literal notranslate"><span class="pre">.append_columns()</span></code>
(and also <code class="docutils literal notranslate"><span class="pre">.clear()</span></code>, which empties the table).</p>
<p>For example, a node table could be constructed using <code class="docutils literal notranslate"><span class="pre">.add_row()</span></code> as
follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">n</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NodeTable</span><span class="p">()</span>
<span class="n">sv</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">]</span>
<span class="n">tv</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]</span>
<span class="n">pv</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sv</span><span class="p">,</span> <span class="n">tv</span><span class="p">,</span> <span class="n">pv</span><span class="p">):</span>
    <span class="n">n</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">s</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="n">p</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
<p>obtaining:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
<span class="go">id    flags    population    individual    time    metadata</span>
<span class="go">0    1    0    -1    0.0</span>
<span class="go">1    1    0    -1    0.0</span>
<span class="go">2    1    0    -1    0.0</span>
<span class="go">3    0    0    -1    0.4</span>
<span class="go">4    0    0    -1    0.5</span>
<span class="go">5    0    0    -1    0.7</span>
<span class="go">6    0    0    -1    1.0</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">.add_row()</span></code> method is natural (and should be reasonably efficient) if
new records appear one-by-one. In the example above it would have been more
natural to use <code class="docutils literal notranslate"><span class="pre">.set_columns()</span></code> — equivalently:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">n</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NodeTable</span><span class="p">()</span>
<span class="n">n</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">sv</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="n">pv</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="n">tv</span><span class="p">)</span>
</pre></div>
</div>
<p>Since columns cannot be modified directly as properties of the tables,
they must be extracted, modified, then replaced.
For example, here we add 1.4 to every <code class="docutils literal notranslate"><span class="pre">time</span></code> except the first
in the node table constructed above (using <code class="docutils literal notranslate"><span class="pre">numpy</span></code> indexing):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tn</span> <span class="o">=</span> <span class="n">n</span><span class="o">.</span><span class="n">time</span>
<span class="n">tn</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="n">tn</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">+</span> <span class="mf">1.4</span>
<span class="n">n</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">n</span><span class="o">.</span><span class="n">flags</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="n">n</span><span class="o">.</span><span class="n">population</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="n">tn</span><span class="p">)</span>
</pre></div>
</div>
<p>The result is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
<span class="go">id    flags    population    individual    time    metadata</span>
<span class="go">0    1    0    -1    0.0</span>
<span class="go">1    1    0    -1    1.4</span>
<span class="go">2    1    0    -1    1.4</span>
<span class="go">3    0    0    -1    1.8</span>
<span class="go">4    0    0    -1    1.9</span>
<span class="go">5    0    0    -1    2.1</span>
<span class="go">6    0    0    -1    2.4</span>
</pre></div>
</div>
</div>
<div class="section" id="overview-of-the-tables-format">
<h2>Overview of the Tables Format<a class="headerlink" href="#overview-of-the-tables-format" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="data-model.html#sec-table-definitions"><span class="std std-ref">Table Definitions</span></a> section gives a precise
definition of how a tree sequence is stored in a collection of tables.
Here we give an overview. Consider the following sequence of trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>time ago
--------
   1.0         6
             ┏━┻━━┓
             ┃    ┃
   0.7       ┃    ╋                     5
             ┃    ┃                   ┏━┻━┓
   0.5       ┃    4         4         ┃   4
             ┃  ┏━┻━┓     ┏━┻━┓       ┃ ┏━┻━┓
             ┃  ┃   ┃     ┃   ╋       ┃ ┃   ┃
   0.4       ┃  ┃   ┃     ┃   3       ┃ ┃   ┃
             ┃  ┃   ┃     ┃ ┏━┻━┓     ┃ ┃   ┃
             ┃  ┃   ┃     ┃ ┃   ╋     ┃ ┃   ┃
   0.0       0  1   2     1 0   2     0 1   2

position 0.0          0.2         0.8         1.0
</pre></div>
</div>
<p>Ancestral recombination events have produced three different trees
that relate the three sampled genomes <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, and <code class="docutils literal notranslate"><span class="pre">2</span></code> to each other
along the chromosome of length 1.0.</p>
<p>Each node in each of the above trees represents a particular ancestral genome
(a <em>haploid</em> genome; diploid individuals would be represented by two nodes).
We record when each of nodes lived in a <a class="reference internal" href="python-api.html#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">NodeTable</span><span class="p">:</span>

<span class="nb">id</span>      <span class="n">flags</span>    <span class="n">population</span>   <span class="n">time</span>
<span class="mi">0</span>       <span class="mi">1</span>        <span class="mi">0</span>            <span class="mi">0</span>
<span class="mi">1</span>       <span class="mi">1</span>        <span class="mi">0</span>            <span class="mi">0</span>
<span class="mi">2</span>       <span class="mi">1</span>        <span class="mi">0</span>            <span class="mi">0</span>
<span class="mi">3</span>       <span class="mi">0</span>        <span class="mi">0</span>            <span class="mf">0.4</span>
<span class="mi">4</span>       <span class="mi">0</span>        <span class="mi">0</span>            <span class="mf">0.5</span>
<span class="mi">5</span>       <span class="mi">0</span>        <span class="mi">0</span>            <span class="mf">0.7</span>
<span class="mi">6</span>       <span class="mi">0</span>        <span class="mi">0</span>            <span class="mf">1.0</span>
</pre></div>
</div>
<p>Importantly, the first column, <code class="docutils literal notranslate"><span class="pre">id</span></code>, is not actually recorded, and is
only shown when printing out node tables (as here) for convenience.
The second column, <code class="docutils literal notranslate"><span class="pre">flags</span></code> records a <code class="docutils literal notranslate"><span class="pre">1</span></code> for the individuals that are <em>samples</em>,
i.e., whose entire genealogical history is recorded by these trees.
(Note that the trees above record that node 3 inherited from node 4
on the middle portion of the genome, but not on the ends.)</p>
<p>We next need to record each tree’s edges. Since some edges are present
in more than one tree (e.g., node 1 inherits from node 4 across
the entire sequence), we record in the <a class="reference internal" href="python-api.html#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> each edge
and the genomic region for which it appears in the trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">EdgeTable</span><span class="p">:</span>

<span class="n">left</span>    <span class="n">right</span>   <span class="n">parent</span>  <span class="n">children</span>
<span class="mf">0.2</span>     <span class="mf">0.8</span>     <span class="mi">3</span>       <span class="mi">0</span>
<span class="mf">0.2</span>     <span class="mf">0.8</span>     <span class="mi">3</span>       <span class="mi">2</span>
<span class="mf">0.0</span>     <span class="mf">1.0</span>     <span class="mi">4</span>       <span class="mi">1</span>
<span class="mf">0.0</span>     <span class="mf">0.2</span>     <span class="mi">4</span>       <span class="mi">2</span>
<span class="mf">0.8</span>     <span class="mf">1.0</span>     <span class="mi">4</span>       <span class="mi">2</span>
<span class="mf">0.2</span>     <span class="mf">0.8</span>     <span class="mi">4</span>       <span class="mi">3</span>
<span class="mf">0.8</span>     <span class="mf">1.0</span>     <span class="mi">5</span>       <span class="mi">0</span>
<span class="mf">0.8</span>     <span class="mf">1.0</span>     <span class="mi">5</span>       <span class="mi">4</span>
<span class="mf">0.0</span>     <span class="mf">0.2</span>     <span class="mi">6</span>       <span class="mi">0</span>
<span class="mf">0.0</span>     <span class="mf">0.2</span>     <span class="mi">6</span>       <span class="mi">4</span>
</pre></div>
</div>
<p>Since node 3 is most recent, the edge that says that nodes 0 and 2 inherit
from node 3 on the interval between 0.2 and 0.8 comes first.  Next are the
edges from node 4: there are four of these, as the edge from node 4 to node
1 is shared across the entire sequence, and for each of the three
genomic intervals there is an additional child node. At this
point, we know the full tree on the middle interval.  Finally, edges
specifying the common ancestor of 0 and 4 on the remaining intervals (parents 6
and 5 respectively) allow us to construct all trees across the entire interval.</p>
<p>There are three mutations in the depiction above,
marked by <code class="docutils literal notranslate"><span class="pre">╋</span></code>: one above node <code class="docutils literal notranslate"><span class="pre">4</span></code> on the first tree,
and the other two above nodes <code class="docutils literal notranslate"><span class="pre">2</span></code> and <code class="docutils literal notranslate"><span class="pre">3</span></code> on the second tree.
Suppose that the first mutation occurs at position 0.1 and the mutations in the
second tree both occurred at the same position, at 0.5 (with a back mutation).
To record the inheritance patterns of these, we need only record
the positions on the genome at which they occurred,
and on which edge (equivalently, above which node) they occurred.
The positions are recorded in the <a class="reference internal" href="python-api.html#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SiteTable</span><span class="p">:</span>

<span class="nb">id</span>    <span class="n">position</span>    <span class="n">ancestral_state</span>
<span class="mi">0</span>    <span class="mf">0.1</span>         <span class="mi">0</span>
<span class="mi">1</span>    <span class="mf">0.5</span>         <span class="mi">0</span>
</pre></div>
</div>
<p>As with node tables, the <code class="docutils literal notranslate"><span class="pre">id</span></code> column is <strong>not</strong> actually recorded, but is
implied by the position in the table.  The results of the
actual mutations are then recorded:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MutationTable</span><span class="p">:</span>

<span class="n">site</span>    <span class="n">node</span>    <span class="n">derived_state</span>
<span class="mi">0</span>        <span class="mi">4</span>        <span class="mi">1</span>
<span class="mi">1</span>        <span class="mi">3</span>        <span class="mi">1</span>
<span class="mi">1</span>        <span class="mi">2</span>        <span class="mi">0</span>
</pre></div>
</div>
<p>This would then result in the following (two-locus) haplotypes for the three
samples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>  <span class="n">haplotype</span>
<span class="o">------</span>  <span class="o">---------</span>
<span class="mi">0</span>       <span class="mi">01</span>
<span class="mi">1</span>       <span class="mi">10</span>
<span class="mi">2</span>       <span class="mi">10</span>
</pre></div>
</div>
<p>To create these tables, and the corresponding tree sequence, we would
create a <a class="reference internal" href="python-api.html#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>, and then use its
<a class="reference internal" href="python-api.html#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a> method:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TableCollection</span><span class="p">(</span><span class="n">sequence_length</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>

<span class="c1"># Nodes</span>
<span class="n">sv</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">]</span>
<span class="n">tv</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]</span>

<span class="k">for</span> <span class="n">is_sample</span><span class="p">,</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sv</span><span class="p">,</span> <span class="n">tv</span><span class="p">):</span>
 <span class="n">flags</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NODE_IS_SAMPLE</span> <span class="k">if</span> <span class="n">is_sample</span> <span class="k">else</span> <span class="mi">0</span>
 <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">flags</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="n">t</span><span class="p">)</span>

<span class="c1"># Edges</span>
<span class="n">lv</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">]</span>
<span class="n">rv</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">]</span>
<span class="n">pv</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">]</span>
<span class="n">cv</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>

<span class="k">for</span> <span class="n">l</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">lv</span><span class="p">,</span> <span class="n">rv</span><span class="p">,</span> <span class="n">pv</span><span class="p">,</span> <span class="n">cv</span><span class="p">):</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="n">l</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="n">r</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">p</span><span class="p">,</span> <span class="n">child</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>

<span class="c1"># Sites</span>
<span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;0&#39;</span><span class="p">]):</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="n">p</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="n">a</span><span class="p">)</span>

<span class="c1"># Mutations</span>
<span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;0&#39;</span><span class="p">]):</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">site</span><span class="o">=</span><span class="n">s</span><span class="p">,</span> <span class="n">node</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">derived_state</span><span class="o">=</span><span class="n">d</span><span class="p">)</span>
</pre></div>
</div>
<p>We can then finally obtain the tree sequence:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
  <span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s1">&#39;unicode&#39;</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>  6
┏━┻┓
┃  4
┃ ┏┻┓
0 1 2

  4
┏━┻┓
┃  3
┃ ┏┻┓
1 0 2

  5
┏━┻┓
┃  4
┃ ┏┻┓
0 1 2
</pre></div>
</div>
</div>
<div class="section" id="working-with-metadata">
<span id="sec-tutorial-metadata"></span><h2>Working with Metadata<a class="headerlink" href="#working-with-metadata" title="Permalink to this headline">¶</a></h2>
<p>Metadata is information associated with entities that tskit doesn’t use or interpret,
but which is useful to pass on to downstream analysis such as sample ids, dates etc. See
<a class="reference internal" href="metadata.html#sec-metadata"><span class="std std-ref">Metadata</span></a> for a full discussion. Each table has a
<a class="reference internal" href="python-api.html#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> accessed as <code class="docutils literal notranslate"><span class="pre">table.metadata_schema</span></code>. which details the
contents and encoding of the metadata for each row. A metadata schema is a JSON document
that conforms to <a class="reference external" href="https://json-schema.org/understanding-json-schema/">JSON Schema</a>
(The full schema for tskit is at <a class="reference internal" href="metadata.html#sec-metadata-schema-schema"><span class="std std-ref">Full metaschema</span></a>).</p>
<div class="section" id="reading-metadata-and-schemas">
<h3>Reading metadata and schemas<a class="headerlink" href="#reading-metadata-and-schemas" title="Permalink to this headline">¶</a></h3>
<p>Metadata is automatically decoded using the schema when accessed via a
<a class="reference internal" href="python-api.html#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> or <a class="reference internal" href="python-api.html#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> Python API. For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ts</span><span class="o">.</span><span class="n">individual</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;accession&#39;: &#39;ERS0001&#39;, &#39;pcr&#39;: True}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;accession&#39;: &#39;ERS0001&#39;, &#39;pcr&#39;: True}</span>
</pre></div>
</div>
<p>Viewing the <a class="reference internal" href="python-api.html#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for a table can help with understanding
its metadata as it can contain descriptions and constraints:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pprint</span><span class="o">.</span><span class="n">pprint</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">table_metadata_schemas</span><span class="o">.</span><span class="n">individual</span><span class="p">)</span>
<span class="go">{&#39;additionalProperties&#39;: False,</span>
<span class="go"> &#39;codec&#39;: &#39;json&#39;,</span>
<span class="go"> &#39;properties&#39;: {&#39;accession&#39;: {&#39;description&#39;: &#39;ENA accession number&#39;,</span>
<span class="go">                              &#39;type&#39;: &#39;string&#39;},</span>
<span class="go">                &#39;pcr&#39;: {&#39;description&#39;: &#39;Was PCR used on this sample&#39;,</span>
<span class="go">                        &#39;name&#39;: &#39;PCR Used&#39;,</span>
<span class="go">                        &#39;type&#39;: &#39;boolean&#39;}},</span>
<span class="go"> &#39;required&#39;: [&#39;accession&#39;, &#39;pcr&#39;],</span>
<span class="go"> &#39;type&#39;: &#39;object&#39;}</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">pprint</span><span class="o">.</span><span class="n">pprint</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">metadata_schema</span><span class="p">)</span>
<span class="go">(.... as above)</span>
</pre></div>
</div>
<p>The top-level metadata and schemas for the entire tree sequence is similarly
accessed with <code class="docutils literal notranslate"><span class="pre">ts.metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">ts.metadata_schema</span></code>.</p>
<p>If there is no schema (i.e. it is equal to <code class="docutils literal notranslate"><span class="pre">MetadataSchema(None)</span></code>) for a table
or top-level metadata, then no decoding is performed and <code class="docutils literal notranslate"><span class="pre">bytes</span></code> will be returned.</p>
</div>
<div class="section" id="modifying-metadata-and-schemas">
<h3>Modifying metadata and schemas<a class="headerlink" href="#modifying-metadata-and-schemas" title="Permalink to this headline">¶</a></h3>
<p>If you are creating or modifying a tree sequence you may want to record or add to the
metadata. When doing so the schema must be modified first, as it is then used to
validate and encode the metadata.</p>
<p>Schemas in tskit are held in a <a class="reference internal" href="python-api.html#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a>.
A Python dict representation of the schema is passed to its constructor, which
will validate the schema:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">schema</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">({</span><span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;json&quot;</span> <span class="o">......</span> <span class="p">})</span>
</pre></div>
</div>
<p>This <a class="reference internal" href="python-api.html#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> can then be assigned to a table or the top-level
tree sequence e.g.
<a class="reference internal" href="python-api.html#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TableCollection</span><span class="p">(</span><span class="n">sequence_length</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">schema</span>
</pre></div>
</div>
<p>This will overwrite any existing schema. Note that this will not validate any existing
metadata against the new schema. Now that the table has a schema calls to
<a class="reference internal" href="python-api.html#tskit.IndividualTable.add_row" title="tskit.IndividualTable.add_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_row</span></code></a> will validate and encode the metadata:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="s2">&quot;Bob1234&quot;</span><span class="p">,</span> <span class="s2">&quot;pcr&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">})</span>
</pre></div>
</div>
<p>and if we try to add metadata that doesn’t fit the schema, such as accidentally using a
string instead of a proper Python boolean, we’ll get an error:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="s2">&quot;Bob1234&quot;</span><span class="p">,</span> <span class="s2">&quot;pcr&quot;</span><span class="p">:</span> <span class="s2">&quot;false&quot;</span><span class="p">})</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Traceback</span> <span class="p">(</span><span class="n">most</span> <span class="n">recent</span> <span class="n">call</span> <span class="n">last</span><span class="p">):</span>
  <span class="n">File</span> <span class="s2">&quot;/home/benj/projects/tskit/python/tskit/metadata.py&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">181</span><span class="p">,</span> <span class="ow">in</span> <span class="n">validate_and_encode_row</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_row</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
  <span class="n">File</span> <span class="s2">&quot;/home/benj/projects/tskit/env/lib/python3.8/site-packages/jsonschema/validators.py&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">353</span><span class="p">,</span> <span class="ow">in</span> <span class="n">validate</span>
    <span class="k">raise</span> <span class="n">error</span>
<span class="n">jsonschema</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">ValidationError</span><span class="p">:</span> <span class="s1">&#39;false&#39;</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">of</span> <span class="nb">type</span> <span class="s1">&#39;boolean&#39;</span>

<span class="n">Failed</span> <span class="n">validating</span> <span class="s1">&#39;type&#39;</span> <span class="ow">in</span> <span class="n">schema</span><span class="p">[</span><span class="s1">&#39;properties&#39;</span><span class="p">][</span><span class="s1">&#39;pcr&#39;</span><span class="p">]:</span>
    <span class="p">{</span><span class="s1">&#39;description&#39;</span><span class="p">:</span> <span class="s1">&#39;Was PCR used on this sample&#39;</span><span class="p">,</span>
     <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;PCR Used&#39;</span><span class="p">,</span>
     <span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;boolean&#39;</span><span class="p">}</span>

<span class="n">On</span> <span class="n">instance</span><span class="p">[</span><span class="s1">&#39;pcr&#39;</span><span class="p">]:</span>
    <span class="s1">&#39;false&#39;</span>

<span class="n">The</span> <span class="n">above</span> <span class="n">exception</span> <span class="n">was</span> <span class="n">the</span> <span class="n">direct</span> <span class="n">cause</span> <span class="n">of</span> <span class="n">the</span> <span class="n">following</span> <span class="n">exception</span><span class="p">:</span>

<span class="n">Traceback</span> <span class="p">(</span><span class="n">most</span> <span class="n">recent</span> <span class="n">call</span> <span class="n">last</span><span class="p">):</span>
  <span class="n">File</span> <span class="s2">&quot;&lt;stdin&gt;&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">1</span><span class="p">,</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
  <span class="n">File</span> <span class="s2">&quot;/home/benj/projects/tskit/python/tskit/tables.py&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">409</span><span class="p">,</span> <span class="ow">in</span> <span class="n">add_row</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata_schema</span><span class="o">.</span><span class="n">validate_and_encode_row</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span>
  <span class="n">File</span> <span class="s2">&quot;/home/benj/projects/tskit/python/tskit/metadata.py&quot;</span><span class="p">,</span> <span class="n">line</span> <span class="mi">183</span><span class="p">,</span> <span class="ow">in</span> <span class="n">validate_and_encode_row</span>
    <span class="k">raise</span> <span class="n">exceptions</span><span class="o">.</span><span class="n">MetadataValidationError</span> <span class="kn">from</span> <span class="nn">ve</span>
<span class="n">tskit</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">MetadataValidationError</span>
</pre></div>
</div>
<p>To set the top-level metadata, just assign it. Validation and encoding happen as specified by the
top-level metadata schema (Note that provenance info should go in <a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">Provenance</span></a>
not metadata):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;mean_coverage&quot;</span><span class="p">:</span> <span class="mf">200.5</span><span class="p">}</span>
</pre></div>
</div>
<p>To modify a schema, for example to add a key; first get the dict representation, modify
then write back:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">schema_dict</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">metadata_schema</span><span class="o">.</span><span class="n">schema</span>
<span class="n">schema_dict</span><span class="p">[</span><span class="s2">&quot;properties&quot;</span><span class="p">][</span><span class="s2">&quot;newKey&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">}</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="n">schema_dict</span><span class="p">)</span>
</pre></div>
</div>
<p>To modify the metadata of rows in tables use the <a class="reference internal" href="#sec-tutorial-metadata-bulk"><span class="std std-ref">Metadata for bulk table methods</span></a>.</p>
</div>
<div class="section" id="viewing-raw-metadata">
<h3>Viewing raw metadata<a class="headerlink" href="#viewing-raw-metadata" title="Permalink to this headline">¶</a></h3>
<p>If you need to see the raw (i.e., bytes) metadata, you just need to remove the
schema, for instance:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;accession&#39;</span><span class="p">:</span> <span class="s1">&#39;abc123&#39;</span><span class="p">,</span> <span class="s1">&#39;newKey&#39;</span><span class="p">:</span><span class="mi">25</span><span class="p">}})</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span>
<span class="c1"># {&#39;accession&#39;: &#39;abc123&#39;, &#39;newKey&#39;: &#39;25&#39;}</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">MetadataSchema</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span>
<span class="c1"># b&#39;abc123\x00\x00\x19&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="metadata-for-bulk-table-methods">
<span id="sec-tutorial-metadata-bulk"></span><h3>Metadata for bulk table methods<a class="headerlink" href="#metadata-for-bulk-table-methods" title="Permalink to this headline">¶</a></h3>
<p>In the interests of efficiency each table’s
<a class="reference internal" href="python-api.html#tskit.NodeTable.packset_metadata" title="tskit.NodeTable.packset_metadata"><code class="xref py py-meth docutils literal notranslate"><span class="pre">packset_metadata</span></code></a> method, as well as the more
general <a class="reference internal" href="python-api.html#tskit.NodeTable.set_columns" title="tskit.NodeTable.set_columns"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_columns</span></code></a> and
<a class="reference internal" href="python-api.html#tskit.NodeTable.append_columns" title="tskit.NodeTable.append_columns"><code class="xref py py-meth docutils literal notranslate"><span class="pre">append_columns</span></code></a> do not attempt to validate or
encode metadata. You can call
<a class="reference internal" href="python-api.html#tskit.MetadataSchema.validate_and_encode_row" title="tskit.MetadataSchema.validate_and_encode_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MetadataSchema.validate_and_encode_row</span></code></a>
directly to prepare metadata for these methods:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">metadata_column</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">{</span><span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="s2">&quot;etho1234&quot;</span><span class="p">,</span> <span class="s2">&quot;pcr&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
    <span class="p">{</span><span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="s2">&quot;richard1235&quot;</span><span class="p">,</span> <span class="s2">&quot;pcr&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">},</span>
    <span class="p">{</span><span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="s2">&quot;albert1236&quot;</span><span class="p">,</span> <span class="s2">&quot;pcr&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
<span class="p">]</span>
<span class="n">encoded_metadata_column</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">table</span><span class="o">.</span><span class="n">metadata_schema</span><span class="o">.</span><span class="n">validate_and_encode_row</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">metadata_column</span>
<span class="p">]</span>
<span class="n">metadata</span><span class="p">,</span> <span class="n">metadata_offset</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">pack_bytes</span><span class="p">(</span><span class="n">encoded_metadata_column</span><span class="p">)</span>
<span class="n">table</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span> <span class="n">metadata_offset</span><span class="o">=</span><span class="n">metadata_offset</span><span class="p">)</span>
</pre></div>
</div>
<p>Or if all columns do not need to be set:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">table</span><span class="o">.</span><span class="n">packset_metadata</span><span class="p">(</span>
    <span class="p">[</span><span class="n">table</span><span class="o">.</span><span class="n">metadata_schema</span><span class="o">.</span><span class="n">validate_and_encode_row</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">metadata_column</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="binary-metadata">
<span id="sec-tutorial-metadata-binary"></span><h3>Binary metadata<a class="headerlink" href="#binary-metadata" title="Permalink to this headline">¶</a></h3>
<p>To disable the validation and encoding of metadata and store raw bytes pass <code class="docutils literal notranslate"><span class="pre">None</span></code> to
<a class="reference internal" href="python-api.html#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">table</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="n">table</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="sa">b</span><span class="s2">&quot;SOME CUSTOM BYTES #!@&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="sa">b</span><span class="s1">&#39;SOME CUSTOM BYTES #!@&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="calculating-ld">
<h2>Calculating LD<a class="headerlink" href="#calculating-ld" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">tskit</span></code> API provides methods to efficiently calculate
population genetics statistics. For example, the <a class="reference internal" href="python-api.html#tskit.LdCalculator" title="tskit.LdCalculator"><code class="xref py py-class docutils literal notranslate"><span class="pre">LdCalculator</span></code></a>
class allows us to compute pairwise <a class="reference external" href="https://en.wikipedia.org/wiki/Linkage_disequilibrium">linkage disequilibrium</a> coefficients.
Here we use the <a class="reference internal" href="python-api.html#tskit.LdCalculator.r2_matrix" title="tskit.LdCalculator.r2_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">LdCalculator.r2_matrix()</span></code></a> method to easily make an
LD plot using <a class="reference external" href="http://matplotlib.org/">matplotlib</a>. (Thanks to
the excellent <a class="reference external" href="http://scikit-allel.readthedocs.io/en/latest/index.html">scikit-allel</a>
for the basic <a class="reference external" href="http://scikit-allel.readthedocs.io/en/latest/_modules/allel/stats/ld.html#plot_pairwise_ld">plotting code</a>
used here.)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>
<span class="kn">import</span> <span class="nn">tskit</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">pyplot</span>


<span class="k">def</span> <span class="nf">ld_matrix_example</span><span class="p">():</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ld_calc</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">LdCalculator</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
    <span class="n">A</span> <span class="o">=</span> <span class="n">ld_calc</span><span class="o">.</span><span class="n">r2_matrix</span><span class="p">()</span>
    <span class="c1"># Now plot this matrix.</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">A</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.dpi&quot;</span><span class="p">]</span>
    <span class="n">x</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">pyplot</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">(</span><span class="n">pad</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">im</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">A</span><span class="p">,</span> <span class="n">interpolation</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="s2">&quot;top&quot;</span><span class="p">,</span> <span class="s2">&quot;bottom&quot;</span><span class="p">,</span> <span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="s2">&quot;right&quot;</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">pyplot</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im</span><span class="p">,</span> <span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">pyplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;ld.svg&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/ld.svg"><img alt="An example LD matrix plot." src="_images/ld.svg" width="800px" /></a>
</div>
<div class="section" id="working-with-threads">
<span id="sec-tutorial-threads"></span><h2>Working with threads<a class="headerlink" href="#working-with-threads" title="Permalink to this headline">¶</a></h2>
<p>When performing large calculations it’s often useful to split the
work over multiple processes or threads. The <code class="docutils literal notranslate"><span class="pre">tskit</span></code> API can
be used without issues across multiple processes, and the Python
<a class="reference external" href="https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">multiprocessing</span></code></a> module often provides a very effective way to
work with many replicate simulations in parallel.</p>
<p>When we wish to work with a single very large dataset, however, threads can
offer better resource usage because of the shared memory space. The Python
<a class="reference external" href="https://docs.python.org/3/library/threading.html#module-threading" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">threading</span></code></a> library gives a very simple interface to lightweight CPU
threads and allows us to perform several CPU intensive tasks in parallel. The
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> API is designed to allow multiple threads to work in parallel when
CPU intensive tasks are being undertaken.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the CPython implementation the <a class="reference external" href="https://wiki.python.org/moin/GlobalInterpreterLock">Global Interpreter Lock</a> ensures that
only one thread executes Python bytecode at one time. This means that
Python code does not parallelise well across threads, but avoids a large
number of nasty pitfalls associated with multiple threads updating
data structures in parallel. Native C extensions like <code class="docutils literal notranslate"><span class="pre">numpy</span></code> and <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
release the GIL while expensive tasks are being performed, therefore
allowing these calculations to proceed in parallel.</p>
</div>
<p>In the following example we wish to find all mutations that are in approximate
LD (<span class="math notranslate nohighlight">\(r^2 \geq 0.5\)</span>) with a given set of mutations. We parallelise this
by splitting the input array between a number of threads, and use the
<a class="reference internal" href="python-api.html#tskit.LdCalculator.r2_array" title="tskit.LdCalculator.r2_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">LdCalculator.r2_array()</span></code></a> method to compute the <span class="math notranslate nohighlight">\(r^2\)</span> value
both up and downstream of each focal mutation, filter out those that
exceed our threshold, and store the results in a dictionary. We also
use the very cool <a class="reference external" href="https://pypi.python.org/pypi/tqdm">tqdm</a> module to give us a
progress bar on this computation.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">threading</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">tqdm</span>
<span class="kn">import</span> <span class="nn">msprime</span>
<span class="kn">import</span> <span class="nn">tskit</span>


<span class="k">def</span> <span class="nf">find_ld_sites</span><span class="p">(</span>
    <span class="n">tree_sequence</span><span class="p">,</span> <span class="n">focal_mutations</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="mf">1e6</span><span class="p">,</span> <span class="n">r2_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">num_threads</span><span class="o">=</span><span class="mi">8</span>
<span class="p">):</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">progress_bar</span> <span class="o">=</span> <span class="n">tqdm</span><span class="o">.</span><span class="n">tqdm</span><span class="p">(</span><span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">focal_mutations</span><span class="p">))</span>
    <span class="n">num_threads</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">num_threads</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">focal_mutations</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">thread_worker</span><span class="p">(</span><span class="n">thread_index</span><span class="p">):</span>
        <span class="n">ld_calc</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">LdCalculator</span><span class="p">(</span><span class="n">tree_sequence</span><span class="p">)</span>
        <span class="n">chunk_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">focal_mutations</span><span class="p">)</span> <span class="o">/</span> <span class="n">num_threads</span><span class="p">))</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">thread_index</span> <span class="o">*</span> <span class="n">chunk_size</span>
        <span class="k">for</span> <span class="n">focal_mutation</span> <span class="ow">in</span> <span class="n">focal_mutations</span><span class="p">[</span><span class="n">start</span> <span class="p">:</span> <span class="n">start</span> <span class="o">+</span> <span class="n">chunk_size</span><span class="p">]:</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">ld_calc</span><span class="o">.</span><span class="n">r2_array</span><span class="p">(</span>
                <span class="n">focal_mutation</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="n">max_distance</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="n">tskit</span><span class="o">.</span><span class="n">REVERSE</span>
            <span class="p">)</span>
            <span class="n">rev_indexes</span> <span class="o">=</span> <span class="n">focal_mutation</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">a</span> <span class="o">&gt;=</span> <span class="n">r2_threshold</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">a</span> <span class="o">=</span> <span class="n">ld_calc</span><span class="o">.</span><span class="n">r2_array</span><span class="p">(</span>
                <span class="n">focal_mutation</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="n">max_distance</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="n">tskit</span><span class="o">.</span><span class="n">FORWARD</span>
            <span class="p">)</span>
            <span class="n">fwd_indexes</span> <span class="o">=</span> <span class="n">focal_mutation</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">a</span> <span class="o">&gt;=</span> <span class="n">r2_threshold</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">indexes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">rev_indexes</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">fwd_indexes</span><span class="p">))</span>
            <span class="n">results</span><span class="p">[</span><span class="n">focal_mutation</span><span class="p">]</span> <span class="o">=</span> <span class="n">indexes</span>
            <span class="n">progress_bar</span><span class="o">.</span><span class="n">update</span><span class="p">()</span>

    <span class="n">threads</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">threading</span><span class="o">.</span><span class="n">Thread</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">thread_worker</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">j</span><span class="p">,))</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_threads</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">threads</span><span class="p">:</span>
        <span class="n">t</span><span class="o">.</span><span class="n">start</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">threads</span><span class="p">:</span>
        <span class="n">t</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
    <span class="n">progress_bar</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">results</span>


<span class="k">def</span> <span class="nf">threads_example</span><span class="p">():</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span>
        <span class="n">sample_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
        <span class="n">Ne</span><span class="o">=</span><span class="mf">1e4</span><span class="p">,</span>
        <span class="n">length</span><span class="o">=</span><span class="mf">1e7</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">2e-8</span><span class="p">,</span>
        <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">2e-8</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
            <span class="n">mutation</span> <span class="o">=</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">counts</span><span class="p">[</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">num_samples</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">node</span><span class="p">)</span>
    <span class="n">doubletons</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">counts</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">find_ld_sites</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">doubletons</span><span class="p">,</span> <span class="n">num_threads</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Found LD sites for&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">),</span> <span class="s2">&quot;doubleton sites out of&quot;</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, we first simulate 1000 samples of 10 megabases and find all
doubleton mutations in the resulting tree sequence. We then call the
<code class="docutils literal notranslate"><span class="pre">find_ld_sites()</span></code> function to find all mutations that are within 1 megabase
of these doubletons and have an <span class="math notranslate nohighlight">\(r^2\)</span> statistic of greater than 0.5.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">find_ld_sites()</span></code> function performs these calculations in parallel using
8 threads. The real work is done in the nested <code class="docutils literal notranslate"><span class="pre">thread_worker()</span></code> function,
which is called once by each thread. In the thread worker, we first allocate an
instance of the <a class="reference internal" href="python-api.html#tskit.LdCalculator" title="tskit.LdCalculator"><code class="xref py py-class docutils literal notranslate"><span class="pre">LdCalculator</span></code></a> class. (It is <strong>critically important</strong>
that each thread has its own instance of <a class="reference internal" href="python-api.html#tskit.LdCalculator" title="tskit.LdCalculator"><code class="xref py py-class docutils literal notranslate"><span class="pre">LdCalculator</span></code></a>, as the threads
will not work efficiently otherwise.) After this, each thread works out the
slice of the input array that it is responsible for, and then iterates over
each focal mutation in turn. After the <span class="math notranslate nohighlight">\(r^2\)</span> values have been calculated,
we then find the indexes of the mutations corresponding to values greater than
0.5 using <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html#numpy.nonzero" title="(in NumPy v1.19)"><code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.nonzero()</span></code></a>. Finally, the thread stores the resulting array
of mutation indexes in the <code class="docutils literal notranslate"><span class="pre">results</span></code> dictionary, and moves on to the next
focal mutation.</p>
<p>Running this example we get:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">threads_example</span><span class="p">()</span>
<span class="go">100%|████████████████████████████████████████████████| 4045/4045 [00:09&lt;00:00, 440.29it/s]</span>
<span class="go">Found LD sites for 4045 doubleton mutations out of 60100</span>
</pre></div>
</div>
</div>
<div class="section" id="parsimony">
<span id="sec-tutorial-parsimony"></span><h2>Parsimony<a class="headerlink" href="#parsimony" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="python-api.html#tskit.Tree.map_mutations" title="tskit.Tree.map_mutations"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.map_mutations()</span></code></a> method finds a parsimonious explanation for a
set of discrete character observations on the samples in a tree using classical phylogenetic
algorithms.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="n">alleles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="s2">&quot;green&quot;</span><span class="p">]</span>
<span class="n">genotypes</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
<span class="n">node_colours</span> <span class="o">=</span> <span class="p">{</span><span class="n">j</span><span class="p">:</span> <span class="n">alleles</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)}</span>
<span class="n">ancestral_state</span><span class="p">,</span> <span class="n">mutations</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">map_mutations</span><span class="p">(</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">alleles</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Ancestral state = &quot;</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="p">)</span>
<span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation: node = </span><span class="si">{</span><span class="n">mut</span><span class="o">.</span><span class="n">node</span><span class="si">}</span><span class="s2"> derived_state = </span><span class="si">{</span><span class="n">mut</span><span class="o">.</span><span class="n">derived_state</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="s2">&quot;_static/parsimony1.svg&quot;</span><span class="p">,</span> <span class="n">node_colours</span><span class="o">=</span><span class="n">node_colours</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/parsimony1.svg" src="_images/parsimony1.svg" /><p>We get:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Ancestral</span> <span class="n">state</span> <span class="o">=</span>  <span class="n">red</span>
<span class="n">Mutation</span><span class="p">:</span> <span class="n">node</span> <span class="o">=</span> <span class="mi">5</span> <span class="n">derived_state</span> <span class="o">=</span> <span class="n">green</span>
<span class="n">Mutation</span><span class="p">:</span> <span class="n">node</span> <span class="o">=</span> <span class="mi">4</span> <span class="n">derived_state</span> <span class="o">=</span> <span class="n">blue</span>
</pre></div>
</div>
<p>So, the algorithm has concluded, quite reasonably, that the most parsimonious
description of this state is that the ancestral state is red and there was
a mutation to blue and green over nodes 4 and 5.</p>
<div class="section" id="building-tables">
<h3>Building tables<a class="headerlink" href="#building-tables" title="Permalink to this headline">¶</a></h3>
<p>One of the main uses of <a class="reference internal" href="python-api.html#tskit.Tree.map_mutations" title="tskit.Tree.map_mutations"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.map_mutations()</span></code></a> is to position mutations on a tree
to encode observed data. In the following example we show how a set
of tables can be updated using the <a class="reference internal" href="python-api.html#sec-tables-api"><span class="std std-ref">Tables API</span></a>; here we
infer the location of mutations in an simulated tree sequence, and recompute
the node and edge tables exactly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">23</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">mutate</span><span class="p">(</span>
    <span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">InfiniteSites</span><span class="p">(</span><span class="n">msprime</span><span class="o">.</span><span class="n">NUCLEOTIDES</span><span class="p">),</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="n">tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="c1"># Reinfer the sites and mutations from the variants.</span>
<span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
<span class="k">for</span> <span class="n">var</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="n">ancestral_state</span><span class="p">,</span> <span class="n">mutations</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">map_mutations</span><span class="p">(</span><span class="n">var</span><span class="o">.</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">var</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">var</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="n">ancestral_state</span><span class="p">)</span>
    <span class="n">parent_offset</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
        <span class="n">parent</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">parent</span>
        <span class="k">if</span> <span class="n">parent</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
            <span class="n">parent</span> <span class="o">+=</span> <span class="n">parent_offset</span>
        <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span>
            <span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">node</span><span class="o">=</span><span class="n">mutation</span><span class="o">.</span><span class="n">node</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">parent</span><span class="p">,</span>
            <span class="n">derived_state</span><span class="o">=</span><span class="n">mutation</span><span class="o">.</span><span class="n">derived_state</span><span class="p">)</span>

<span class="k">assert</span> <span class="n">tables</span><span class="o">.</span><span class="n">sites</span> <span class="o">==</span> <span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">sites</span>
<span class="k">assert</span> <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span> <span class="o">==</span> <span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">mutations</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="p">)</span>
</pre></div>
</div>
<p>The output is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">id</span>      <span class="n">position</span>        <span class="n">ancestral_state</span> <span class="n">metadata</span>
<span class="mi">0</span>       <span class="mf">0.25849808</span>      <span class="n">T</span>
<span class="mi">1</span>       <span class="mf">0.26682728</span>      <span class="n">G</span>
<span class="mi">2</span>       <span class="mf">0.32053644</span>      <span class="n">C</span>
<span class="mi">3</span>       <span class="mf">0.40730783</span>      <span class="n">T</span>
<span class="mi">4</span>       <span class="mf">0.49856117</span>      <span class="n">G</span>
<span class="mi">5</span>       <span class="mf">0.58679698</span>      <span class="n">A</span>
<span class="mi">6</span>       <span class="mf">0.61927097</span>      <span class="n">A</span>
<span class="mi">7</span>       <span class="mf">0.71975423</span>      <span class="n">T</span>
<span class="mi">8</span>       <span class="mf">0.94773061</span>      <span class="n">C</span>
<span class="nb">id</span>      <span class="n">site</span>    <span class="n">node</span>    <span class="n">derived_state</span>   <span class="n">parent</span>  <span class="n">metadata</span>
<span class="mi">0</span>       <span class="mi">0</span>       <span class="mi">8</span>       <span class="n">C</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">1</span>       <span class="mi">1</span>       <span class="mi">7</span>       <span class="n">C</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">2</span>       <span class="mi">2</span>       <span class="mi">4</span>       <span class="n">A</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">3</span>       <span class="mi">3</span>       <span class="mi">9</span>       <span class="n">C</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">4</span>       <span class="mi">4</span>       <span class="mi">8</span>       <span class="n">A</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">5</span>       <span class="mi">5</span>       <span class="mi">8</span>       <span class="n">T</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">6</span>       <span class="mi">6</span>       <span class="mi">5</span>       <span class="n">G</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">7</span>       <span class="mi">7</span>       <span class="mi">8</span>       <span class="n">A</span>       <span class="o">-</span><span class="mi">1</span>
<span class="mi">8</span>       <span class="mi">8</span>       <span class="mi">3</span>       <span class="n">T</span>       <span class="o">-</span><span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="missing-data">
<h3>Missing data<a class="headerlink" href="#missing-data" title="Permalink to this headline">¶</a></h3>
<p>The Fitch parsimony algorithm in <a class="reference internal" href="python-api.html#tskit.Tree.map_mutations" title="tskit.Tree.map_mutations"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.map_mutations()</span></code></a> can also take missing data
into account when finding a set of parsimonious state transitions. We do this by
specifying the special value <code class="docutils literal notranslate"><span class="pre">-1</span></code> as the state, which is treated by the algorithm as
“could be anything”.</p>
<p>For example, here we state that sample 0 is missing, and use the colour white to indicate
this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="n">alleles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="s2">&quot;green&quot;</span><span class="p">,</span> <span class="s2">&quot;white&quot;</span><span class="p">]</span>
<span class="n">genotypes</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
<span class="n">node_colours</span> <span class="o">=</span> <span class="p">{</span><span class="n">j</span><span class="p">:</span> <span class="n">alleles</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)}</span>
<span class="n">ancestral_state</span><span class="p">,</span> <span class="n">mutations</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">map_mutations</span><span class="p">(</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">alleles</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Ancestral state = &quot;</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="p">)</span>
<span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation: node = </span><span class="si">{</span><span class="n">mut</span><span class="o">.</span><span class="n">node</span><span class="si">}</span><span class="s2"> derived_state = </span><span class="si">{</span><span class="n">mut</span><span class="o">.</span><span class="n">derived_state</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="s2">&quot;_static/parsimony2.svg&quot;</span><span class="p">,</span> <span class="n">node_colours</span><span class="o">=</span><span class="n">node_colours</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/parsimony2.svg" src="_images/parsimony2.svg" /><p>As before, we get:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Ancestral</span> <span class="n">state</span> <span class="o">=</span>  <span class="n">red</span>
<span class="n">Mutation</span><span class="p">:</span> <span class="n">node</span> <span class="o">=</span> <span class="mi">5</span> <span class="n">derived_state</span> <span class="o">=</span> <span class="n">green</span>
<span class="n">Mutation</span><span class="p">:</span> <span class="n">node</span> <span class="o">=</span> <span class="mi">4</span> <span class="n">derived_state</span> <span class="o">=</span> <span class="n">blue</span>
</pre></div>
</div>
<p>The algorithm decided, again, quite reasonably, that the most parsimonious explanation
for the input data is the same as before. Thus, if we used this information to fill
out mutation table as above, we would impute the missing value for 0 as red.</p>
<p>The output of the algorithm can be a little surprising at times. Consider this example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="n">alleles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="s2">&quot;white&quot;</span><span class="p">]</span>
<span class="n">genotypes</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">node_colours</span> <span class="o">=</span> <span class="p">{</span><span class="n">j</span><span class="p">:</span> <span class="n">alleles</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)}</span>
<span class="n">ancestral_state</span><span class="p">,</span> <span class="n">mutations</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">map_mutations</span><span class="p">(</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">alleles</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Ancestral state = &quot;</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="p">)</span>
<span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mutation: node = </span><span class="si">{</span><span class="n">mut</span><span class="o">.</span><span class="n">node</span><span class="si">}</span><span class="s2"> derived_state = </span><span class="si">{</span><span class="n">mut</span><span class="o">.</span><span class="n">derived_state</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="s2">&quot;_static/parsimony3.svg&quot;</span><span class="p">,</span> <span class="n">node_colours</span><span class="o">=</span><span class="n">node_colours</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/parsimony3.svg" src="_images/parsimony3.svg" /><p>The output we get is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Ancestral</span> <span class="n">state</span> <span class="o">=</span>  <span class="n">red</span>
<span class="n">Mutation</span><span class="p">:</span> <span class="n">node</span> <span class="o">=</span> <span class="mi">6</span> <span class="n">derived_state</span> <span class="o">=</span> <span class="n">blue</span>
</pre></div>
</div>
<p>Note that this is putting a mutation to blue over node 6, <strong>not</strong> node 0 as
we might expect. Thus, we impute here that node 1 is blue. It is important
to remember that the algorithm is minimising the number of state transitions;
this may not correspond always to what we might consider the most parsimonious
explanation.</p>
</div>
</div>
<div class="section" id="computing-statistics">
<span id="sec-tutorial-stats"></span><h2>Computing statistics<a class="headerlink" href="#computing-statistics" title="Permalink to this headline">¶</a></h2>
<p>Tskit provides an extensive and flexible interface for computing population
genetic statistics, which is documented in detail in the <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">general statistics</span></a> section. This tutorial aims to give a quick overview of
how the APIs work how to use them effectively.</p>
<p>First, lets simulate a tree sequence to work with which has roughly human
parameters for 10 thousand samples and 10Mb chromosomes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span>
    <span class="mi">10</span><span class="o">**</span><span class="mi">4</span><span class="p">,</span> <span class="n">Ne</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">4</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">7</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
</pre></div>
</div>
<p>We end up with 36K trees 39K segregating sites. We’d now like to compute some statistics on
this dataset.</p>
<div class="section" id="one-way-statistics">
<h3>One-way statistics<a class="headerlink" href="#one-way-statistics" title="Permalink to this headline">¶</a></h3>
<p>We refer to statistics that are defined with respect to a single set of
samples as “one-way”. An example of such a statistic is diversity, which
is computed using the <a class="reference internal" href="python-api.html#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.diversity()</span></code></a> method:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Average diversity per unit sequence length = </span><span class="si">{:.3G}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="n">Average</span> <span class="n">diversity</span> <span class="n">per</span> <span class="n">unit</span> <span class="n">sequence</span> <span class="n">length</span> <span class="o">=</span> <span class="mf">0.000401</span>
</pre></div>
</div>
<p>This tells the average diversity across the whole sequence and returns a single
number. We’ll usually want to compute statistics in
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a> along the genome and we
use the <code class="docutils literal notranslate"><span class="pre">windows</span></code> argument to do this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">windows</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">(</span><span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">windows</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="p">[</span>       <span class="mf">0.</span>  <span class="mf">2500000.</span>  <span class="mf">5000000.</span>  <span class="mf">7500000.</span> <span class="mf">10000000.</span><span class="p">]</span>
<span class="p">[</span><span class="mf">0.00041602</span> <span class="mf">0.00039112</span> <span class="mf">0.00041554</span> <span class="mf">0.00038329</span><span class="p">]</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">windows</span></code> argument takes a numpy array specifying the breakpoints
along the genome. Here, we use numpy to create four equally spaced windows
of size 2.5 megabases (the windows array contains k + 1 elements to define
k windows). Because we have asked for values in windows, tskit now returns
a numpy array rather than a single value. (See
<a class="reference internal" href="stats.html#sec-stats-output-dimensions"><span class="std std-ref">Output dimensions</span></a> for a full description of how the output
dimensions of statistics are determined by the <code class="docutils literal notranslate"><span class="pre">windows</span></code> argument.)</p>
<p>Suppose we wanted to compute diversity within a specific subset of samples.
We can do this using the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()[:</span><span class="mi">100</span><span class="p">]</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">(</span><span class="n">sample_sets</span><span class="o">=</span><span class="n">A</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="mf">0.00040166573737371227</span>
</pre></div>
</div>
<p>Here, we’ve computed the average diversity within the first hundred samples across
the whole genome. As we’ve not specified any windows, this is again a single value.</p>
<p>We can also compute diversity in <em>multiple</em> sample sets at the same time by providing
a list of sample sets as an argument:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()[:</span><span class="mi">100</span><span class="p">]</span>
<span class="n">B</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()[</span><span class="mi">100</span><span class="p">:</span><span class="mi">200</span><span class="p">]</span>
<span class="n">C</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()[</span><span class="mi">200</span><span class="p">:</span><span class="mi">300</span><span class="p">]</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">(</span><span class="n">sample_sets</span><span class="o">=</span><span class="p">[</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="p">[</span><span class="mf">0.00040167</span> <span class="mf">0.00040008</span> <span class="mf">0.00040103</span><span class="p">]</span>
</pre></div>
</div>
<p>Because we’ve computed multiple statistics concurrently, tskit returns a numpy array
of these statistics. We have asked for diversity within three different sample sets,
and tskit therefore returns an array with three values. (In general, the
dimensions of the input determines the dimensions of the output: see
<a class="reference internal" href="stats.html#sec-stats-output-dimensions"><span class="std std-ref">Output dimensions</span></a> for a detailed description of the rules.)</p>
<p>We can also compute multiple statistics in multiple windows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">(</span><span class="n">sample_sets</span><span class="o">=</span><span class="p">[</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">],</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;shape = &quot;</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="n">shape</span> <span class="o">=</span>  <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="p">[[</span><span class="mf">0.0004139</span>  <span class="mf">0.00041567</span> <span class="mf">0.00041774</span><span class="p">]</span>
 <span class="p">[</span><span class="mf">0.00039148</span> <span class="mf">0.00039152</span> <span class="mf">0.00038997</span><span class="p">]</span>
 <span class="p">[</span><span class="mf">0.00042019</span> <span class="mf">0.00041039</span> <span class="mf">0.00041475</span><span class="p">]</span>
 <span class="p">[</span><span class="mf">0.0003811</span>  <span class="mf">0.00038274</span> <span class="mf">0.00038166</span><span class="p">]]</span>
</pre></div>
</div>
<p>We have computed diversity within three different sample sets across four
genomic windows, and our output is therefore a 2D numpy array with four
rows and three columns: each row contains the diversity values within
A, B and C for a particular window.</p>
</div>
<div class="section" id="multi-way-statistics">
<h3>Multi-way statistics<a class="headerlink" href="#multi-way-statistics" title="Permalink to this headline">¶</a></h3>
<p>Many population genetic statistics compare multiple sets of samples to
each other. For example, the <a class="reference internal" href="python-api.html#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.divergence()</span></code></a> method computes
the divergence between two subsets of samples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()[:</span><span class="mi">100</span><span class="p">]</span>
<span class="n">B</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()[:</span><span class="mi">100</span><span class="p">]</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">divergence</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="mf">0.00039764908000000676</span>
</pre></div>
</div>
<p>The divergence between two sets of samples A and B is a single number,
and we we again return a single floating point value as the result. We can also
compute this in windows along the genome, as before:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">divergence</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">],</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="p">[</span><span class="mf">0.00040976</span> <span class="mf">0.00038756</span> <span class="mf">0.00041599</span> <span class="mf">0.00037728</span><span class="p">]</span>
</pre></div>
</div>
<p>Again, as we have defined four genomic windows along the sequence, the result is
numpy array with four values.</p>
<p>A powerful feature of tskit’s stats API is that we can compute the divergences
between multiple sets of samples simultaneously using the <code class="docutils literal notranslate"><span class="pre">indexes</span></code> argument:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">divergence</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">],</span> <span class="n">indexes</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="p">[</span><span class="mf">0.00039765</span> <span class="mf">0.00040181</span><span class="p">]</span>
</pre></div>
</div>
<p>Here, we’ve specified three sample sets A, B and C and we’ve computed the
divergences between A and B,  and between A and C. The <code class="docutils literal notranslate"><span class="pre">indexes</span></code> argument is used
to specify which pairs of sets we are interested in. In this example
we’ve computed two different divergence values and the output is therefore
a numpy array of length 2.</p>
<p>As before, we can combine computing multiple statistics in multiple windows
to return a 2D numpy array:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">windows</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">divergence</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">],</span> <span class="n">indexes</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)],</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="p">[[</span><span class="mf">0.00040976</span> <span class="mf">0.0004161</span> <span class="p">]</span>
 <span class="p">[</span><span class="mf">0.00038756</span> <span class="mf">0.00039025</span><span class="p">]</span>
 <span class="p">[</span><span class="mf">0.00041599</span> <span class="mf">0.00041847</span><span class="p">]</span>
 <span class="p">[</span><span class="mf">0.00037728</span> <span class="mf">0.0003824</span> <span class="p">]]</span>
</pre></div>
</div>
<p>Each row again corresponds to a window, which contains the average divergence
values between the chosen sets.</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">indexes</span></code> parameter is 1D array, we interpret this as specifying
a single statistic and remove the empty outer dimension:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">divergence</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">],</span> <span class="n">indexes</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="mf">0.00039764908000000676</span>
</pre></div>
</div>
<p>It’s important to note that we don’t <strong>have</strong> to remove empty dimensions: tskit
will only do this if you explicitly ask it to. Here, for example, we can keep the
output as an array with one value if we wish:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">divergence</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">,</span> <span class="n">C</span><span class="p">],</span> <span class="n">indexes</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="p">[</span><span class="n">Output</span><span class="p">]</span>

<span class="p">[</span><span class="mf">0.00039765</span><span class="p">]</span>
</pre></div>
</div>
<p>Please see <a class="reference internal" href="stats.html#sec-stats-sample-sets"><span class="std std-ref">Sample sets and indexes</span></a> for a
full description of the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments.</p>
</div>
</div>
<div class="section" id="allele-frequency-spectra">
<span id="sec-tutorial-afs"></span><h2>Allele frequency spectra<a class="headerlink" href="#allele-frequency-spectra" title="Permalink to this headline">¶</a></h2>
<p>The allele frequency spectrum is a fundamental tool in population genetics, and
tskit provides a flexible and powerful approach to computing such spectra.
Suppose we have simulated the following tree and site table:</p>
<img alt="_images/afs1.svg" src="_images/afs1.svg" /><div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">id</span>      <span class="n">position</span>        <span class="n">ancestral_state</span> <span class="n">metadata</span>
<span class="mi">0</span>       <span class="mf">0.30043643</span>      <span class="mi">0</span>
<span class="mi">1</span>       <span class="mf">0.32220794</span>      <span class="mi">0</span>
<span class="mi">2</span>       <span class="mf">0.36507027</span>      <span class="mi">0</span>
<span class="mi">3</span>       <span class="mf">0.50940255</span>      <span class="mi">0</span>
<span class="mi">4</span>       <span class="mf">0.51327137</span>      <span class="mi">0</span>
<span class="mi">5</span>       <span class="mf">0.51400861</span>      <span class="mi">0</span>
<span class="mi">6</span>       <span class="mf">0.54796110</span>      <span class="mi">0</span>
<span class="mi">7</span>       <span class="mf">0.75929404</span>      <span class="mi">0</span>
<span class="mi">8</span>       <span class="mf">0.80591800</span>      <span class="mi">0</span>
<span class="mi">9</span>       <span class="mf">0.92324208</span>      <span class="mi">0</span>
</pre></div>
</div>
<p>Computing the allele frequency spectrum is then easy:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">afs</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">allele_frequency_spectrum</span><span class="p">(</span><span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>which looks like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[[</span><span class="mf">0.</span> <span class="mf">2.</span> <span class="mf">6.</span> <span class="mf">1.</span> <span class="mf">1.</span> <span class="mf">0.</span> <span class="mf">0.</span><span class="p">]]</span>
</pre></div>
</div>
<p>This tells us that we have two singletons, six doubletons and one 3-ton and
one 4-ton. Note
that the first element of the returned AFS array does <em>not</em> correspond to
the singletons (see below for why). Because we have simulated these mutations,
we know the ancestral and derived states we have set <code class="docutils literal notranslate"><span class="pre">polarised</span></code> to True. We
can get the “folded” AFS by setting polarised to False. Because we want simple
counts here and not averaged values, we set <code class="docutils literal notranslate"><span class="pre">span_normalise=False</span></code>: by
default, windowed statistics are divided by the sequence length, so they are
comparable between windows.</p>
<p>The returned value here is actually a 2D array, and this is because we can
also perform these computations in windows along the genome:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">afs</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">allele_frequency_spectrum</span><span class="p">(</span>
    <span class="n">windows</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">afs</span><span class="p">)</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[[</span><span class="mf">0.</span> <span class="mf">1.</span> <span class="mf">1.</span> <span class="mf">1.</span> <span class="mf">0.</span> <span class="mf">0.</span> <span class="mf">0.</span><span class="p">]</span>
 <span class="p">[</span><span class="mf">0.</span> <span class="mf">1.</span> <span class="mf">5.</span> <span class="mf">0.</span> <span class="mf">1.</span> <span class="mf">0.</span> <span class="mf">0.</span><span class="p">]]</span>
</pre></div>
</div>
<p>This time, we’ve asked for the number of sites at each frequency in two
equal windows. Now we can see that in the first half of the sequence we
have three sites (compare with the site table above): one singleton,
one doubleton and one tripleton.</p>
<div class="section" id="joint-spectra">
<h3>Joint spectra<a class="headerlink" href="#joint-spectra" title="Permalink to this headline">¶</a></h3>
<p>We can also compute allele frequencies within multiple sets of samples,
the <em>joint allele frequency spectra</em>.</p>
<img alt="_images/afs2.svg" src="_images/afs2.svg" /><p>Here we’ve marked the samples as either blue or green (we can imagine
these belonging to different populations, for example). We can then compute
the joint AFS based on these two sets:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">afs</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">allele_frequency_spectrum</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">]],</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">afs</span><span class="p">)</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[[[</span><span class="mf">0.</span> <span class="mf">2.</span> <span class="mf">0.</span> <span class="mf">0.</span><span class="p">]</span>
  <span class="p">[</span><span class="mf">0.</span> <span class="mf">6.</span> <span class="mf">0.</span> <span class="mf">0.</span><span class="p">]</span>
  <span class="p">[</span><span class="mf">0.</span> <span class="mf">1.</span> <span class="mf">1.</span> <span class="mf">0.</span><span class="p">]</span>
  <span class="p">[</span><span class="mf">0.</span> <span class="mf">0.</span> <span class="mf">0.</span> <span class="mf">0.</span><span class="p">]]]</span>
</pre></div>
</div>
<p>Now, each window in our AFS is a 2D numpy array, where each dimension
corresponds to frequencies within the different sets. So, we see for example
that there are six sites that are singletons in both sets, 1 site
that is a doubleton in both sets, and 2 sites that singletons in [1, 4, 5]
and not present in the other sample set.</p>
</div>
<div class="section" id="branch-length-spectra">
<h3>Branch length spectra<a class="headerlink" href="#branch-length-spectra" title="Permalink to this headline">¶</a></h3>
<p>Up to now we’ve used the <a class="reference internal" href="python-api.html#tskit.TreeSequence.allele_frequency_spectrum" title="tskit.TreeSequence.allele_frequency_spectrum"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.allele_frequency_spectrum()</span></code></a> method
to summarise the number of sites that occur at different frequencies. We can also
use this approach to compute the total branch lengths subtending a given
number of samples by setting <code class="docutils literal notranslate"><span class="pre">mode=&quot;branch&quot;</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">afs</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">allele_frequency_spectrum</span><span class="p">(</span>
    <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;branch&quot;</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">span_normalise</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">afs</span><span class="p">)</span>
</pre></div>
</div>
<p>giving:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[[</span><span class="mf">0.</span> <span class="mf">4.86089166</span> <span class="mf">5.39638988</span> <span class="mf">2.55239269</span> <span class="mf">2.07444286</span> <span class="mf">0.</span> <span class="mf">0.</span><span class="p">]]</span>
</pre></div>
</div>
<p>Thus, the total branch length over example one sample is 4.86, over two is
5.39, and so on.</p>
</div>
<div class="section" id="zeroth-and-final-entries-in-the-afs">
<span id="sec-tutorial-afs-zeroth-entry"></span><h3>Zeroth and final entries in the AFS<a class="headerlink" href="#zeroth-and-final-entries-in-the-afs" title="Permalink to this headline">¶</a></h3>
<p>The zeroth element of the AFS is significant when we are working with
sample sets that are a subset of all samples in the tree sequence.
For example, in the following we compute the AFS within the sample set
[0, 1, 2]:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">afs</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">allele_frequency_spectrum</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]],</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;branch&quot;</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">afs</span><span class="p">)</span>
</pre></div>
</div>
<p>getting:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[[</span><span class="mf">4.33184862</span> <span class="mf">5.30022646</span> <span class="mf">5.252042</span>   <span class="mf">0.</span>        <span class="p">]]</span>
</pre></div>
</div>
<p>Thus, the total branch length over 0, 1 and 2 is 5.3, and over pairs from this set
is 5.25. What does the zeroth value of 4.33 signify? This is the total branch length
over all samples that are <strong>not</strong> in this sample set. By including this value, we
maintain the property that for each tree, the sum of the AFS for any sample set
is always equal to the total branch length. For example, here we compute:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;sum afs          = &quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">afs</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;total branch len = &quot;</span><span class="p">,</span> <span class="n">tree</span><span class="o">.</span><span class="n">total_branch_length</span><span class="p">)</span>
</pre></div>
</div>
<p>getting:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">sum</span> <span class="n">afs</span>          <span class="o">=</span>  <span class="mf">14.884117086717392</span>
<span class="n">total</span> <span class="n">branch</span> <span class="nb">len</span> <span class="o">=</span>  <span class="mf">14.884117086717396</span>
</pre></div>
</div>
<p>The final entry of the AFS is similar: it counts alleles (for mode=”site”) or
branches (for mode=”branch”) that are ancestral to all of the given sample set,
but are still polymorphic in the entire set of samples of the tree sequence.
Note, however, that alleles fixed among all the samples, e.g., ones above
the root of the tree, will not be included.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="changelogs.html" class="btn btn-neutral float-right" title="Changelogs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="development.html" class="btn btn-neutral float-left" title="Development" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2018-2020, Tskit developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>